// rclnodejs/generated/ur_msgs/ur_msgs__msg__RobotStateRTMsg.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Float64Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float64.js');
const RobotStateRTMsgRefStruct = StructType({
  time: primitiveTypes.float64,
  q_target: Float64Wrapper.refObjectArrayType,
  qd_target: Float64Wrapper.refObjectArrayType,
  qdd_target: Float64Wrapper.refObjectArrayType,
  i_target: Float64Wrapper.refObjectArrayType,
  m_target: Float64Wrapper.refObjectArrayType,
  q_actual: Float64Wrapper.refObjectArrayType,
  qd_actual: Float64Wrapper.refObjectArrayType,
  i_actual: Float64Wrapper.refObjectArrayType,
  tool_acc_values: Float64Wrapper.refObjectArrayType,
  tcp_force: Float64Wrapper.refObjectArrayType,
  tool_vector: Float64Wrapper.refObjectArrayType,
  tcp_speed: Float64Wrapper.refObjectArrayType,
  digital_input_bits: primitiveTypes.float64,
  motor_temperatures: Float64Wrapper.refObjectArrayType,
  controller_timer: primitiveTypes.float64,
  test_value: primitiveTypes.float64,
  robot_mode: primitiveTypes.float64,
  joint_modes: Float64Wrapper.refObjectArrayType,
});
const RobotStateRTMsgRefArray = ArrayType(RobotStateRTMsgRefStruct);
const RobotStateRTMsgRefStructArray = StructType({
  data: RobotStateRTMsgRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class RobotStateRTMsgWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new RobotStateRTMsgRefStruct(other._refObject.toObject());
      this._wrapperFields.q_target = Float64Wrapper.createArray();
      this._wrapperFields.q_target.copy(other._wrapperFields.q_target);
      this._wrapperFields.qd_target = Float64Wrapper.createArray();
      this._wrapperFields.qd_target.copy(other._wrapperFields.qd_target);
      this._wrapperFields.qdd_target = Float64Wrapper.createArray();
      this._wrapperFields.qdd_target.copy(other._wrapperFields.qdd_target);
      this._wrapperFields.i_target = Float64Wrapper.createArray();
      this._wrapperFields.i_target.copy(other._wrapperFields.i_target);
      this._wrapperFields.m_target = Float64Wrapper.createArray();
      this._wrapperFields.m_target.copy(other._wrapperFields.m_target);
      this._wrapperFields.q_actual = Float64Wrapper.createArray();
      this._wrapperFields.q_actual.copy(other._wrapperFields.q_actual);
      this._wrapperFields.qd_actual = Float64Wrapper.createArray();
      this._wrapperFields.qd_actual.copy(other._wrapperFields.qd_actual);
      this._wrapperFields.i_actual = Float64Wrapper.createArray();
      this._wrapperFields.i_actual.copy(other._wrapperFields.i_actual);
      this._wrapperFields.tool_acc_values = Float64Wrapper.createArray();
      this._wrapperFields.tool_acc_values.copy(other._wrapperFields.tool_acc_values);
      this._wrapperFields.tcp_force = Float64Wrapper.createArray();
      this._wrapperFields.tcp_force.copy(other._wrapperFields.tcp_force);
      this._wrapperFields.tool_vector = Float64Wrapper.createArray();
      this._wrapperFields.tool_vector.copy(other._wrapperFields.tool_vector);
      this._wrapperFields.tcp_speed = Float64Wrapper.createArray();
      this._wrapperFields.tcp_speed.copy(other._wrapperFields.tcp_speed);
      this._wrapperFields.motor_temperatures = Float64Wrapper.createArray();
      this._wrapperFields.motor_temperatures.copy(other._wrapperFields.motor_temperatures);
      this._wrapperFields.joint_modes = Float64Wrapper.createArray();
      this._wrapperFields.joint_modes.copy(other._wrapperFields.joint_modes);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new RobotStateRTMsgRefStruct();
    this._wrapperFields.q_target = Float64Wrapper.createArray();
    this._wrapperFields.qd_target = Float64Wrapper.createArray();
    this._wrapperFields.qdd_target = Float64Wrapper.createArray();
    this._wrapperFields.i_target = Float64Wrapper.createArray();
    this._wrapperFields.m_target = Float64Wrapper.createArray();
    this._wrapperFields.q_actual = Float64Wrapper.createArray();
    this._wrapperFields.qd_actual = Float64Wrapper.createArray();
    this._wrapperFields.i_actual = Float64Wrapper.createArray();
    this._wrapperFields.tool_acc_values = Float64Wrapper.createArray();
    this._wrapperFields.tcp_force = Float64Wrapper.createArray();
    this._wrapperFields.tool_vector = Float64Wrapper.createArray();
    this._wrapperFields.tcp_speed = Float64Wrapper.createArray();
    this._wrapperFields.motor_temperatures = Float64Wrapper.createArray();
    this._wrapperFields.joint_modes = Float64Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new RobotStateRTMsgWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new RobotStateRTMsgArrayWrapper();
  }
  static get ArrayType() {
    return RobotStateRTMsgArrayWrapper;
  }
  static get refObjectArrayType() {
    return RobotStateRTMsgRefStructArray;
  }
  static get refObjectType() {
    return RobotStateRTMsgRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.q_target.fill(Float64Array.from([]));
    }
    this._wrapperFields.q_target.freeze(own);
    this._refObject.q_target = this._wrapperFields.q_target.refObject;
    if (own) {
      this._wrapperFields.qd_target.fill(Float64Array.from([]));
    }
    this._wrapperFields.qd_target.freeze(own);
    this._refObject.qd_target = this._wrapperFields.qd_target.refObject;
    if (own) {
      this._wrapperFields.qdd_target.fill(Float64Array.from([]));
    }
    this._wrapperFields.qdd_target.freeze(own);
    this._refObject.qdd_target = this._wrapperFields.qdd_target.refObject;
    if (own) {
      this._wrapperFields.i_target.fill(Float64Array.from([]));
    }
    this._wrapperFields.i_target.freeze(own);
    this._refObject.i_target = this._wrapperFields.i_target.refObject;
    if (own) {
      this._wrapperFields.m_target.fill(Float64Array.from([]));
    }
    this._wrapperFields.m_target.freeze(own);
    this._refObject.m_target = this._wrapperFields.m_target.refObject;
    if (own) {
      this._wrapperFields.q_actual.fill(Float64Array.from([]));
    }
    this._wrapperFields.q_actual.freeze(own);
    this._refObject.q_actual = this._wrapperFields.q_actual.refObject;
    if (own) {
      this._wrapperFields.qd_actual.fill(Float64Array.from([]));
    }
    this._wrapperFields.qd_actual.freeze(own);
    this._refObject.qd_actual = this._wrapperFields.qd_actual.refObject;
    if (own) {
      this._wrapperFields.i_actual.fill(Float64Array.from([]));
    }
    this._wrapperFields.i_actual.freeze(own);
    this._refObject.i_actual = this._wrapperFields.i_actual.refObject;
    if (own) {
      this._wrapperFields.tool_acc_values.fill(Float64Array.from([]));
    }
    this._wrapperFields.tool_acc_values.freeze(own);
    this._refObject.tool_acc_values = this._wrapperFields.tool_acc_values.refObject;
    if (own) {
      this._wrapperFields.tcp_force.fill(Float64Array.from([]));
    }
    this._wrapperFields.tcp_force.freeze(own);
    this._refObject.tcp_force = this._wrapperFields.tcp_force.refObject;
    if (own) {
      this._wrapperFields.tool_vector.fill(Float64Array.from([]));
    }
    this._wrapperFields.tool_vector.freeze(own);
    this._refObject.tool_vector = this._wrapperFields.tool_vector.refObject;
    if (own) {
      this._wrapperFields.tcp_speed.fill(Float64Array.from([]));
    }
    this._wrapperFields.tcp_speed.freeze(own);
    this._refObject.tcp_speed = this._wrapperFields.tcp_speed.refObject;
    if (own) {
      this._wrapperFields.motor_temperatures.fill(Float64Array.from([]));
    }
    this._wrapperFields.motor_temperatures.freeze(own);
    this._refObject.motor_temperatures = this._wrapperFields.motor_temperatures.refObject;
    if (own) {
      this._wrapperFields.joint_modes.fill(Float64Array.from([]));
    }
    this._wrapperFields.joint_modes.freeze(own);
    this._refObject.joint_modes = this._wrapperFields.joint_modes.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.time = refObject.time;
    this._wrapperFields.q_target.copyRefObject(refObject.q_target);
    this._wrapperFields.qd_target.copyRefObject(refObject.qd_target);
    this._wrapperFields.qdd_target.copyRefObject(refObject.qdd_target);
    this._wrapperFields.i_target.copyRefObject(refObject.i_target);
    this._wrapperFields.m_target.copyRefObject(refObject.m_target);
    this._wrapperFields.q_actual.copyRefObject(refObject.q_actual);
    this._wrapperFields.qd_actual.copyRefObject(refObject.qd_actual);
    this._wrapperFields.i_actual.copyRefObject(refObject.i_actual);
    this._wrapperFields.tool_acc_values.copyRefObject(refObject.tool_acc_values);
    this._wrapperFields.tcp_force.copyRefObject(refObject.tcp_force);
    this._wrapperFields.tool_vector.copyRefObject(refObject.tool_vector);
    this._wrapperFields.tcp_speed.copyRefObject(refObject.tcp_speed);
    this._refObject.digital_input_bits = refObject.digital_input_bits;
    this._wrapperFields.motor_temperatures.copyRefObject(refObject.motor_temperatures);
    this._refObject.controller_timer = refObject.controller_timer;
    this._refObject.test_value = refObject.test_value;
    this._refObject.robot_mode = refObject.robot_mode;
    this._wrapperFields.joint_modes.copyRefObject(refObject.joint_modes);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.q_target.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.q_target);
      if (!Float64Wrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.q_target, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.qd_target.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.qd_target);
      if (!Float64Wrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.qd_target, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.qdd_target.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.qdd_target);
      if (!Float64Wrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.qdd_target, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.i_target.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.i_target);
      if (!Float64Wrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.i_target, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.m_target.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.m_target);
      if (!Float64Wrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.m_target, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.q_actual.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.q_actual);
      if (!Float64Wrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.q_actual, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.qd_actual.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.qd_actual);
      if (!Float64Wrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.qd_actual, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.i_actual.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.i_actual);
      if (!Float64Wrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.i_actual, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.tool_acc_values.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.tool_acc_values);
      if (!Float64Wrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.tool_acc_values, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.tcp_force.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.tcp_force);
      if (!Float64Wrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.tcp_force, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.tool_vector.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.tool_vector);
      if (!Float64Wrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.tool_vector, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.tcp_speed.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.tcp_speed);
      if (!Float64Wrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.tcp_speed, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.motor_temperatures.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.motor_temperatures);
      if (!Float64Wrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.motor_temperatures, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.joint_modes.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.joint_modes);
      if (!Float64Wrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.joint_modes, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    RobotStateRTMsgWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'ur_msgs', subFolder: 'msg', interfaceName: 'RobotStateRTMsg'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get time() {
    return this._refObject.time;
  }
  set time(value) {
    this._refObject.time = value;
  }
  get q_target() {
    return this._wrapperFields['q_target'].data;
  }
  set q_target(value) {
    this._wrapperFields['q_target'].fill(value);
  }
  get qd_target() {
    return this._wrapperFields['qd_target'].data;
  }
  set qd_target(value) {
    this._wrapperFields['qd_target'].fill(value);
  }
  get qdd_target() {
    return this._wrapperFields['qdd_target'].data;
  }
  set qdd_target(value) {
    this._wrapperFields['qdd_target'].fill(value);
  }
  get i_target() {
    return this._wrapperFields['i_target'].data;
  }
  set i_target(value) {
    this._wrapperFields['i_target'].fill(value);
  }
  get m_target() {
    return this._wrapperFields['m_target'].data;
  }
  set m_target(value) {
    this._wrapperFields['m_target'].fill(value);
  }
  get q_actual() {
    return this._wrapperFields['q_actual'].data;
  }
  set q_actual(value) {
    this._wrapperFields['q_actual'].fill(value);
  }
  get qd_actual() {
    return this._wrapperFields['qd_actual'].data;
  }
  set qd_actual(value) {
    this._wrapperFields['qd_actual'].fill(value);
  }
  get i_actual() {
    return this._wrapperFields['i_actual'].data;
  }
  set i_actual(value) {
    this._wrapperFields['i_actual'].fill(value);
  }
  get tool_acc_values() {
    return this._wrapperFields['tool_acc_values'].data;
  }
  set tool_acc_values(value) {
    this._wrapperFields['tool_acc_values'].fill(value);
  }
  get tcp_force() {
    return this._wrapperFields['tcp_force'].data;
  }
  set tcp_force(value) {
    this._wrapperFields['tcp_force'].fill(value);
  }
  get tool_vector() {
    return this._wrapperFields['tool_vector'].data;
  }
  set tool_vector(value) {
    this._wrapperFields['tool_vector'].fill(value);
  }
  get tcp_speed() {
    return this._wrapperFields['tcp_speed'].data;
  }
  set tcp_speed(value) {
    this._wrapperFields['tcp_speed'].fill(value);
  }
  get digital_input_bits() {
    return this._refObject.digital_input_bits;
  }
  set digital_input_bits(value) {
    this._refObject.digital_input_bits = value;
  }
  get motor_temperatures() {
    return this._wrapperFields['motor_temperatures'].data;
  }
  set motor_temperatures(value) {
    this._wrapperFields['motor_temperatures'].fill(value);
  }
  get controller_timer() {
    return this._refObject.controller_timer;
  }
  set controller_timer(value) {
    this._refObject.controller_timer = value;
  }
  get test_value() {
    return this._refObject.test_value;
  }
  set test_value(value) {
    this._refObject.test_value = value;
  }
  get robot_mode() {
    return this._refObject.robot_mode;
  }
  set robot_mode(value) {
    this._refObject.robot_mode = value;
  }
  get joint_modes() {
    return this._wrapperFields['joint_modes'].data;
  }
  set joint_modes(value) {
    this._wrapperFields['joint_modes'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new RobotStateRTMsgRefStruct(refObject.toObject());
    this._wrapperFields.q_target.copyRefObject(this._refObject.q_target);
    this._wrapperFields.qd_target.copyRefObject(this._refObject.qd_target);
    this._wrapperFields.qdd_target.copyRefObject(this._refObject.qdd_target);
    this._wrapperFields.i_target.copyRefObject(this._refObject.i_target);
    this._wrapperFields.m_target.copyRefObject(this._refObject.m_target);
    this._wrapperFields.q_actual.copyRefObject(this._refObject.q_actual);
    this._wrapperFields.qd_actual.copyRefObject(this._refObject.qd_actual);
    this._wrapperFields.i_actual.copyRefObject(this._refObject.i_actual);
    this._wrapperFields.tool_acc_values.copyRefObject(this._refObject.tool_acc_values);
    this._wrapperFields.tcp_force.copyRefObject(this._refObject.tcp_force);
    this._wrapperFields.tool_vector.copyRefObject(this._refObject.tool_vector);
    this._wrapperFields.tcp_speed.copyRefObject(this._refObject.tcp_speed);
    this._wrapperFields.motor_temperatures.copyRefObject(this._refObject.motor_temperatures);
    this._wrapperFields.joint_modes.copyRefObject(this._refObject.joint_modes);
  }
  copy(other) {
    this._refObject = new RobotStateRTMsgRefStruct(other._refObject.toObject());
    this._wrapperFields.q_target.copy(other._wrapperFields.q_target);
    this._wrapperFields.qd_target.copy(other._wrapperFields.qd_target);
    this._wrapperFields.qdd_target.copy(other._wrapperFields.qdd_target);
    this._wrapperFields.i_target.copy(other._wrapperFields.i_target);
    this._wrapperFields.m_target.copy(other._wrapperFields.m_target);
    this._wrapperFields.q_actual.copy(other._wrapperFields.q_actual);
    this._wrapperFields.qd_actual.copy(other._wrapperFields.qd_actual);
    this._wrapperFields.i_actual.copy(other._wrapperFields.i_actual);
    this._wrapperFields.tool_acc_values.copy(other._wrapperFields.tool_acc_values);
    this._wrapperFields.tcp_force.copy(other._wrapperFields.tcp_force);
    this._wrapperFields.tool_vector.copy(other._wrapperFields.tool_vector);
    this._wrapperFields.tcp_speed.copy(other._wrapperFields.tcp_speed);
    this._wrapperFields.motor_temperatures.copy(other._wrapperFields.motor_temperatures);
    this._wrapperFields.joint_modes.copy(other._wrapperFields.joint_modes);
  }
  static get classType() {
    return RobotStateRTMsgWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"time","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"q_target","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"qd_target","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"qdd_target","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"i_target","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"m_target","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"q_actual","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"qd_actual","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"i_actual","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"tool_acc_values","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"tcp_force","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"tool_vector","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"tcp_speed","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"digital_input_bits","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"motor_temperatures","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"controller_timer","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"test_value","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"robot_mode","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"joint_modes","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"ur_msgs","type":"RobotStateRTMsg","stringUpperBound":null,"isPrimitiveType":false},"msgName":"RobotStateRTMsg"};
  }
  hasMember(name) {
    let memberNames = ["time","q_target","qd_target","qdd_target","i_target","m_target","q_actual","qd_actual","i_actual","tool_acc_values","tcp_force","tool_vector","tcp_speed","digital_input_bits","motor_temperatures","controller_timer","test_value","robot_mode","joint_modes"];
    return memberNames.indexOf(name) !== -1;
  }
}
class RobotStateRTMsgArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof RobotStateRTMsgWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new RobotStateRTMsgWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RobotStateRTMsgArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RobotStateRTMsgArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new RobotStateRTMsgRefArray(size);
    this._refObject = new RobotStateRTMsgRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new RobotStateRTMsgWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof RobotStateRTMsgArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "RobotStateRTMsgArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      RobotStateRTMsgWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return RobotStateRTMsgWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return RobotStateRTMsgArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = RobotStateRTMsgWrapper;
