// rclnodejs/generated/ur_msgs/ur_msgs__srv__SetPayload_Request.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Vector3Wrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Vector3.js');
const SetPayload_RequestRefStruct = StructType({
  mass: primitiveTypes.float32,
  center_of_gravity: Vector3Wrapper.refObjectType,
});
const SetPayload_RequestRefArray = ArrayType(SetPayload_RequestRefStruct);
const SetPayload_RequestRefStructArray = StructType({
  data: SetPayload_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class SetPayload_RequestWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new SetPayload_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.center_of_gravity =  new Vector3Wrapper(other._wrapperFields.center_of_gravity);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new SetPayload_RequestRefStruct();
    this._wrapperFields.center_of_gravity = new Vector3Wrapper();
  }
  static createFromRefObject(refObject) {
    let self = new SetPayload_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SetPayload_RequestArrayWrapper();
  }
  static get ArrayType() {
    return SetPayload_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return SetPayload_RequestRefStructArray;
  }
  static get refObjectType() {
    return SetPayload_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.center_of_gravity.freeze(own);
    this._refObject.center_of_gravity = this._wrapperFields.center_of_gravity.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.mass = refObject.mass;
    this._wrapperFields.center_of_gravity.copyRefObject(refObject.center_of_gravity);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    Vector3Wrapper.freeStruct(refObject.center_of_gravity);
  }
  static destroyRawROS(msg) {
    SetPayload_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'ur_msgs', subFolder: 'srv', interfaceName: 'SetPayload_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get mass() {
    return this._refObject.mass;
  }
  set mass(value) {
    this._refObject.mass = value;
  }
  get center_of_gravity() {
    return this._wrapperFields.center_of_gravity;
  }
  set center_of_gravity(value) {
    if (value instanceof Vector3Wrapper) {
      this._wrapperFields.center_of_gravity.copy(value);
    } else {
      this._wrapperFields.center_of_gravity.copy(new Vector3Wrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new SetPayload_RequestRefStruct(refObject.toObject());
    this._wrapperFields.center_of_gravity.copyRefObject(this._refObject.center_of_gravity);
  }
  copy(other) {
    this._refObject = new SetPayload_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.center_of_gravity.copy(other._wrapperFields.center_of_gravity);
  }
  static get classType() {
    return SetPayload_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"mass","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"center_of_gravity","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Vector3","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"ur_msgs","type":"SetPayload_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SetPayload_Request"};
  }
  hasMember(name) {
    let memberNames = ["mass","center_of_gravity"];
    return memberNames.indexOf(name) !== -1;
  }
}
class SetPayload_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SetPayload_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SetPayload_RequestWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetPayload_RequestArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetPayload_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new SetPayload_RequestRefArray(size);
    this._refObject = new SetPayload_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SetPayload_RequestWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SetPayload_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SetPayload_RequestArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SetPayload_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SetPayload_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SetPayload_RequestArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = SetPayload_RequestWrapper;
