// rclnodejs/generated/ur_msgs/ur_msgs__srv__SetFrictionModelParameters_Request.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const FrictionModelParametersWrapper = require('../../generated/ur_msgs/ur_msgs__msg__FrictionModelParameters.js');
const SetFrictionModelParameters_RequestRefStruct = StructType({
  parameters: FrictionModelParametersWrapper.refObjectType,
});
const SetFrictionModelParameters_RequestRefArray = ArrayType(SetFrictionModelParameters_RequestRefStruct);
const SetFrictionModelParameters_RequestRefStructArray = StructType({
  data: SetFrictionModelParameters_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class SetFrictionModelParameters_RequestWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new SetFrictionModelParameters_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.parameters =  new FrictionModelParametersWrapper(other._wrapperFields.parameters);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new SetFrictionModelParameters_RequestRefStruct();
    this._wrapperFields.parameters = new FrictionModelParametersWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new SetFrictionModelParameters_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SetFrictionModelParameters_RequestArrayWrapper();
  }
  static get ArrayType() {
    return SetFrictionModelParameters_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return SetFrictionModelParameters_RequestRefStructArray;
  }
  static get refObjectType() {
    return SetFrictionModelParameters_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.parameters.freeze(own);
    this._refObject.parameters = this._wrapperFields.parameters.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.parameters.copyRefObject(refObject.parameters);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    FrictionModelParametersWrapper.freeStruct(refObject.parameters);
  }
  static destroyRawROS(msg) {
    SetFrictionModelParameters_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'ur_msgs', subFolder: 'srv', interfaceName: 'SetFrictionModelParameters_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get parameters() {
    return this._wrapperFields.parameters;
  }
  set parameters(value) {
    if (value instanceof FrictionModelParametersWrapper) {
      this._wrapperFields.parameters.copy(value);
    } else {
      this._wrapperFields.parameters.copy(new FrictionModelParametersWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new SetFrictionModelParameters_RequestRefStruct(refObject.toObject());
    this._wrapperFields.parameters.copyRefObject(this._refObject.parameters);
  }
  copy(other) {
    this._refObject = new SetFrictionModelParameters_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.parameters.copy(other._wrapperFields.parameters);
  }
  static get classType() {
    return SetFrictionModelParameters_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"parameters","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"ur_msgs","type":"FrictionModelParameters","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"ur_msgs","type":"SetFrictionModelParameters_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SetFrictionModelParameters_Request"};
  }
  hasMember(name) {
    let memberNames = ["parameters"];
    return memberNames.indexOf(name) !== -1;
  }
}
class SetFrictionModelParameters_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SetFrictionModelParameters_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SetFrictionModelParameters_RequestWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetFrictionModelParameters_RequestArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetFrictionModelParameters_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new SetFrictionModelParameters_RequestRefArray(size);
    this._refObject = new SetFrictionModelParameters_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SetFrictionModelParameters_RequestWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SetFrictionModelParameters_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SetFrictionModelParameters_RequestArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SetFrictionModelParameters_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SetFrictionModelParameters_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SetFrictionModelParameters_RequestArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = SetFrictionModelParameters_RequestWrapper;
