// rclnodejs/generated/ur_msgs/ur_msgs__srv__SetIO_Request.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const SetIO_RequestRefStruct = StructType({
  fun: primitiveTypes.int8,
  pin: primitiveTypes.int8,
  state: primitiveTypes.float32,
});
const SetIO_RequestRefArray = ArrayType(SetIO_RequestRefStruct);
const SetIO_RequestRefStructArray = StructType({
  data: SetIO_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class SetIO_RequestWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new SetIO_RequestRefStruct(other._refObject.toObject());
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new SetIO_RequestRefStruct();
  }
  static createFromRefObject(refObject) {
    let self = new SetIO_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SetIO_RequestArrayWrapper();
  }
  static get ArrayType() {
    return SetIO_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return SetIO_RequestRefStructArray;
  }
  static get refObjectType() {
    return SetIO_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.fun = refObject.fun;
    this._refObject.pin = refObject.pin;
    this._refObject.state = refObject.state;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destroyRawROS(msg) {
    SetIO_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'ur_msgs', subFolder: 'srv', interfaceName: 'SetIO_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get fun() {
    return this._refObject.fun;
  }
  set fun(value) {
    this._refObject.fun = value;
  }
  get pin() {
    return this._refObject.pin;
  }
  set pin(value) {
    this._refObject.pin = value;
  }
  get state() {
    return this._refObject.state;
  }
  set state(value) {
    this._refObject.state = value;
  }
  copyRefObject(refObject) {
    this._refObject = new SetIO_RequestRefStruct(refObject.toObject());
  }
  copy(other) {
    this._refObject = new SetIO_RequestRefStruct(other._refObject.toObject());
  }
  static get classType() {
    return SetIO_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"int8","name":"PIN_AOUT0","value":0},{"type":"int8","name":"PIN_AOUT1","value":1},{"type":"int8","name":"PIN_DOUT0","value":0},{"type":"int8","name":"PIN_DOUT1","value":1},{"type":"int8","name":"PIN_DOUT2","value":2},{"type":"int8","name":"PIN_DOUT3","value":3},{"type":"int8","name":"PIN_DOUT4","value":4},{"type":"int8","name":"PIN_DOUT5","value":5},{"type":"int8","name":"PIN_DOUT6","value":6},{"type":"int8","name":"PIN_DOUT7","value":7},{"type":"int8","name":"PIN_CONF_OUT0","value":8},{"type":"int8","name":"PIN_CONF_OUT1","value":9},{"type":"int8","name":"PIN_CONF_OUT2","value":10},{"type":"int8","name":"PIN_CONF_OUT3","value":11},{"type":"int8","name":"PIN_CONF_OUT4","value":12},{"type":"int8","name":"PIN_CONF_OUT5","value":13},{"type":"int8","name":"PIN_CONF_OUT6","value":14},{"type":"int8","name":"PIN_CONF_OUT7","value":15},{"type":"int8","name":"PIN_TOOL_DOUT0","value":16},{"type":"int8","name":"PIN_TOOL_DOUT1","value":17},{"type":"int8","name":"FUN_SET_DIGITAL_OUT","value":1},{"type":"int8","name":"FUN_SET_FLAG","value":2},{"type":"int8","name":"FUN_SET_ANALOG_OUT","value":3},{"type":"int8","name":"FUN_SET_TOOL_VOLTAGE","value":4},{"type":"int8","name":"STATE_OFF","value":0},{"type":"int8","name":"STATE_ON","value":1},{"type":"int8","name":"STATE_TOOL_VOLTAGE_0V","value":0},{"type":"int8","name":"STATE_TOOL_VOLTAGE_12V","value":12},{"type":"int8","name":"STATE_TOOL_VOLTAGE_24V","value":24}],"fields":[{"name":"fun","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pin","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"state","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"ur_msgs","type":"SetIO_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SetIO_Request"};
  }
  hasMember(name) {
    let memberNames = ["fun","pin","state"];
    return memberNames.indexOf(name) !== -1;
  }
}
class SetIO_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SetIO_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SetIO_RequestWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetIO_RequestArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetIO_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new SetIO_RequestRefArray(size);
    this._refObject = new SetIO_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SetIO_RequestWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SetIO_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SetIO_RequestArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SetIO_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SetIO_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SetIO_RequestArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
Object.defineProperty(SetIO_RequestWrapper, "PIN_AOUT0", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetIO_RequestWrapper, "PIN_AOUT1", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetIO_RequestWrapper, "PIN_DOUT0", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetIO_RequestWrapper, "PIN_DOUT1", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetIO_RequestWrapper, "PIN_DOUT2", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetIO_RequestWrapper, "PIN_DOUT3", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetIO_RequestWrapper, "PIN_DOUT4", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetIO_RequestWrapper, "PIN_DOUT5", {value: 5, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetIO_RequestWrapper, "PIN_DOUT6", {value: 6, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetIO_RequestWrapper, "PIN_DOUT7", {value: 7, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetIO_RequestWrapper, "PIN_CONF_OUT0", {value: 8, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetIO_RequestWrapper, "PIN_CONF_OUT1", {value: 9, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetIO_RequestWrapper, "PIN_CONF_OUT2", {value: 10, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetIO_RequestWrapper, "PIN_CONF_OUT3", {value: 11, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetIO_RequestWrapper, "PIN_CONF_OUT4", {value: 12, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetIO_RequestWrapper, "PIN_CONF_OUT5", {value: 13, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetIO_RequestWrapper, "PIN_CONF_OUT6", {value: 14, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetIO_RequestWrapper, "PIN_CONF_OUT7", {value: 15, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetIO_RequestWrapper, "PIN_TOOL_DOUT0", {value: 16, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetIO_RequestWrapper, "PIN_TOOL_DOUT1", {value: 17, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetIO_RequestWrapper, "FUN_SET_DIGITAL_OUT", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetIO_RequestWrapper, "FUN_SET_FLAG", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetIO_RequestWrapper, "FUN_SET_ANALOG_OUT", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetIO_RequestWrapper, "FUN_SET_TOOL_VOLTAGE", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetIO_RequestWrapper, "STATE_OFF", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetIO_RequestWrapper, "STATE_ON", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetIO_RequestWrapper, "STATE_TOOL_VOLTAGE_0V", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetIO_RequestWrapper, "STATE_TOOL_VOLTAGE_12V", {value: 12, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetIO_RequestWrapper, "STATE_TOOL_VOLTAGE_24V", {value: 24, writable: false, enumerable: true, configurable: true});
module.exports = SetIO_RequestWrapper;
