// rclnodejs/generated/ur_msgs/ur_msgs__action__FollowJointTrajectoryUntil_Result.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const FollowJointTrajectoryUntil_ResultRefStruct = StructType({
  error_code: primitiveTypes.int32,
  until_condition_result: primitiveTypes.int32,
  error_string: primitiveTypes.string,
});
const FollowJointTrajectoryUntil_ResultRefArray = ArrayType(FollowJointTrajectoryUntil_ResultRefStruct);
const FollowJointTrajectoryUntil_ResultRefStructArray = StructType({
  data: FollowJointTrajectoryUntil_ResultRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class FollowJointTrajectoryUntil_ResultWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new FollowJointTrajectoryUntil_ResultRefStruct(other._refObject.toObject());
      this._wrapperFields.error_string =  new StringWrapper(other._wrapperFields.error_string);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new FollowJointTrajectoryUntil_ResultRefStruct();
    this._wrapperFields.error_string = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new FollowJointTrajectoryUntil_ResultWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new FollowJointTrajectoryUntil_ResultArrayWrapper();
  }
  static get ArrayType() {
    return FollowJointTrajectoryUntil_ResultArrayWrapper;
  }
  static get refObjectArrayType() {
    return FollowJointTrajectoryUntil_ResultRefStructArray;
  }
  static get refObjectType() {
    return FollowJointTrajectoryUntil_ResultRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.error_string.freeze(own);
    }
    this._refObject.error_string = this._wrapperFields.error_string.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.error_code = refObject.error_code;
    this._refObject.until_condition_result = refObject.until_condition_result;
    this._wrapperFields.error_string.data = refObject.error_string.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.error_string);
  }
  static destroyRawROS(msg) {
    FollowJointTrajectoryUntil_ResultWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'ur_msgs', subFolder: 'action', interfaceName: 'FollowJointTrajectoryUntil_Result'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get error_code() {
    return this._refObject.error_code;
  }
  set error_code(value) {
    this._refObject.error_code = value;
  }
  get until_condition_result() {
    return this._refObject.until_condition_result;
  }
  set until_condition_result(value) {
    this._refObject.until_condition_result = value;
  }
  get error_string() {
    return this._wrapperFields.error_string.data;
  }
  set error_string(value) {
    this._wrapperFields.error_string.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new FollowJointTrajectoryUntil_ResultRefStruct(refObject.toObject());
    this._wrapperFields.error_string.copyRefObject(this._refObject.error_string);
  }
  copy(other) {
    this._refObject = new FollowJointTrajectoryUntil_ResultRefStruct(other._refObject.toObject());
    this._wrapperFields.error_string.copy(other._wrapperFields.error_string);
  }
  static get classType() {
    return FollowJointTrajectoryUntil_ResultWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"int32","name":"SUCCESSFUL","value":0},{"type":"int32","name":"INVALID_GOAL","value":-1},{"type":"int32","name":"INVALID_JOINTS","value":-2},{"type":"int32","name":"OLD_HEADER_TIMESTAMP","value":-3},{"type":"int32","name":"PATH_TOLERANCE_VIOLATED","value":-4},{"type":"int32","name":"GOAL_TOLERANCE_VIOLATED","value":-5},{"type":"int32","name":"TRIGGERED","value":0},{"type":"int32","name":"NOT_TRIGGERED","value":1}],"fields":[{"name":"error_code","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"until_condition_result","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"error_string","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"ur_msgs","type":"FollowJointTrajectoryUntil_Result","stringUpperBound":null,"isPrimitiveType":false},"msgName":"FollowJointTrajectoryUntil_Result"};
  }
  hasMember(name) {
    let memberNames = ["error_code","until_condition_result","error_string"];
    return memberNames.indexOf(name) !== -1;
  }
}
class FollowJointTrajectoryUntil_ResultArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof FollowJointTrajectoryUntil_ResultWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new FollowJointTrajectoryUntil_ResultWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FollowJointTrajectoryUntil_ResultArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FollowJointTrajectoryUntil_ResultArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new FollowJointTrajectoryUntil_ResultRefArray(size);
    this._refObject = new FollowJointTrajectoryUntil_ResultRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new FollowJointTrajectoryUntil_ResultWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof FollowJointTrajectoryUntil_ResultArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "FollowJointTrajectoryUntil_ResultArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      FollowJointTrajectoryUntil_ResultWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return FollowJointTrajectoryUntil_ResultWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return FollowJointTrajectoryUntil_ResultArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
Object.defineProperty(FollowJointTrajectoryUntil_ResultWrapper, "SUCCESSFUL", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FollowJointTrajectoryUntil_ResultWrapper, "INVALID_GOAL", {value: -1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FollowJointTrajectoryUntil_ResultWrapper, "INVALID_JOINTS", {value: -2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FollowJointTrajectoryUntil_ResultWrapper, "OLD_HEADER_TIMESTAMP", {value: -3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FollowJointTrajectoryUntil_ResultWrapper, "PATH_TOLERANCE_VIOLATED", {value: -4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FollowJointTrajectoryUntil_ResultWrapper, "GOAL_TOLERANCE_VIOLATED", {value: -5, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FollowJointTrajectoryUntil_ResultWrapper, "TRIGGERED", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FollowJointTrajectoryUntil_ResultWrapper, "NOT_TRIGGERED", {value: 1, writable: false, enumerable: true, configurable: true});
module.exports = FollowJointTrajectoryUntil_ResultWrapper;
