// rclnodejs/generated/ur_msgs/ur_msgs__action__ToolContact.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
class ur_msgs__action__ToolContact {
  static get Goal() {
    return require('./ur_msgs__action__ToolContact_Goal.js');
  }
  static get Result() {
    return require('./ur_msgs__action__ToolContact_Result.js');
  }
  static get Feedback() {
    return require('./ur_msgs__action__ToolContact_Feedback.js');
  }
  static get impl() {
    return {
      get SendGoalService() {
        return require('./ur_msgs__action__ToolContact_SendGoal.js');
      },
      get GetResultService() {
        return require('./ur_msgs__action__ToolContact_GetResult.js');
      },
      get FeedbackMessage() {
        return require('./ur_msgs__action__ToolContact_FeedbackMessage.js');
      },
      get CancelGoal() {
        return require('../action_msgs/action_msgs__srv__CancelGoal.js');
      },
      get GoalStatusArray() {
        return require('../action_msgs/action_msgs__msg__GoalStatusArray.js');
      }
    };
  }
  static type() {
    return {pkgName: 'ur_msgs', subFolder: 'action', interfaceName: 'ToolContact'};
  }
}
module.exports = ur_msgs__action__ToolContact;
