// rclnodejs/generated/ur_msgs/ur_msgs__msg__IOStates.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const DigitalWrapper = require('../../generated/ur_msgs/ur_msgs__msg__Digital.js');
const AnalogWrapper = require('../../generated/ur_msgs/ur_msgs__msg__Analog.js');
const IOStatesRefStruct = StructType({
  digital_in_states: DigitalWrapper.refObjectArrayType,
  digital_out_states: DigitalWrapper.refObjectArrayType,
  flag_states: DigitalWrapper.refObjectArrayType,
  analog_in_states: AnalogWrapper.refObjectArrayType,
  analog_out_states: AnalogWrapper.refObjectArrayType,
});
const IOStatesRefArray = ArrayType(IOStatesRefStruct);
const IOStatesRefStructArray = StructType({
  data: IOStatesRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class IOStatesWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new IOStatesRefStruct(other._refObject.toObject());
      this._wrapperFields.digital_in_states = DigitalWrapper.createArray();
      this._wrapperFields.digital_in_states.copy(other._wrapperFields.digital_in_states);
      this._wrapperFields.digital_out_states = DigitalWrapper.createArray();
      this._wrapperFields.digital_out_states.copy(other._wrapperFields.digital_out_states);
      this._wrapperFields.flag_states = DigitalWrapper.createArray();
      this._wrapperFields.flag_states.copy(other._wrapperFields.flag_states);
      this._wrapperFields.analog_in_states = AnalogWrapper.createArray();
      this._wrapperFields.analog_in_states.copy(other._wrapperFields.analog_in_states);
      this._wrapperFields.analog_out_states = AnalogWrapper.createArray();
      this._wrapperFields.analog_out_states.copy(other._wrapperFields.analog_out_states);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new IOStatesRefStruct();
    this._wrapperFields.digital_in_states = DigitalWrapper.createArray();
    this._wrapperFields.digital_out_states = DigitalWrapper.createArray();
    this._wrapperFields.flag_states = DigitalWrapper.createArray();
    this._wrapperFields.analog_in_states = AnalogWrapper.createArray();
    this._wrapperFields.analog_out_states = AnalogWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new IOStatesWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new IOStatesArrayWrapper();
  }
  static get ArrayType() {
    return IOStatesArrayWrapper;
  }
  static get refObjectArrayType() {
    return IOStatesRefStructArray;
  }
  static get refObjectType() {
    return IOStatesRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.digital_in_states.fill([]);
    }
    this._wrapperFields.digital_in_states.freeze(own);
    this._refObject.digital_in_states = this._wrapperFields.digital_in_states.refObject;
    if (own) {
      this._wrapperFields.digital_out_states.fill([]);
    }
    this._wrapperFields.digital_out_states.freeze(own);
    this._refObject.digital_out_states = this._wrapperFields.digital_out_states.refObject;
    if (own) {
      this._wrapperFields.flag_states.fill([]);
    }
    this._wrapperFields.flag_states.freeze(own);
    this._refObject.flag_states = this._wrapperFields.flag_states.refObject;
    if (own) {
      this._wrapperFields.analog_in_states.fill([]);
    }
    this._wrapperFields.analog_in_states.freeze(own);
    this._refObject.analog_in_states = this._wrapperFields.analog_in_states.refObject;
    if (own) {
      this._wrapperFields.analog_out_states.fill([]);
    }
    this._wrapperFields.analog_out_states.freeze(own);
    this._refObject.analog_out_states = this._wrapperFields.analog_out_states.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.digital_in_states.copyRefObject(refObject.digital_in_states);
    this._wrapperFields.digital_out_states.copyRefObject(refObject.digital_out_states);
    this._wrapperFields.flag_states.copyRefObject(refObject.flag_states);
    this._wrapperFields.analog_in_states.copyRefObject(refObject.analog_in_states);
    this._wrapperFields.analog_out_states.copyRefObject(refObject.analog_out_states);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.digital_in_states.size != 0) {
      DigitalWrapper.ArrayType.freeArray(refObject.digital_in_states);
      if (!DigitalWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.digital_in_states, DigitalWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.digital_out_states.size != 0) {
      DigitalWrapper.ArrayType.freeArray(refObject.digital_out_states);
      if (!DigitalWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.digital_out_states, DigitalWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.flag_states.size != 0) {
      DigitalWrapper.ArrayType.freeArray(refObject.flag_states);
      if (!DigitalWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.flag_states, DigitalWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.analog_in_states.size != 0) {
      AnalogWrapper.ArrayType.freeArray(refObject.analog_in_states);
      if (!AnalogWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.analog_in_states, AnalogWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.analog_out_states.size != 0) {
      AnalogWrapper.ArrayType.freeArray(refObject.analog_out_states);
      if (!AnalogWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.analog_out_states, AnalogWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    IOStatesWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'ur_msgs', subFolder: 'msg', interfaceName: 'IOStates'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get digital_in_states() {
    return this._wrapperFields.digital_in_states;
  }
  set digital_in_states(value) {
    this._wrapperFields.digital_in_states.fill(value);
  }
  get digital_out_states() {
    return this._wrapperFields.digital_out_states;
  }
  set digital_out_states(value) {
    this._wrapperFields.digital_out_states.fill(value);
  }
  get flag_states() {
    return this._wrapperFields.flag_states;
  }
  set flag_states(value) {
    this._wrapperFields.flag_states.fill(value);
  }
  get analog_in_states() {
    return this._wrapperFields.analog_in_states;
  }
  set analog_in_states(value) {
    this._wrapperFields.analog_in_states.fill(value);
  }
  get analog_out_states() {
    return this._wrapperFields.analog_out_states;
  }
  set analog_out_states(value) {
    this._wrapperFields.analog_out_states.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new IOStatesRefStruct(refObject.toObject());
    this._wrapperFields.digital_in_states.copyRefObject(this._refObject.digital_in_states);
    this._wrapperFields.digital_out_states.copyRefObject(this._refObject.digital_out_states);
    this._wrapperFields.flag_states.copyRefObject(this._refObject.flag_states);
    this._wrapperFields.analog_in_states.copyRefObject(this._refObject.analog_in_states);
    this._wrapperFields.analog_out_states.copyRefObject(this._refObject.analog_out_states);
  }
  copy(other) {
    this._refObject = new IOStatesRefStruct(other._refObject.toObject());
    this._wrapperFields.digital_in_states.copy(other._wrapperFields.digital_in_states);
    this._wrapperFields.digital_out_states.copy(other._wrapperFields.digital_out_states);
    this._wrapperFields.flag_states.copy(other._wrapperFields.flag_states);
    this._wrapperFields.analog_in_states.copy(other._wrapperFields.analog_in_states);
    this._wrapperFields.analog_out_states.copy(other._wrapperFields.analog_out_states);
  }
  static get classType() {
    return IOStatesWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"digital_in_states","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"ur_msgs","type":"Digital","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"digital_out_states","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"ur_msgs","type":"Digital","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"flag_states","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"ur_msgs","type":"Digital","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"analog_in_states","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"ur_msgs","type":"Analog","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"analog_out_states","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"ur_msgs","type":"Analog","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"ur_msgs","type":"IOStates","stringUpperBound":null,"isPrimitiveType":false},"msgName":"IOStates"};
  }
  hasMember(name) {
    let memberNames = ["digital_in_states","digital_out_states","flag_states","analog_in_states","analog_out_states"];
    return memberNames.indexOf(name) !== -1;
  }
}
class IOStatesArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof IOStatesWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new IOStatesWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to IOStatesArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to IOStatesArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new IOStatesRefArray(size);
    this._refObject = new IOStatesRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new IOStatesWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof IOStatesArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "IOStatesArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      IOStatesWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return IOStatesWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return IOStatesArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = IOStatesWrapper;
