// rclnodejs/generated/ur_msgs/ur_msgs__msg__ToolDataMsg.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const ToolDataMsgRefStruct = StructType({
  analog_input_range2: primitiveTypes.int8,
  analog_input_range3: primitiveTypes.int8,
  analog_input2: primitiveTypes.float64,
  analog_input3: primitiveTypes.float64,
  tool_voltage_48v: primitiveTypes.float32,
  tool_output_voltage: primitiveTypes.uint8,
  tool_current: primitiveTypes.float32,
  tool_temperature: primitiveTypes.float32,
  tool_mode: primitiveTypes.uint8,
});
const ToolDataMsgRefArray = ArrayType(ToolDataMsgRefStruct);
const ToolDataMsgRefStructArray = StructType({
  data: ToolDataMsgRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class ToolDataMsgWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new ToolDataMsgRefStruct(other._refObject.toObject());
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new ToolDataMsgRefStruct();
  }
  static createFromRefObject(refObject) {
    let self = new ToolDataMsgWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ToolDataMsgArrayWrapper();
  }
  static get ArrayType() {
    return ToolDataMsgArrayWrapper;
  }
  static get refObjectArrayType() {
    return ToolDataMsgRefStructArray;
  }
  static get refObjectType() {
    return ToolDataMsgRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.analog_input_range2 = refObject.analog_input_range2;
    this._refObject.analog_input_range3 = refObject.analog_input_range3;
    this._refObject.analog_input2 = refObject.analog_input2;
    this._refObject.analog_input3 = refObject.analog_input3;
    this._refObject.tool_voltage_48v = refObject.tool_voltage_48v;
    this._refObject.tool_output_voltage = refObject.tool_output_voltage;
    this._refObject.tool_current = refObject.tool_current;
    this._refObject.tool_temperature = refObject.tool_temperature;
    this._refObject.tool_mode = refObject.tool_mode;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destroyRawROS(msg) {
    ToolDataMsgWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'ur_msgs', subFolder: 'msg', interfaceName: 'ToolDataMsg'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get analog_input_range2() {
    return this._refObject.analog_input_range2;
  }
  set analog_input_range2(value) {
    this._refObject.analog_input_range2 = value;
  }
  get analog_input_range3() {
    return this._refObject.analog_input_range3;
  }
  set analog_input_range3(value) {
    this._refObject.analog_input_range3 = value;
  }
  get analog_input2() {
    return this._refObject.analog_input2;
  }
  set analog_input2(value) {
    this._refObject.analog_input2 = value;
  }
  get analog_input3() {
    return this._refObject.analog_input3;
  }
  set analog_input3(value) {
    this._refObject.analog_input3 = value;
  }
  get tool_voltage_48v() {
    return this._refObject.tool_voltage_48v;
  }
  set tool_voltage_48v(value) {
    this._refObject.tool_voltage_48v = value;
  }
  get tool_output_voltage() {
    return this._refObject.tool_output_voltage;
  }
  set tool_output_voltage(value) {
    this._refObject.tool_output_voltage = value;
  }
  get tool_current() {
    return this._refObject.tool_current;
  }
  set tool_current(value) {
    this._refObject.tool_current = value;
  }
  get tool_temperature() {
    return this._refObject.tool_temperature;
  }
  set tool_temperature(value) {
    this._refObject.tool_temperature = value;
  }
  get tool_mode() {
    return this._refObject.tool_mode;
  }
  set tool_mode(value) {
    this._refObject.tool_mode = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ToolDataMsgRefStruct(refObject.toObject());
  }
  copy(other) {
    this._refObject = new ToolDataMsgRefStruct(other._refObject.toObject());
  }
  static get classType() {
    return ToolDataMsgWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"int8","name":"ANALOG_INPUT_RANGE_CURRENT","value":0},{"type":"int8","name":"ANALOG_INPUT_RANGE_VOLTAGE","value":1},{"type":"uint8","name":"TOOL_BOOTLOADER_MODE","value":249},{"type":"uint8","name":"TOOL_RUNNING_MODE","value":253},{"type":"uint8","name":"TOOL_IDLE_MODE","value":255}],"fields":[{"name":"analog_input_range2","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"analog_input_range3","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"analog_input2","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"analog_input3","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"tool_voltage_48v","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"tool_output_voltage","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"tool_current","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"tool_temperature","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"tool_mode","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"ur_msgs","type":"ToolDataMsg","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ToolDataMsg"};
  }
  hasMember(name) {
    let memberNames = ["analog_input_range2","analog_input_range3","analog_input2","analog_input3","tool_voltage_48v","tool_output_voltage","tool_current","tool_temperature","tool_mode"];
    return memberNames.indexOf(name) !== -1;
  }
}
class ToolDataMsgArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ToolDataMsgWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ToolDataMsgWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ToolDataMsgArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ToolDataMsgArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new ToolDataMsgRefArray(size);
    this._refObject = new ToolDataMsgRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ToolDataMsgWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ToolDataMsgArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ToolDataMsgArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ToolDataMsgWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ToolDataMsgWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ToolDataMsgArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
Object.defineProperty(ToolDataMsgWrapper, "ANALOG_INPUT_RANGE_CURRENT", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ToolDataMsgWrapper, "ANALOG_INPUT_RANGE_VOLTAGE", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ToolDataMsgWrapper, "TOOL_BOOTLOADER_MODE", {value: 249, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ToolDataMsgWrapper, "TOOL_RUNNING_MODE", {value: 253, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ToolDataMsgWrapper, "TOOL_IDLE_MODE", {value: 255, writable: false, enumerable: true, configurable: true});
module.exports = ToolDataMsgWrapper;
