// rclnodejs/generated/ur_msgs/ur_msgs__action__ToolContact_SendGoal_Request.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const UUIDWrapper = require('../../generated/unique_identifier_msgs/unique_identifier_msgs__msg__UUID.js');
const ToolContact_GoalWrapper = require('../../generated/ur_msgs/ur_msgs__action__ToolContact_Goal.js');
const ToolContact_SendGoal_RequestRefStruct = StructType({
  goal_id: UUIDWrapper.refObjectType,
  goal: ToolContact_GoalWrapper.refObjectType,
});
const ToolContact_SendGoal_RequestRefArray = ArrayType(ToolContact_SendGoal_RequestRefStruct);
const ToolContact_SendGoal_RequestRefStructArray = StructType({
  data: ToolContact_SendGoal_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class ToolContact_SendGoal_RequestWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new ToolContact_SendGoal_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.goal_id =  new UUIDWrapper(other._wrapperFields.goal_id);
      this._wrapperFields.goal =  new ToolContact_GoalWrapper(other._wrapperFields.goal);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new ToolContact_SendGoal_RequestRefStruct();
    this._wrapperFields.goal_id = new UUIDWrapper();
    this._wrapperFields.goal = new ToolContact_GoalWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new ToolContact_SendGoal_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ToolContact_SendGoal_RequestArrayWrapper();
  }
  static get ArrayType() {
    return ToolContact_SendGoal_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return ToolContact_SendGoal_RequestRefStructArray;
  }
  static get refObjectType() {
    return ToolContact_SendGoal_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.goal_id.freeze(own);
    this._refObject.goal_id = this._wrapperFields.goal_id.refObject;
    this._wrapperFields.goal.freeze(own);
    this._refObject.goal = this._wrapperFields.goal.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.goal_id.copyRefObject(refObject.goal_id);
    this._wrapperFields.goal.copyRefObject(refObject.goal);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    UUIDWrapper.freeStruct(refObject.goal_id);
    ToolContact_GoalWrapper.freeStruct(refObject.goal);
  }
  static destroyRawROS(msg) {
    ToolContact_SendGoal_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'ur_msgs', subFolder: 'action', interfaceName: 'ToolContact_SendGoal_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get goal_id() {
    return this._wrapperFields.goal_id;
  }
  set goal_id(value) {
    if (value instanceof UUIDWrapper) {
      this._wrapperFields.goal_id.copy(value);
    } else {
      this._wrapperFields.goal_id.copy(new UUIDWrapper(value));
    }
  }
  get goal() {
    return this._wrapperFields.goal;
  }
  set goal(value) {
    if (value instanceof ToolContact_GoalWrapper) {
      this._wrapperFields.goal.copy(value);
    } else {
      this._wrapperFields.goal.copy(new ToolContact_GoalWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new ToolContact_SendGoal_RequestRefStruct(refObject.toObject());
    this._wrapperFields.goal_id.copyRefObject(this._refObject.goal_id);
    this._wrapperFields.goal.copyRefObject(this._refObject.goal);
  }
  copy(other) {
    this._refObject = new ToolContact_SendGoal_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.goal_id.copy(other._wrapperFields.goal_id);
    this._wrapperFields.goal.copy(other._wrapperFields.goal);
  }
  static get classType() {
    return ToolContact_SendGoal_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"goal_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":null,"pkgName":"unique_identifier_msgs","type":"UUID","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"goal","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":null,"pkgName":"ur_msgs","type":"ToolContact_Goal","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"ur_msgs","type":"ToolContact_SendGoal_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ToolContact_SendGoal_Request"};
  }
  hasMember(name) {
    let memberNames = ["goal_id","goal"];
    return memberNames.indexOf(name) !== -1;
  }
}
class ToolContact_SendGoal_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ToolContact_SendGoal_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ToolContact_SendGoal_RequestWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ToolContact_SendGoal_RequestArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ToolContact_SendGoal_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new ToolContact_SendGoal_RequestRefArray(size);
    this._refObject = new ToolContact_SendGoal_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ToolContact_SendGoal_RequestWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ToolContact_SendGoal_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ToolContact_SendGoal_RequestArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ToolContact_SendGoal_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ToolContact_SendGoal_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ToolContact_SendGoal_RequestArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = ToolContact_SendGoal_RequestWrapper;
