// rclnodejs/generated/ur_msgs/ur_msgs__msg__FrictionModelParameters.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Float64Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float64.js');
const FrictionModelParametersRefStruct = StructType({
  viscous_scale: Float64Wrapper.refObjectArrayType,
  coulomb_scale: Float64Wrapper.refObjectArrayType,
});
const FrictionModelParametersRefArray = ArrayType(FrictionModelParametersRefStruct);
const FrictionModelParametersRefStructArray = StructType({
  data: FrictionModelParametersRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class FrictionModelParametersWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new FrictionModelParametersRefStruct(other._refObject.toObject());
      this._wrapperFields.viscous_scale = Float64Wrapper.createArray();
      this._wrapperFields.viscous_scale.copy(other._wrapperFields.viscous_scale);
      this._wrapperFields.coulomb_scale = Float64Wrapper.createArray();
      this._wrapperFields.coulomb_scale.copy(other._wrapperFields.coulomb_scale);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new FrictionModelParametersRefStruct();
    this._wrapperFields.viscous_scale = Float64Wrapper.createArray();
    this._wrapperFields.coulomb_scale = Float64Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new FrictionModelParametersWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new FrictionModelParametersArrayWrapper();
  }
  static get ArrayType() {
    return FrictionModelParametersArrayWrapper;
  }
  static get refObjectArrayType() {
    return FrictionModelParametersRefStructArray;
  }
  static get refObjectType() {
    return FrictionModelParametersRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.viscous_scale.fill(Float64Array.from([]));
    }
    this._wrapperFields.viscous_scale.freeze(own);
    this._refObject.viscous_scale = this._wrapperFields.viscous_scale.refObject;
    if (own) {
      this._wrapperFields.coulomb_scale.fill(Float64Array.from([]));
    }
    this._wrapperFields.coulomb_scale.freeze(own);
    this._refObject.coulomb_scale = this._wrapperFields.coulomb_scale.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.viscous_scale.copyRefObject(refObject.viscous_scale);
    this._wrapperFields.coulomb_scale.copyRefObject(refObject.coulomb_scale);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.viscous_scale.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.viscous_scale);
      if (!Float64Wrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.viscous_scale, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.coulomb_scale.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.coulomb_scale);
      if (!Float64Wrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.coulomb_scale, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    FrictionModelParametersWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'ur_msgs', subFolder: 'msg', interfaceName: 'FrictionModelParameters'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get viscous_scale() {
    return this._wrapperFields['viscous_scale'].data;
  }
  set viscous_scale(value) {
    this._wrapperFields['viscous_scale'].fill(value);
  }
  get coulomb_scale() {
    return this._wrapperFields['coulomb_scale'].data;
  }
  set coulomb_scale(value) {
    this._wrapperFields['coulomb_scale'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new FrictionModelParametersRefStruct(refObject.toObject());
    this._wrapperFields.viscous_scale.copyRefObject(this._refObject.viscous_scale);
    this._wrapperFields.coulomb_scale.copyRefObject(this._refObject.coulomb_scale);
  }
  copy(other) {
    this._refObject = new FrictionModelParametersRefStruct(other._refObject.toObject());
    this._wrapperFields.viscous_scale.copy(other._wrapperFields.viscous_scale);
    this._wrapperFields.coulomb_scale.copy(other._wrapperFields.coulomb_scale);
  }
  static get classType() {
    return FrictionModelParametersWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"viscous_scale","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"coulomb_scale","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"ur_msgs","type":"FrictionModelParameters","stringUpperBound":null,"isPrimitiveType":false},"msgName":"FrictionModelParameters"};
  }
  hasMember(name) {
    let memberNames = ["viscous_scale","coulomb_scale"];
    return memberNames.indexOf(name) !== -1;
  }
}
class FrictionModelParametersArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof FrictionModelParametersWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new FrictionModelParametersWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FrictionModelParametersArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FrictionModelParametersArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new FrictionModelParametersRefArray(size);
    this._refObject = new FrictionModelParametersRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new FrictionModelParametersWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof FrictionModelParametersArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "FrictionModelParametersArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      FrictionModelParametersWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return FrictionModelParametersWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return FrictionModelParametersArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = FrictionModelParametersWrapper;
