// rclnodejs/generated/ur_msgs/ur_msgs__msg__MasterboardDataMsg.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const MasterboardDataMsgRefStruct = StructType({
  digital_input_bits: primitiveTypes.uint32,
  digital_output_bits: primitiveTypes.uint32,
  analog_input_range0: primitiveTypes.int8,
  analog_input_range1: primitiveTypes.int8,
  analog_input0: primitiveTypes.float64,
  analog_input1: primitiveTypes.float64,
  analog_output_domain0: primitiveTypes.int8,
  analog_output_domain1: primitiveTypes.int8,
  analog_output0: primitiveTypes.float64,
  analog_output1: primitiveTypes.float64,
  masterboard_temperature: primitiveTypes.float32,
  robot_voltage_48v: primitiveTypes.float32,
  robot_current: primitiveTypes.float32,
  master_io_current: primitiveTypes.float32,
  master_safety_state: primitiveTypes.uint8,
  master_onoff_state: primitiveTypes.uint8,
});
const MasterboardDataMsgRefArray = ArrayType(MasterboardDataMsgRefStruct);
const MasterboardDataMsgRefStructArray = StructType({
  data: MasterboardDataMsgRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class MasterboardDataMsgWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new MasterboardDataMsgRefStruct(other._refObject.toObject());
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new MasterboardDataMsgRefStruct();
  }
  static createFromRefObject(refObject) {
    let self = new MasterboardDataMsgWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MasterboardDataMsgArrayWrapper();
  }
  static get ArrayType() {
    return MasterboardDataMsgArrayWrapper;
  }
  static get refObjectArrayType() {
    return MasterboardDataMsgRefStructArray;
  }
  static get refObjectType() {
    return MasterboardDataMsgRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.digital_input_bits = refObject.digital_input_bits;
    this._refObject.digital_output_bits = refObject.digital_output_bits;
    this._refObject.analog_input_range0 = refObject.analog_input_range0;
    this._refObject.analog_input_range1 = refObject.analog_input_range1;
    this._refObject.analog_input0 = refObject.analog_input0;
    this._refObject.analog_input1 = refObject.analog_input1;
    this._refObject.analog_output_domain0 = refObject.analog_output_domain0;
    this._refObject.analog_output_domain1 = refObject.analog_output_domain1;
    this._refObject.analog_output0 = refObject.analog_output0;
    this._refObject.analog_output1 = refObject.analog_output1;
    this._refObject.masterboard_temperature = refObject.masterboard_temperature;
    this._refObject.robot_voltage_48v = refObject.robot_voltage_48v;
    this._refObject.robot_current = refObject.robot_current;
    this._refObject.master_io_current = refObject.master_io_current;
    this._refObject.master_safety_state = refObject.master_safety_state;
    this._refObject.master_onoff_state = refObject.master_onoff_state;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destroyRawROS(msg) {
    MasterboardDataMsgWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'ur_msgs', subFolder: 'msg', interfaceName: 'MasterboardDataMsg'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get digital_input_bits() {
    return this._refObject.digital_input_bits;
  }
  set digital_input_bits(value) {
    this._refObject.digital_input_bits = value;
  }
  get digital_output_bits() {
    return this._refObject.digital_output_bits;
  }
  set digital_output_bits(value) {
    this._refObject.digital_output_bits = value;
  }
  get analog_input_range0() {
    return this._refObject.analog_input_range0;
  }
  set analog_input_range0(value) {
    this._refObject.analog_input_range0 = value;
  }
  get analog_input_range1() {
    return this._refObject.analog_input_range1;
  }
  set analog_input_range1(value) {
    this._refObject.analog_input_range1 = value;
  }
  get analog_input0() {
    return this._refObject.analog_input0;
  }
  set analog_input0(value) {
    this._refObject.analog_input0 = value;
  }
  get analog_input1() {
    return this._refObject.analog_input1;
  }
  set analog_input1(value) {
    this._refObject.analog_input1 = value;
  }
  get analog_output_domain0() {
    return this._refObject.analog_output_domain0;
  }
  set analog_output_domain0(value) {
    this._refObject.analog_output_domain0 = value;
  }
  get analog_output_domain1() {
    return this._refObject.analog_output_domain1;
  }
  set analog_output_domain1(value) {
    this._refObject.analog_output_domain1 = value;
  }
  get analog_output0() {
    return this._refObject.analog_output0;
  }
  set analog_output0(value) {
    this._refObject.analog_output0 = value;
  }
  get analog_output1() {
    return this._refObject.analog_output1;
  }
  set analog_output1(value) {
    this._refObject.analog_output1 = value;
  }
  get masterboard_temperature() {
    return this._refObject.masterboard_temperature;
  }
  set masterboard_temperature(value) {
    this._refObject.masterboard_temperature = value;
  }
  get robot_voltage_48v() {
    return this._refObject.robot_voltage_48v;
  }
  set robot_voltage_48v(value) {
    this._refObject.robot_voltage_48v = value;
  }
  get robot_current() {
    return this._refObject.robot_current;
  }
  set robot_current(value) {
    this._refObject.robot_current = value;
  }
  get master_io_current() {
    return this._refObject.master_io_current;
  }
  set master_io_current(value) {
    this._refObject.master_io_current = value;
  }
  get master_safety_state() {
    return this._refObject.master_safety_state;
  }
  set master_safety_state(value) {
    this._refObject.master_safety_state = value;
  }
  get master_onoff_state() {
    return this._refObject.master_onoff_state;
  }
  set master_onoff_state(value) {
    this._refObject.master_onoff_state = value;
  }
  copyRefObject(refObject) {
    this._refObject = new MasterboardDataMsgRefStruct(refObject.toObject());
  }
  copy(other) {
    this._refObject = new MasterboardDataMsgRefStruct(other._refObject.toObject());
  }
  static get classType() {
    return MasterboardDataMsgWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"digital_input_bits","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"digital_output_bits","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"analog_input_range0","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"analog_input_range1","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"analog_input0","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"analog_input1","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"analog_output_domain0","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"analog_output_domain1","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"analog_output0","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"analog_output1","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"masterboard_temperature","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"robot_voltage_48v","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"robot_current","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"master_io_current","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"master_safety_state","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"master_onoff_state","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"ur_msgs","type":"MasterboardDataMsg","stringUpperBound":null,"isPrimitiveType":false},"msgName":"MasterboardDataMsg"};
  }
  hasMember(name) {
    let memberNames = ["digital_input_bits","digital_output_bits","analog_input_range0","analog_input_range1","analog_input0","analog_input1","analog_output_domain0","analog_output_domain1","analog_output0","analog_output1","masterboard_temperature","robot_voltage_48v","robot_current","master_io_current","master_safety_state","master_onoff_state"];
    return memberNames.indexOf(name) !== -1;
  }
}
class MasterboardDataMsgArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MasterboardDataMsgWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MasterboardDataMsgWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MasterboardDataMsgArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MasterboardDataMsgArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new MasterboardDataMsgRefArray(size);
    this._refObject = new MasterboardDataMsgRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MasterboardDataMsgWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MasterboardDataMsgArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MasterboardDataMsgArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MasterboardDataMsgWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MasterboardDataMsgWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MasterboardDataMsgArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = MasterboardDataMsgWrapper;
