// rclnodejs/generated/ur_msgs/ur_msgs__srv__SetGravity_Request.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Vector3StampedWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Vector3Stamped.js');
const SetGravity_RequestRefStruct = StructType({
  gravity: Vector3StampedWrapper.refObjectType,
});
const SetGravity_RequestRefArray = ArrayType(SetGravity_RequestRefStruct);
const SetGravity_RequestRefStructArray = StructType({
  data: SetGravity_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class SetGravity_RequestWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new SetGravity_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.gravity =  new Vector3StampedWrapper(other._wrapperFields.gravity);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new SetGravity_RequestRefStruct();
    this._wrapperFields.gravity = new Vector3StampedWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new SetGravity_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SetGravity_RequestArrayWrapper();
  }
  static get ArrayType() {
    return SetGravity_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return SetGravity_RequestRefStructArray;
  }
  static get refObjectType() {
    return SetGravity_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.gravity.freeze(own);
    this._refObject.gravity = this._wrapperFields.gravity.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.gravity.copyRefObject(refObject.gravity);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    Vector3StampedWrapper.freeStruct(refObject.gravity);
  }
  static destroyRawROS(msg) {
    SetGravity_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'ur_msgs', subFolder: 'srv', interfaceName: 'SetGravity_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get gravity() {
    return this._wrapperFields.gravity;
  }
  set gravity(value) {
    if (value instanceof Vector3StampedWrapper) {
      this._wrapperFields.gravity.copy(value);
    } else {
      this._wrapperFields.gravity.copy(new Vector3StampedWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new SetGravity_RequestRefStruct(refObject.toObject());
    this._wrapperFields.gravity.copyRefObject(this._refObject.gravity);
  }
  copy(other) {
    this._refObject = new SetGravity_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.gravity.copy(other._wrapperFields.gravity);
  }
  static get classType() {
    return SetGravity_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"gravity","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Vector3Stamped","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"ur_msgs","type":"SetGravity_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SetGravity_Request"};
  }
  hasMember(name) {
    let memberNames = ["gravity"];
    return memberNames.indexOf(name) !== -1;
  }
}
class SetGravity_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SetGravity_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SetGravity_RequestWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetGravity_RequestArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetGravity_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new SetGravity_RequestRefArray(size);
    this._refObject = new SetGravity_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SetGravity_RequestWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SetGravity_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SetGravity_RequestArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SetGravity_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SetGravity_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SetGravity_RequestArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = SetGravity_RequestWrapper;
