// rclnodejs/generated/ur_msgs/ur_msgs__srv__SetForceMode_Request.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PoseStampedWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__PoseStamped.js');
const WrenchWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Wrench.js');
const TwistWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Twist.js');
const Float32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float32.js');
const SetForceMode_RequestRefStruct = StructType({
  task_frame: PoseStampedWrapper.refObjectType,
  selection_vector_x: primitiveTypes.bool,
  selection_vector_y: primitiveTypes.bool,
  selection_vector_z: primitiveTypes.bool,
  selection_vector_rx: primitiveTypes.bool,
  selection_vector_ry: primitiveTypes.bool,
  selection_vector_rz: primitiveTypes.bool,
  wrench: WrenchWrapper.refObjectType,
  type: primitiveTypes.uint8,
  speed_limits: TwistWrapper.refObjectType,
  deviation_limits: ArrayType(primitiveTypes.float32, 6),
  damping_factor: primitiveTypes.float32,
  gain_scaling: primitiveTypes.float32,
});
const SetForceMode_RequestRefArray = ArrayType(SetForceMode_RequestRefStruct);
const SetForceMode_RequestRefStructArray = StructType({
  data: SetForceMode_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class SetForceMode_RequestWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new SetForceMode_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.task_frame =  new PoseStampedWrapper(other._wrapperFields.task_frame);
      this._wrapperFields.wrench =  new WrenchWrapper(other._wrapperFields.wrench);
      this._wrapperFields.speed_limits =  new TwistWrapper(other._wrapperFields.speed_limits);
      this._wrapperFields.deviation_limits = Float32Wrapper.createArray();
      this._wrapperFields.deviation_limits.copy(other._wrapperFields.deviation_limits);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
    this._refObject.type = 2;
    this._wrapperFields.deviation_limits.fill(Float32Array.from([0.01,0.01,0.01,0.01,0.01,0.01]));
    this._refObject.damping_factor = 0.025;
    this._refObject.gain_scaling = 0.5;
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new SetForceMode_RequestRefStruct();
    this._wrapperFields.task_frame = new PoseStampedWrapper();
    this._wrapperFields.wrench = new WrenchWrapper();
    this._wrapperFields.speed_limits = new TwistWrapper();
    this._wrapperFields.deviation_limits = Float32Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new SetForceMode_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SetForceMode_RequestArrayWrapper();
  }
  static get ArrayType() {
    return SetForceMode_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return SetForceMode_RequestRefStructArray;
  }
  static get refObjectType() {
    return SetForceMode_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.task_frame.freeze(own);
    this._refObject.task_frame = this._wrapperFields.task_frame.refObject;
    this._wrapperFields.wrench.freeze(own);
    this._refObject.wrench = this._wrapperFields.wrench.refObject;
    this._wrapperFields.speed_limits.freeze(own);
    this._refObject.speed_limits = this._wrapperFields.speed_limits.refObject;
    this._refObject.deviation_limits = Array.from(this._wrapperFields.deviation_limits.data);
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.task_frame.copyRefObject(refObject.task_frame);
    this._refObject.selection_vector_x = refObject.selection_vector_x;
    this._refObject.selection_vector_y = refObject.selection_vector_y;
    this._refObject.selection_vector_z = refObject.selection_vector_z;
    this._refObject.selection_vector_rx = refObject.selection_vector_rx;
    this._refObject.selection_vector_ry = refObject.selection_vector_ry;
    this._refObject.selection_vector_rz = refObject.selection_vector_rz;
    this._wrapperFields.wrench.copyRefObject(refObject.wrench);
    this._refObject.type = refObject.type;
    this._wrapperFields.speed_limits.copyRefObject(refObject.speed_limits);
    this._wrapperFields.deviation_limits.fill(refObject.deviation_limits.toArray());
    this._refObject.damping_factor = refObject.damping_factor;
    this._refObject.gain_scaling = refObject.gain_scaling;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    PoseStampedWrapper.freeStruct(refObject.task_frame);
    WrenchWrapper.freeStruct(refObject.wrench);
    TwistWrapper.freeStruct(refObject.speed_limits);
  }
  static destroyRawROS(msg) {
    SetForceMode_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'ur_msgs', subFolder: 'srv', interfaceName: 'SetForceMode_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get task_frame() {
    return this._wrapperFields.task_frame;
  }
  set task_frame(value) {
    if (value instanceof PoseStampedWrapper) {
      this._wrapperFields.task_frame.copy(value);
    } else {
      this._wrapperFields.task_frame.copy(new PoseStampedWrapper(value));
    }
  }
  get selection_vector_x() {
    return this._refObject.selection_vector_x;
  }
  set selection_vector_x(value) {
    this._refObject.selection_vector_x = value;
  }
  get selection_vector_y() {
    return this._refObject.selection_vector_y;
  }
  set selection_vector_y(value) {
    this._refObject.selection_vector_y = value;
  }
  get selection_vector_z() {
    return this._refObject.selection_vector_z;
  }
  set selection_vector_z(value) {
    this._refObject.selection_vector_z = value;
  }
  get selection_vector_rx() {
    return this._refObject.selection_vector_rx;
  }
  set selection_vector_rx(value) {
    this._refObject.selection_vector_rx = value;
  }
  get selection_vector_ry() {
    return this._refObject.selection_vector_ry;
  }
  set selection_vector_ry(value) {
    this._refObject.selection_vector_ry = value;
  }
  get selection_vector_rz() {
    return this._refObject.selection_vector_rz;
  }
  set selection_vector_rz(value) {
    this._refObject.selection_vector_rz = value;
  }
  get wrench() {
    return this._wrapperFields.wrench;
  }
  set wrench(value) {
    if (value instanceof WrenchWrapper) {
      this._wrapperFields.wrench.copy(value);
    } else {
      this._wrapperFields.wrench.copy(new WrenchWrapper(value));
    }
  }
  get type() {
    return this._refObject.type;
  }
  set type(value) {
    this._refObject.type = value;
  }
  get speed_limits() {
    return this._wrapperFields.speed_limits;
  }
  set speed_limits(value) {
    if (value instanceof TwistWrapper) {
      this._wrapperFields.speed_limits.copy(value);
    } else {
      this._wrapperFields.speed_limits.copy(new TwistWrapper(value));
    }
  }
  get deviation_limits() {
    return this._wrapperFields['deviation_limits'].data;
  }
  set deviation_limits(value) {
    if (value.length !== 6) {
      throw new RangeError('The length of the array must be 6.');
    }
    this._wrapperFields['deviation_limits'].fill(value);
  }
  get damping_factor() {
    return this._refObject.damping_factor;
  }
  set damping_factor(value) {
    this._refObject.damping_factor = value;
  }
  get gain_scaling() {
    return this._refObject.gain_scaling;
  }
  set gain_scaling(value) {
    this._refObject.gain_scaling = value;
  }
  copyRefObject(refObject) {
    this._refObject = new SetForceMode_RequestRefStruct(refObject.toObject());
    this._wrapperFields.task_frame.copyRefObject(this._refObject.task_frame);
    this._wrapperFields.wrench.copyRefObject(this._refObject.wrench);
    this._wrapperFields.speed_limits.copyRefObject(this._refObject.speed_limits);
    this._wrapperFields.deviation_limits.fill(refObject.deviation_limits.toArray());
  }
  copy(other) {
    this._refObject = new SetForceMode_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.task_frame.copy(other._wrapperFields.task_frame);
    this._wrapperFields.wrench.copy(other._wrapperFields.wrench);
    this._wrapperFields.speed_limits.copy(other._wrapperFields.speed_limits);
    this._wrapperFields.deviation_limits.copy(other._wrapperFields.deviation_limits);
  }
  static get classType() {
    return SetForceMode_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"TCP_TO_ORIGIN","value":1},{"type":"uint8","name":"NO_TRANSFORM","value":2},{"type":"uint8","name":"TCP_VELOCITY_TO_X_Y","value":3}],"fields":[{"name":"task_frame","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"PoseStamped","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"selection_vector_x","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":false},{"name":"selection_vector_y","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":false},{"name":"selection_vector_z","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":false},{"name":"selection_vector_rx","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":false},{"name":"selection_vector_ry","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":false},{"name":"selection_vector_rz","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":false},{"name":"wrench","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Wrench","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":2},{"name":"speed_limits","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Twist","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"deviation_limits","type":{"isArray":true,"arraySize":6,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":[0.01,0.01,0.01,0.01,0.01,0.01]},{"name":"damping_factor","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":0.025},{"name":"gain_scaling","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":0.5}],"baseType":{"pkgName":"ur_msgs","type":"SetForceMode_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SetForceMode_Request"};
  }
  hasMember(name) {
    let memberNames = ["task_frame","selection_vector_x","selection_vector_y","selection_vector_z","selection_vector_rx","selection_vector_ry","selection_vector_rz","wrench","type","speed_limits","deviation_limits","damping_factor","gain_scaling"];
    return memberNames.indexOf(name) !== -1;
  }
}
class SetForceMode_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SetForceMode_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SetForceMode_RequestWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetForceMode_RequestArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetForceMode_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new SetForceMode_RequestRefArray(size);
    this._refObject = new SetForceMode_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SetForceMode_RequestWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SetForceMode_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SetForceMode_RequestArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SetForceMode_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SetForceMode_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SetForceMode_RequestArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
Object.defineProperty(SetForceMode_RequestWrapper, "TCP_TO_ORIGIN", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetForceMode_RequestWrapper, "NO_TRANSFORM", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SetForceMode_RequestWrapper, "TCP_VELOCITY_TO_X_Y", {value: 3, writable: false, enumerable: true, configurable: true});
module.exports = SetForceMode_RequestWrapper;
