// rclnodejs/generated/ur_msgs/ur_msgs__msg__Analog.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const AnalogRefStruct = StructType({
  pin: primitiveTypes.uint8,
  domain: primitiveTypes.uint8,
  state: primitiveTypes.float32,
});
const AnalogRefArray = ArrayType(AnalogRefStruct);
const AnalogRefStructArray = StructType({
  data: AnalogRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class AnalogWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new AnalogRefStruct(other._refObject.toObject());
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new AnalogRefStruct();
  }
  static createFromRefObject(refObject) {
    let self = new AnalogWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new AnalogArrayWrapper();
  }
  static get ArrayType() {
    return AnalogArrayWrapper;
  }
  static get refObjectArrayType() {
    return AnalogRefStructArray;
  }
  static get refObjectType() {
    return AnalogRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.pin = refObject.pin;
    this._refObject.domain = refObject.domain;
    this._refObject.state = refObject.state;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destroyRawROS(msg) {
    AnalogWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'ur_msgs', subFolder: 'msg', interfaceName: 'Analog'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get pin() {
    return this._refObject.pin;
  }
  set pin(value) {
    this._refObject.pin = value;
  }
  get domain() {
    return this._refObject.domain;
  }
  set domain(value) {
    this._refObject.domain = value;
  }
  get state() {
    return this._refObject.state;
  }
  set state(value) {
    this._refObject.state = value;
  }
  copyRefObject(refObject) {
    this._refObject = new AnalogRefStruct(refObject.toObject());
  }
  copy(other) {
    this._refObject = new AnalogRefStruct(other._refObject.toObject());
  }
  static get classType() {
    return AnalogWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"CURRENT","value":0},{"type":"uint8","name":"VOLTAGE","value":1}],"fields":[{"name":"pin","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"domain","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"state","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"ur_msgs","type":"Analog","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Analog"};
  }
  hasMember(name) {
    let memberNames = ["pin","domain","state"];
    return memberNames.indexOf(name) !== -1;
  }
}
class AnalogArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof AnalogWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new AnalogWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to AnalogArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to AnalogArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new AnalogRefArray(size);
    this._refObject = new AnalogRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new AnalogWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof AnalogArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "AnalogArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      AnalogWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return AnalogWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return AnalogArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
Object.defineProperty(AnalogWrapper, "CURRENT", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(AnalogWrapper, "VOLTAGE", {value: 1, writable: false, enumerable: true, configurable: true});
module.exports = AnalogWrapper;
