// rclnodejs/generated/ur_msgs/ur_msgs__srv__GetRobotSoftwareVersion_Response.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const GetRobotSoftwareVersion_ResponseRefStruct = StructType({
  major: primitiveTypes.uint32,
  minor: primitiveTypes.uint32,
  bugfix: primitiveTypes.uint32,
  build: primitiveTypes.uint32,
});
const GetRobotSoftwareVersion_ResponseRefArray = ArrayType(GetRobotSoftwareVersion_ResponseRefStruct);
const GetRobotSoftwareVersion_ResponseRefStructArray = StructType({
  data: GetRobotSoftwareVersion_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class GetRobotSoftwareVersion_ResponseWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetRobotSoftwareVersion_ResponseRefStruct(other._refObject.toObject());
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new GetRobotSoftwareVersion_ResponseRefStruct();
  }
  static createFromRefObject(refObject) {
    let self = new GetRobotSoftwareVersion_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetRobotSoftwareVersion_ResponseArrayWrapper();
  }
  static get ArrayType() {
    return GetRobotSoftwareVersion_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetRobotSoftwareVersion_ResponseRefStructArray;
  }
  static get refObjectType() {
    return GetRobotSoftwareVersion_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.major = refObject.major;
    this._refObject.minor = refObject.minor;
    this._refObject.bugfix = refObject.bugfix;
    this._refObject.build = refObject.build;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destroyRawROS(msg) {
    GetRobotSoftwareVersion_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'ur_msgs', subFolder: 'srv', interfaceName: 'GetRobotSoftwareVersion_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get major() {
    return this._refObject.major;
  }
  set major(value) {
    this._refObject.major = value;
  }
  get minor() {
    return this._refObject.minor;
  }
  set minor(value) {
    this._refObject.minor = value;
  }
  get bugfix() {
    return this._refObject.bugfix;
  }
  set bugfix(value) {
    this._refObject.bugfix = value;
  }
  get build() {
    return this._refObject.build;
  }
  set build(value) {
    this._refObject.build = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GetRobotSoftwareVersion_ResponseRefStruct(refObject.toObject());
  }
  copy(other) {
    this._refObject = new GetRobotSoftwareVersion_ResponseRefStruct(other._refObject.toObject());
  }
  static get classType() {
    return GetRobotSoftwareVersion_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"major","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"minor","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"bugfix","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"build","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"ur_msgs","type":"GetRobotSoftwareVersion_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetRobotSoftwareVersion_Response"};
  }
  hasMember(name) {
    let memberNames = ["major","minor","bugfix","build"];
    return memberNames.indexOf(name) !== -1;
  }
}
class GetRobotSoftwareVersion_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetRobotSoftwareVersion_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetRobotSoftwareVersion_ResponseWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetRobotSoftwareVersion_ResponseArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetRobotSoftwareVersion_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new GetRobotSoftwareVersion_ResponseRefArray(size);
    this._refObject = new GetRobotSoftwareVersion_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetRobotSoftwareVersion_ResponseWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetRobotSoftwareVersion_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetRobotSoftwareVersion_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetRobotSoftwareVersion_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetRobotSoftwareVersion_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetRobotSoftwareVersion_ResponseArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = GetRobotSoftwareVersion_ResponseWrapper;
