// rclnodejs/generated/ur_msgs/ur_msgs__msg__RobotModeDataMsg.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const RobotModeDataMsgRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  is_robot_connected: primitiveTypes.bool,
  is_real_robot_enabled: primitiveTypes.bool,
  is_power_on_robot: primitiveTypes.bool,
  is_emergency_stopped: primitiveTypes.bool,
  is_protective_stopped: primitiveTypes.bool,
  is_program_running: primitiveTypes.bool,
  is_program_paused: primitiveTypes.bool,
});
const RobotModeDataMsgRefArray = ArrayType(RobotModeDataMsgRefStruct);
const RobotModeDataMsgRefStructArray = StructType({
  data: RobotModeDataMsgRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class RobotModeDataMsgWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new RobotModeDataMsgRefStruct(other._refObject.toObject());
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new RobotModeDataMsgRefStruct();
  }
  static createFromRefObject(refObject) {
    let self = new RobotModeDataMsgWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new RobotModeDataMsgArrayWrapper();
  }
  static get ArrayType() {
    return RobotModeDataMsgArrayWrapper;
  }
  static get refObjectArrayType() {
    return RobotModeDataMsgRefStructArray;
  }
  static get refObjectType() {
    return RobotModeDataMsgRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.timestamp = refObject.timestamp;
    this._refObject.is_robot_connected = refObject.is_robot_connected;
    this._refObject.is_real_robot_enabled = refObject.is_real_robot_enabled;
    this._refObject.is_power_on_robot = refObject.is_power_on_robot;
    this._refObject.is_emergency_stopped = refObject.is_emergency_stopped;
    this._refObject.is_protective_stopped = refObject.is_protective_stopped;
    this._refObject.is_program_running = refObject.is_program_running;
    this._refObject.is_program_paused = refObject.is_program_paused;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destroyRawROS(msg) {
    RobotModeDataMsgWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'ur_msgs', subFolder: 'msg', interfaceName: 'RobotModeDataMsg'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    return BigInt(this._refObject.timestamp);
  }
  set timestamp(value) {
    if (typeof value !== "bigint") {
      throw new TypeError('timestamp must be type of bigint');
    }
    this._refObject.timestamp = value.toString();
  }
  get is_robot_connected() {
    return this._refObject.is_robot_connected;
  }
  set is_robot_connected(value) {
    this._refObject.is_robot_connected = value;
  }
  get is_real_robot_enabled() {
    return this._refObject.is_real_robot_enabled;
  }
  set is_real_robot_enabled(value) {
    this._refObject.is_real_robot_enabled = value;
  }
  get is_power_on_robot() {
    return this._refObject.is_power_on_robot;
  }
  set is_power_on_robot(value) {
    this._refObject.is_power_on_robot = value;
  }
  get is_emergency_stopped() {
    return this._refObject.is_emergency_stopped;
  }
  set is_emergency_stopped(value) {
    this._refObject.is_emergency_stopped = value;
  }
  get is_protective_stopped() {
    return this._refObject.is_protective_stopped;
  }
  set is_protective_stopped(value) {
    this._refObject.is_protective_stopped = value;
  }
  get is_program_running() {
    return this._refObject.is_program_running;
  }
  set is_program_running(value) {
    this._refObject.is_program_running = value;
  }
  get is_program_paused() {
    return this._refObject.is_program_paused;
  }
  set is_program_paused(value) {
    this._refObject.is_program_paused = value;
  }
  copyRefObject(refObject) {
    this._refObject = new RobotModeDataMsgRefStruct(refObject.toObject());
  }
  copy(other) {
    this._refObject = new RobotModeDataMsgRefStruct(other._refObject.toObject());
  }
  static get classType() {
    return RobotModeDataMsgWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"is_robot_connected","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"is_real_robot_enabled","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"is_power_on_robot","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"is_emergency_stopped","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"is_protective_stopped","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"is_program_running","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"is_program_paused","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"ur_msgs","type":"RobotModeDataMsg","stringUpperBound":null,"isPrimitiveType":false},"msgName":"RobotModeDataMsg"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","is_robot_connected","is_real_robot_enabled","is_power_on_robot","is_emergency_stopped","is_protective_stopped","is_program_running","is_program_paused"];
    return memberNames.indexOf(name) !== -1;
  }
}
class RobotModeDataMsgArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof RobotModeDataMsgWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new RobotModeDataMsgWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RobotModeDataMsgArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RobotModeDataMsgArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new RobotModeDataMsgRefArray(size);
    this._refObject = new RobotModeDataMsgRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new RobotModeDataMsgWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof RobotModeDataMsgArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "RobotModeDataMsgArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      RobotModeDataMsgWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return RobotModeDataMsgWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return RobotModeDataMsgArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = RobotModeDataMsgWrapper;
