// @chart-sg/node-red-ros2-manager/scripts/validate-installation.js
#!/usr/bin/env node

/**
 * Post-installation validation for Node-RED ROS2 Manager
 * Checks if rclnodejs was properly installed with ROS2 environment
 * Provides helpful guidance if installation is broken
 */

const fs = require('fs');
const path = require('path');

console.log('🔍 Validating ROS2 Manager installation...');

// Function to check if rclnodejs is functional
function validateRclnodejs() {
  try {
    // Try to actually require rclnodejs (not just resolve)
    const rclnodejs = require('rclnodejs');
    return { available: true, error: null };
  } catch (error) {
    return { available: false, error: error.message };
  }
}

// Function to detect Node-RED directory
function findNodeRedDirectory() {
  const possiblePaths = [
    process.env.NODE_RED_HOME,
    path.join(process.env.HOME || process.env.USERPROFILE, '.node-red'),
    process.cwd()
  ];
  
  for (const dirPath of possiblePaths) {
    if (dirPath && fs.existsSync(dirPath)) {
      const packageJsonPath = path.join(dirPath, 'package.json');
      if (fs.existsSync(packageJsonPath)) {
        return dirPath;
      }
    }
  }
  return null;
}

// Check ROS2 environment
function checkROS2Environment() {
  const rosDistro = process.env.ROS_DISTRO;
  const amentPrefix = process.env.AMENT_PREFIX_PATH;
  
  return {
    hasROS2: !!rosDistro,
    hasRMF: amentPrefix && (amentPrefix.includes('rmf') || amentPrefix.includes('RMF')),
    distro: rosDistro
  };
}

// Check if rclnodejs has generated message types
function checkMessageGeneration() {
  const nodeRedDir = findNodeRedDirectory();
  if (!nodeRedDir) return { generated: false, totalCount: 0, rmfCount: 0 };
  
  const generatedPath = path.join(nodeRedDir, 'node_modules', 'rclnodejs', 'generated');
  
  try {
    if (!fs.existsSync(generatedPath)) {
      return { generated: false, totalCount: 0, rmfCount: 0 };
    }
    
    const generatedDirs = fs.readdirSync(generatedPath);
    const rmfDirs = generatedDirs.filter(dir => dir.startsWith('rmf_'));
    
    return {
      generated: generatedDirs.length > 0,
      totalCount: generatedDirs.length,
      rmfCount: rmfDirs.length
    };
  } catch (error) {
    return { generated: false, totalCount: 0, rmfCount: 0 };
  }
}

// Main validation
function main() {
  const ros2Env = checkROS2Environment();
  const rclnodeJS = validateRclnodejs();
  const messageTypes = checkMessageGeneration();
  const nodeRedDir = findNodeRedDirectory();
  
  console.log('\n📋 Installation Status:');
  console.log('======================');
  
  // Check rclnodejs functionality
  if (rclnodeJS.available) {
    console.log('✅ rclnodejs: Functional');
  } else {
    console.log('❌ rclnodejs: Broken');
    console.log(`   Error: ${rclnodeJS.error}`);
  }
  
  // Check environment
  console.log(`🌍 ROS2 Environment: ${ros2Env.hasROS2 ? `✅ (${ros2Env.distro})` : '❌ Not sourced'}`);
  console.log(`🤖 RMF Workspace: ${ros2Env.hasRMF ? '✅ Detected' : '❌ Not sourced'}`);
  
  // Check message types
  console.log(`📨 Message Types: ${messageTypes.generated ? `✅ ${messageTypes.totalCount} total, ${messageTypes.rmfCount} RMF` : '❌ None generated'}`);
  
  // Determine overall status
  const isFullyFunctional = rclnodeJS.available && messageTypes.generated && messageTypes.rmfCount > 0;
  const needsSetup = !rclnodeJS.available || !messageTypes.generated;
  
  console.log('\n🎯 Overall Status:');
  console.log('==================');
  
  if (isFullyFunctional) {
    console.log('✅ ROS2 Manager is fully functional!');
    console.log('🚀 RMF nodes should work correctly in Node-RED');
    return;
  }
  
  if (needsSetup) {
    console.log('⚠️  ROS2 Manager needs setup to function properly');
    console.log('\n🔧 Setup Instructions:');
    console.log('======================');
    
    if (!ros2Env.hasROS2) {
      console.log('\n1️⃣  Source ROS2 Environment:');
      console.log('   # For ROS2 Jazzy (Ubuntu 24.04):');
      console.log('   source /opt/ros/jazzy/setup.bash');
      console.log('   ');
      console.log('   # For ROS2 Humble (Ubuntu 22.04):');
      console.log('   source /opt/ros/humble/setup.bash');
    }
    
    if (!ros2Env.hasRMF) {
      console.log('\n2️⃣  Source RMF Workspace (for RMF users):');
      console.log('   source ~/rmf_ws/install/setup.bash');
      console.log('   # (adjust path to your RMF workspace)');
    }
    
    if (!rclnodeJS.available || !messageTypes.generated) {
      console.log('\n3️⃣  Reinstall rclnodejs with proper environment:');
      console.log(`   cd ${nodeRedDir || '~/.node-red'}`);
      console.log('   npm uninstall rclnodejs');
      console.log('   # Source ROS2/RMF environment first, then:');
      console.log('   npm install rclnodejs');
    }
    
    console.log('\n4️⃣  Restart Node-RED:');
    console.log('   # Make sure to source environment before starting:');
    console.log('   source /opt/ros/jazzy/setup.bash');
    if (ros2Env.hasRMF || messageTypes.rmfCount === 0) {
      console.log('   source ~/rmf_ws/install/setup.bash');
    }
    console.log('   node-red');
    
    console.log('\n💡 This setup only needs to be done once per Node-RED installation.');
    console.log('💡 After setup, RMF nodes will work normally in Node-RED.');
    
    // Don't exit with error - just provide guidance
    console.log('\n⚠️  RMF nodes may not function until setup is complete.');
  }
}

// Only run validation if this script is being executed directly
if (require.main === module) {
  main();
}

module.exports = { validateRclnodejs, checkROS2Environment, checkMessageGeneration };
