// @chart-sg/node-red-ros2-manager/scripts/check-rclnodejs.js
#!/usr/bin/env node

/**
 * Pre-installation check for rclnodejs availability
 * Ensures users have properly installed rclnodejs with ROS2/RMF environment before installing ros2-manager
 */

const fs = require('fs');
const path = require('path');

console.log('🔍 Checking rclnodejs availability...');

// Function to check if rclnodejs is available in node_modules
function checkRclnodejs() {
  try {
    // Try to resolve rclnodejs from the current working directory
    require.resolve('rclnodejs');
    return true;
  } catch (error) {
    return false;
  }
}

// Function to detect Node-RED directory
function findNodeRedDirectory() {
  const possiblePaths = [
    process.env.NODE_RED_HOME,
    path.join(process.env.HOME || process.env.USERPROFILE, '.node-red'),
    process.cwd()
  ];
  
  for (const dirPath of possiblePaths) {
    if (dirPath && fs.existsSync(dirPath)) {
      const packageJsonPath = path.join(dirPath, 'package.json');
      if (fs.existsSync(packageJsonPath)) {
        return dirPath;
      }
    }
  }
  return null;
}

// Check ROS2 environment
function checkROS2Environment() {
  const rosDistro = process.env.ROS_DISTRO;
  const amentPrefix = process.env.AMENT_PREFIX_PATH;
  
  return {
    hasROS2: !!rosDistro,
    hasRMF: amentPrefix && (amentPrefix.includes('rmf') || amentPrefix.includes('RMF')),
    distro: rosDistro
  };
}

// Check if RMF packages are available via ros2 pkg list
function checkRMFPackages() {
  try {
    const { execSync } = require('child_process');
    const packages = execSync('ros2 pkg list', { encoding: 'utf8' });
    const rmfPackages = packages.split('\n').filter(pkg => pkg.startsWith('rmf_'));
    
    return {
      available: rmfPackages.length > 0,
      count: rmfPackages.length,
      hasBuildingMapMsgs: rmfPackages.includes('rmf_building_map_msgs')
    };
  } catch (error) {
    return {
      available: false,
      count: 0,
      hasBuildingMapMsgs: false
    };
  }
}

// Check if rclnodejs has generated RMF message types
function checkRMFMessageGeneration() {
  const nodeRedDir = findNodeRedDirectory();
  if (!nodeRedDir) return { generated: false, rmfCount: 0 };
  
  const generatedPath = path.join(nodeRedDir, 'node_modules', 'rclnodejs', 'generated');
  
  try {
    if (!fs.existsSync(generatedPath)) {
      return { generated: false, rmfCount: 0 };
    }
    
    const generatedDirs = fs.readdirSync(generatedPath);
    const rmfDirs = generatedDirs.filter(dir => dir.startsWith('rmf_'));
    
    return {
      generated: rmfDirs.length > 0,
      rmfCount: rmfDirs.length
    };
  } catch (error) {
    return { generated: false, rmfCount: 0 };
  }
}

// Main check
function main() {
  const ros2Env = checkROS2Environment();
  const hasRclnodejs = checkRclnodejs();
  const nodeRedDir = findNodeRedDirectory();
  
  if (hasRclnodejs) {
    console.log('✅ rclnodejs found - installation will proceed normally');
    if (ros2Env.hasROS2) {
      console.log(`✅ ROS2 environment detected (${ros2Env.distro})`);
    }
    if (ros2Env.hasRMF) {
      console.log('✅ RMF workspace detected');
    }
    return;
  }
  
  // rclnodejs not found - provide helpful instructions
  console.error(`
❌ rclnodejs not found!

Node-RED ROS2 Manager requires rclnodejs to be pre-installed with proper ROS2 environment.

📋 Setup Instructions:

1️⃣  Setup ROS2 Environment:
   For ROS2 Jazzy (Ubuntu 24.04):
     source /opt/ros/jazzy/setup.bash
   
   For ROS2 Humble (Ubuntu 22.04):  
     source /opt/ros/humble/setup.bash

2️⃣  For RMF users, also source RMF workspace:
     source ~/rmf_ws/install/setup.bash

3️⃣  Install rclnodejs in your Node-RED directory:${nodeRedDir ? `
     cd ${nodeRedDir}` : `
     cd ~/.node-red`}
     npm install rclnodejs

4️⃣  Then retry installing this package:
     npm install @chart-sg/node-red-ros2-manager

💡 This setup only needs to be done once per Node-RED installation.

🔧 Current Environment Status:
   ROS2 Sourced: ${ros2Env.hasROS2 ? `✅ (${ros2Env.distro})` : '❌'}
   RMF Workspace: ${ros2Env.hasRMF ? '✅' : '❌'}
   Node-RED Dir: ${nodeRedDir || 'Not detected'}
`);
  
  process.exit(1);
}

// Only run the check if this script is being executed directly
if (require.main === module) {
  main();
}

module.exports = { checkRclnodejs, checkROS2Environment, findNodeRedDirectory };
