// rclnodejs/generated/type_description_interfaces/type_description_interfaces__msg__IndividualTypeDescription.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const FieldWrapper = require('../../generated/type_description_interfaces/type_description_interfaces__msg__Field.js');
const IndividualTypeDescriptionRefStruct = StructType({
  type_name: primitiveTypes.string,
  fields: FieldWrapper.refObjectArrayType,
});
const IndividualTypeDescriptionRefArray = ArrayType(IndividualTypeDescriptionRefStruct);
const IndividualTypeDescriptionRefStructArray = StructType({
  data: IndividualTypeDescriptionRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class IndividualTypeDescriptionWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new IndividualTypeDescriptionRefStruct(other._refObject.toObject());
      this._wrapperFields.type_name =  new StringWrapper(other._wrapperFields.type_name);
      this._wrapperFields.fields = FieldWrapper.createArray();
      this._wrapperFields.fields.copy(other._wrapperFields.fields);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new IndividualTypeDescriptionRefStruct();
    this._wrapperFields.type_name = new StringWrapper();
    this._wrapperFields.fields = FieldWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new IndividualTypeDescriptionWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new IndividualTypeDescriptionArrayWrapper();
  }
  static get ArrayType() {
    return IndividualTypeDescriptionArrayWrapper;
  }
  static get refObjectArrayType() {
    return IndividualTypeDescriptionRefStructArray;
  }
  static get refObjectType() {
    return IndividualTypeDescriptionRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.type_name.freeze(own);
    }
    this._refObject.type_name = this._wrapperFields.type_name.refObject;
    if (own) {
      this._wrapperFields.fields.fill([]);
    }
    this._wrapperFields.fields.freeze(own);
    this._refObject.fields = this._wrapperFields.fields.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.type_name.data = refObject.type_name.data;
    this._wrapperFields.fields.copyRefObject(refObject.fields);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.type_name);
    if (refObject.fields.size != 0) {
      FieldWrapper.ArrayType.freeArray(refObject.fields);
      if (!FieldWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.fields, FieldWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    IndividualTypeDescriptionWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'type_description_interfaces', subFolder: 'msg', interfaceName: 'IndividualTypeDescription'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get type_name() {
    return this._wrapperFields.type_name.data;
  }
  set type_name(value) {
    this._wrapperFields.type_name.data = value;
  }
  get fields() {
    return this._wrapperFields.fields;
  }
  set fields(value) {
    this._wrapperFields.fields.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new IndividualTypeDescriptionRefStruct(refObject.toObject());
    this._wrapperFields.type_name.copyRefObject(this._refObject.type_name);
    this._wrapperFields.fields.copyRefObject(this._refObject.fields);
  }
  copy(other) {
    this._refObject = new IndividualTypeDescriptionRefStruct(other._refObject.toObject());
    this._wrapperFields.type_name.copy(other._wrapperFields.type_name);
    this._wrapperFields.fields.copy(other._wrapperFields.fields);
  }
  static get classType() {
    return IndividualTypeDescriptionWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"type_name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":255,"isPrimitiveType":true},"default_value":null},{"name":"fields","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"type_description_interfaces","type":"Field","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"type_description_interfaces","type":"IndividualTypeDescription","stringUpperBound":null,"isPrimitiveType":false},"msgName":"IndividualTypeDescription"};
  }
  hasMember(name) {
    let memberNames = ["type_name","fields"];
    return memberNames.indexOf(name) !== -1;
  }
}
class IndividualTypeDescriptionArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof IndividualTypeDescriptionWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new IndividualTypeDescriptionWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to IndividualTypeDescriptionArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to IndividualTypeDescriptionArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new IndividualTypeDescriptionRefArray(size);
    this._refObject = new IndividualTypeDescriptionRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new IndividualTypeDescriptionWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof IndividualTypeDescriptionArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "IndividualTypeDescriptionArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      IndividualTypeDescriptionWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return IndividualTypeDescriptionWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return IndividualTypeDescriptionArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = IndividualTypeDescriptionWrapper;
