// rclnodejs/generated/type_description_interfaces/type_description_interfaces__srv__GetTypeDescription_Request.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const GetTypeDescription_RequestRefStruct = StructType({
  type_name: primitiveTypes.string,
  type_hash: primitiveTypes.string,
  include_type_sources: primitiveTypes.bool,
});
const GetTypeDescription_RequestRefArray = ArrayType(GetTypeDescription_RequestRefStruct);
const GetTypeDescription_RequestRefStructArray = StructType({
  data: GetTypeDescription_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class GetTypeDescription_RequestWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetTypeDescription_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.type_name =  new StringWrapper(other._wrapperFields.type_name);
      this._wrapperFields.type_hash =  new StringWrapper(other._wrapperFields.type_hash);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
    this._refObject.include_type_sources = true;
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new GetTypeDescription_RequestRefStruct();
    this._wrapperFields.type_name = new StringWrapper();
    this._wrapperFields.type_hash = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new GetTypeDescription_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetTypeDescription_RequestArrayWrapper();
  }
  static get ArrayType() {
    return GetTypeDescription_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetTypeDescription_RequestRefStructArray;
  }
  static get refObjectType() {
    return GetTypeDescription_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.type_name.freeze(own);
    }
    this._refObject.type_name = this._wrapperFields.type_name.refObject;
    if (own) {
      this._wrapperFields.type_hash.freeze(own);
    }
    this._refObject.type_hash = this._wrapperFields.type_hash.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.type_name.data = refObject.type_name.data;
    this._wrapperFields.type_hash.data = refObject.type_hash.data;
    this._refObject.include_type_sources = refObject.include_type_sources;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.type_name);
    StringWrapper.freeStruct(refObject.type_hash);
  }
  static destroyRawROS(msg) {
    GetTypeDescription_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'type_description_interfaces', subFolder: 'srv', interfaceName: 'GetTypeDescription_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get type_name() {
    return this._wrapperFields.type_name.data;
  }
  set type_name(value) {
    this._wrapperFields.type_name.data = value;
  }
  get type_hash() {
    return this._wrapperFields.type_hash.data;
  }
  set type_hash(value) {
    this._wrapperFields.type_hash.data = value;
  }
  get include_type_sources() {
    return this._refObject.include_type_sources;
  }
  set include_type_sources(value) {
    this._refObject.include_type_sources = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GetTypeDescription_RequestRefStruct(refObject.toObject());
    this._wrapperFields.type_name.copyRefObject(this._refObject.type_name);
    this._wrapperFields.type_hash.copyRefObject(this._refObject.type_hash);
  }
  copy(other) {
    this._refObject = new GetTypeDescription_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.type_name.copy(other._wrapperFields.type_name);
    this._wrapperFields.type_hash.copy(other._wrapperFields.type_hash);
  }
  static get classType() {
    return GetTypeDescription_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"type_name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"type_hash","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"include_type_sources","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":true}],"baseType":{"pkgName":"type_description_interfaces","type":"GetTypeDescription_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetTypeDescription_Request"};
  }
  hasMember(name) {
    let memberNames = ["type_name","type_hash","include_type_sources"];
    return memberNames.indexOf(name) !== -1;
  }
}
class GetTypeDescription_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetTypeDescription_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetTypeDescription_RequestWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetTypeDescription_RequestArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetTypeDescription_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new GetTypeDescription_RequestRefArray(size);
    this._refObject = new GetTypeDescription_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetTypeDescription_RequestWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetTypeDescription_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetTypeDescription_RequestArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetTypeDescription_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetTypeDescription_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetTypeDescription_RequestArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = GetTypeDescription_RequestWrapper;
