// rclnodejs/generated/type_description_interfaces/type_description_interfaces__msg__Field.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const FieldTypeWrapper = require('../../generated/type_description_interfaces/type_description_interfaces__msg__FieldType.js');
const FieldRefStruct = StructType({
  name: primitiveTypes.string,
  type: FieldTypeWrapper.refObjectType,
  default_value: primitiveTypes.string,
});
const FieldRefArray = ArrayType(FieldRefStruct);
const FieldRefStructArray = StructType({
  data: FieldRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class FieldWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new FieldRefStruct(other._refObject.toObject());
      this._wrapperFields.name =  new StringWrapper(other._wrapperFields.name);
      this._wrapperFields.type =  new FieldTypeWrapper(other._wrapperFields.type);
      this._wrapperFields.default_value =  new StringWrapper(other._wrapperFields.default_value);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new FieldRefStruct();
    this._wrapperFields.name = new StringWrapper();
    this._wrapperFields.type = new FieldTypeWrapper();
    this._wrapperFields.default_value = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new FieldWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new FieldArrayWrapper();
  }
  static get ArrayType() {
    return FieldArrayWrapper;
  }
  static get refObjectArrayType() {
    return FieldRefStructArray;
  }
  static get refObjectType() {
    return FieldRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.name.freeze(own);
    }
    this._refObject.name = this._wrapperFields.name.refObject;
    this._wrapperFields.type.freeze(own);
    this._refObject.type = this._wrapperFields.type.refObject;
    if (own) {
      this._wrapperFields.default_value.freeze(own);
    }
    this._refObject.default_value = this._wrapperFields.default_value.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.name.data = refObject.name.data;
    this._wrapperFields.type.copyRefObject(refObject.type);
    this._wrapperFields.default_value.data = refObject.default_value.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.name);
    FieldTypeWrapper.freeStruct(refObject.type);
    StringWrapper.freeStruct(refObject.default_value);
  }
  static destroyRawROS(msg) {
    FieldWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'type_description_interfaces', subFolder: 'msg', interfaceName: 'Field'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get name() {
    return this._wrapperFields.name.data;
  }
  set name(value) {
    this._wrapperFields.name.data = value;
  }
  get type() {
    return this._wrapperFields.type;
  }
  set type(value) {
    if (value instanceof FieldTypeWrapper) {
      this._wrapperFields.type.copy(value);
    } else {
      this._wrapperFields.type.copy(new FieldTypeWrapper(value));
    }
  }
  get default_value() {
    return this._wrapperFields.default_value.data;
  }
  set default_value(value) {
    this._wrapperFields.default_value.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new FieldRefStruct(refObject.toObject());
    this._wrapperFields.name.copyRefObject(this._refObject.name);
    this._wrapperFields.type.copyRefObject(this._refObject.type);
    this._wrapperFields.default_value.copyRefObject(this._refObject.default_value);
  }
  copy(other) {
    this._refObject = new FieldRefStruct(other._refObject.toObject());
    this._wrapperFields.name.copy(other._wrapperFields.name);
    this._wrapperFields.type.copy(other._wrapperFields.type);
    this._wrapperFields.default_value.copy(other._wrapperFields.default_value);
  }
  static get classType() {
    return FieldWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"type_description_interfaces","type":"FieldType","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"default_value","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"type_description_interfaces","type":"Field","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Field"};
  }
  hasMember(name) {
    let memberNames = ["name","type","default_value"];
    return memberNames.indexOf(name) !== -1;
  }
}
class FieldArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof FieldWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new FieldWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FieldArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FieldArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new FieldRefArray(size);
    this._refObject = new FieldRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new FieldWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof FieldArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "FieldArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      FieldWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return FieldWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return FieldArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = FieldWrapper;
