// rclnodejs/generated/type_description_interfaces/type_description_interfaces__msg__TypeSource.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const TypeSourceRefStruct = StructType({
  type_name: primitiveTypes.string,
  encoding: primitiveTypes.string,
  raw_file_contents: primitiveTypes.string,
});
const TypeSourceRefArray = ArrayType(TypeSourceRefStruct);
const TypeSourceRefStructArray = StructType({
  data: TypeSourceRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class TypeSourceWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new TypeSourceRefStruct(other._refObject.toObject());
      this._wrapperFields.type_name =  new StringWrapper(other._wrapperFields.type_name);
      this._wrapperFields.encoding =  new StringWrapper(other._wrapperFields.encoding);
      this._wrapperFields.raw_file_contents =  new StringWrapper(other._wrapperFields.raw_file_contents);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new TypeSourceRefStruct();
    this._wrapperFields.type_name = new StringWrapper();
    this._wrapperFields.encoding = new StringWrapper();
    this._wrapperFields.raw_file_contents = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new TypeSourceWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new TypeSourceArrayWrapper();
  }
  static get ArrayType() {
    return TypeSourceArrayWrapper;
  }
  static get refObjectArrayType() {
    return TypeSourceRefStructArray;
  }
  static get refObjectType() {
    return TypeSourceRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.type_name.freeze(own);
    }
    this._refObject.type_name = this._wrapperFields.type_name.refObject;
    if (own) {
      this._wrapperFields.encoding.freeze(own);
    }
    this._refObject.encoding = this._wrapperFields.encoding.refObject;
    if (own) {
      this._wrapperFields.raw_file_contents.freeze(own);
    }
    this._refObject.raw_file_contents = this._wrapperFields.raw_file_contents.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.type_name.data = refObject.type_name.data;
    this._wrapperFields.encoding.data = refObject.encoding.data;
    this._wrapperFields.raw_file_contents.data = refObject.raw_file_contents.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.type_name);
    StringWrapper.freeStruct(refObject.encoding);
    StringWrapper.freeStruct(refObject.raw_file_contents);
  }
  static destroyRawROS(msg) {
    TypeSourceWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'type_description_interfaces', subFolder: 'msg', interfaceName: 'TypeSource'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get type_name() {
    return this._wrapperFields.type_name.data;
  }
  set type_name(value) {
    this._wrapperFields.type_name.data = value;
  }
  get encoding() {
    return this._wrapperFields.encoding.data;
  }
  set encoding(value) {
    this._wrapperFields.encoding.data = value;
  }
  get raw_file_contents() {
    return this._wrapperFields.raw_file_contents.data;
  }
  set raw_file_contents(value) {
    this._wrapperFields.raw_file_contents.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new TypeSourceRefStruct(refObject.toObject());
    this._wrapperFields.type_name.copyRefObject(this._refObject.type_name);
    this._wrapperFields.encoding.copyRefObject(this._refObject.encoding);
    this._wrapperFields.raw_file_contents.copyRefObject(this._refObject.raw_file_contents);
  }
  copy(other) {
    this._refObject = new TypeSourceRefStruct(other._refObject.toObject());
    this._wrapperFields.type_name.copy(other._wrapperFields.type_name);
    this._wrapperFields.encoding.copy(other._wrapperFields.encoding);
    this._wrapperFields.raw_file_contents.copy(other._wrapperFields.raw_file_contents);
  }
  static get classType() {
    return TypeSourceWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"type_name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"encoding","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"raw_file_contents","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"type_description_interfaces","type":"TypeSource","stringUpperBound":null,"isPrimitiveType":false},"msgName":"TypeSource"};
  }
  hasMember(name) {
    let memberNames = ["type_name","encoding","raw_file_contents"];
    return memberNames.indexOf(name) !== -1;
  }
}
class TypeSourceArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof TypeSourceWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new TypeSourceWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TypeSourceArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TypeSourceArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new TypeSourceRefArray(size);
    this._refObject = new TypeSourceRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new TypeSourceWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof TypeSourceArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "TypeSourceArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      TypeSourceWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return TypeSourceWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return TypeSourceArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = TypeSourceWrapper;
