// rclnodejs/generated/type_description_interfaces/type_description_interfaces__msg__FieldType.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const FieldTypeRefStruct = StructType({
  type_id: primitiveTypes.uint8,
  capacity: primitiveTypes.uint64,
  string_capacity: primitiveTypes.uint64,
  nested_type_name: primitiveTypes.string,
});
const FieldTypeRefArray = ArrayType(FieldTypeRefStruct);
const FieldTypeRefStructArray = StructType({
  data: FieldTypeRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class FieldTypeWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new FieldTypeRefStruct(other._refObject.toObject());
      this._wrapperFields.nested_type_name =  new StringWrapper(other._wrapperFields.nested_type_name);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new FieldTypeRefStruct();
    this._wrapperFields.nested_type_name = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new FieldTypeWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new FieldTypeArrayWrapper();
  }
  static get ArrayType() {
    return FieldTypeArrayWrapper;
  }
  static get refObjectArrayType() {
    return FieldTypeRefStructArray;
  }
  static get refObjectType() {
    return FieldTypeRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.nested_type_name.freeze(own);
    }
    this._refObject.nested_type_name = this._wrapperFields.nested_type_name.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.type_id = refObject.type_id;
    this._refObject.capacity = refObject.capacity;
    this._refObject.string_capacity = refObject.string_capacity;
    this._wrapperFields.nested_type_name.data = refObject.nested_type_name.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.nested_type_name);
  }
  static destroyRawROS(msg) {
    FieldTypeWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'type_description_interfaces', subFolder: 'msg', interfaceName: 'FieldType'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get type_id() {
    return this._refObject.type_id;
  }
  set type_id(value) {
    this._refObject.type_id = value;
  }
  get capacity() {
    return BigInt(this._refObject.capacity);
  }
  set capacity(value) {
    if (typeof value !== "bigint") {
      throw new TypeError('capacity must be type of bigint');
    }
    this._refObject.capacity = value.toString();
  }
  get string_capacity() {
    return BigInt(this._refObject.string_capacity);
  }
  set string_capacity(value) {
    if (typeof value !== "bigint") {
      throw new TypeError('string_capacity must be type of bigint');
    }
    this._refObject.string_capacity = value.toString();
  }
  get nested_type_name() {
    return this._wrapperFields.nested_type_name.data;
  }
  set nested_type_name(value) {
    this._wrapperFields.nested_type_name.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new FieldTypeRefStruct(refObject.toObject());
    this._wrapperFields.nested_type_name.copyRefObject(this._refObject.nested_type_name);
  }
  copy(other) {
    this._refObject = new FieldTypeRefStruct(other._refObject.toObject());
    this._wrapperFields.nested_type_name.copy(other._wrapperFields.nested_type_name);
  }
  static get classType() {
    return FieldTypeWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"FIELD_TYPE_NOT_SET","value":0},{"type":"uint8","name":"FIELD_TYPE_NESTED_TYPE","value":1},{"type":"uint8","name":"FIELD_TYPE_INT8","value":2},{"type":"uint8","name":"FIELD_TYPE_UINT8","value":3},{"type":"uint8","name":"FIELD_TYPE_INT16","value":4},{"type":"uint8","name":"FIELD_TYPE_UINT16","value":5},{"type":"uint8","name":"FIELD_TYPE_INT32","value":6},{"type":"uint8","name":"FIELD_TYPE_UINT32","value":7},{"type":"uint8","name":"FIELD_TYPE_INT64","value":8},{"type":"uint8","name":"FIELD_TYPE_UINT64","value":9},{"type":"uint8","name":"FIELD_TYPE_FLOAT","value":10},{"type":"uint8","name":"FIELD_TYPE_DOUBLE","value":11},{"type":"uint8","name":"FIELD_TYPE_LONG_DOUBLE","value":12},{"type":"uint8","name":"FIELD_TYPE_CHAR","value":13},{"type":"uint8","name":"FIELD_TYPE_WCHAR","value":14},{"type":"uint8","name":"FIELD_TYPE_BOOLEAN","value":15},{"type":"uint8","name":"FIELD_TYPE_BYTE","value":16},{"type":"uint8","name":"FIELD_TYPE_STRING","value":17},{"type":"uint8","name":"FIELD_TYPE_WSTRING","value":18},{"type":"uint8","name":"FIELD_TYPE_FIXED_STRING","value":19},{"type":"uint8","name":"FIELD_TYPE_FIXED_WSTRING","value":20},{"type":"uint8","name":"FIELD_TYPE_BOUNDED_STRING","value":21},{"type":"uint8","name":"FIELD_TYPE_BOUNDED_WSTRING","value":22},{"type":"uint8","name":"FIELD_TYPE_NESTED_TYPE_ARRAY","value":49},{"type":"uint8","name":"FIELD_TYPE_INT8_ARRAY","value":50},{"type":"uint8","name":"FIELD_TYPE_UINT8_ARRAY","value":51},{"type":"uint8","name":"FIELD_TYPE_INT16_ARRAY","value":52},{"type":"uint8","name":"FIELD_TYPE_UINT16_ARRAY","value":53},{"type":"uint8","name":"FIELD_TYPE_INT32_ARRAY","value":54},{"type":"uint8","name":"FIELD_TYPE_UINT32_ARRAY","value":55},{"type":"uint8","name":"FIELD_TYPE_INT64_ARRAY","value":56},{"type":"uint8","name":"FIELD_TYPE_UINT64_ARRAY","value":57},{"type":"uint8","name":"FIELD_TYPE_FLOAT_ARRAY","value":58},{"type":"uint8","name":"FIELD_TYPE_DOUBLE_ARRAY","value":59},{"type":"uint8","name":"FIELD_TYPE_LONG_DOUBLE_ARRAY","value":60},{"type":"uint8","name":"FIELD_TYPE_CHAR_ARRAY","value":61},{"type":"uint8","name":"FIELD_TYPE_WCHAR_ARRAY","value":62},{"type":"uint8","name":"FIELD_TYPE_BOOLEAN_ARRAY","value":63},{"type":"uint8","name":"FIELD_TYPE_BYTE_ARRAY","value":64},{"type":"uint8","name":"FIELD_TYPE_STRING_ARRAY","value":65},{"type":"uint8","name":"FIELD_TYPE_WSTRING_ARRAY","value":66},{"type":"uint8","name":"FIELD_TYPE_FIXED_STRING_ARRAY","value":67},{"type":"uint8","name":"FIELD_TYPE_FIXED_WSTRING_ARRAY","value":68},{"type":"uint8","name":"FIELD_TYPE_BOUNDED_STRING_ARRAY","value":69},{"type":"uint8","name":"FIELD_TYPE_BOUNDED_WSTRING_ARRAY","value":70},{"type":"uint8","name":"FIELD_TYPE_NESTED_TYPE_BOUNDED_SEQUENCE","value":97},{"type":"uint8","name":"FIELD_TYPE_INT8_BOUNDED_SEQUENCE","value":98},{"type":"uint8","name":"FIELD_TYPE_UINT8_BOUNDED_SEQUENCE","value":99},{"type":"uint8","name":"FIELD_TYPE_INT16_BOUNDED_SEQUENCE","value":100},{"type":"uint8","name":"FIELD_TYPE_UINT16_BOUNDED_SEQUENCE","value":101},{"type":"uint8","name":"FIELD_TYPE_INT32_BOUNDED_SEQUENCE","value":102},{"type":"uint8","name":"FIELD_TYPE_UINT32_BOUNDED_SEQUENCE","value":103},{"type":"uint8","name":"FIELD_TYPE_INT64_BOUNDED_SEQUENCE","value":104},{"type":"uint8","name":"FIELD_TYPE_UINT64_BOUNDED_SEQUENCE","value":105},{"type":"uint8","name":"FIELD_TYPE_FLOAT_BOUNDED_SEQUENCE","value":106},{"type":"uint8","name":"FIELD_TYPE_DOUBLE_BOUNDED_SEQUENCE","value":107},{"type":"uint8","name":"FIELD_TYPE_LONG_DOUBLE_BOUNDED_SEQUENCE","value":108},{"type":"uint8","name":"FIELD_TYPE_CHAR_BOUNDED_SEQUENCE","value":109},{"type":"uint8","name":"FIELD_TYPE_WCHAR_BOUNDED_SEQUENCE","value":110},{"type":"uint8","name":"FIELD_TYPE_BOOLEAN_BOUNDED_SEQUENCE","value":111},{"type":"uint8","name":"FIELD_TYPE_BYTE_BOUNDED_SEQUENCE","value":112},{"type":"uint8","name":"FIELD_TYPE_STRING_BOUNDED_SEQUENCE","value":113},{"type":"uint8","name":"FIELD_TYPE_WSTRING_BOUNDED_SEQUENCE","value":114},{"type":"uint8","name":"FIELD_TYPE_FIXED_STRING_BOUNDED_SEQUENCE","value":115},{"type":"uint8","name":"FIELD_TYPE_FIXED_WSTRING_BOUNDED_SEQUENCE","value":116},{"type":"uint8","name":"FIELD_TYPE_BOUNDED_STRING_BOUNDED_SEQUENCE","value":117},{"type":"uint8","name":"FIELD_TYPE_BOUNDED_WSTRING_BOUNDED_SEQUENCE","value":118},{"type":"uint8","name":"FIELD_TYPE_NESTED_TYPE_UNBOUNDED_SEQUENCE","value":145},{"type":"uint8","name":"FIELD_TYPE_INT8_UNBOUNDED_SEQUENCE","value":146},{"type":"uint8","name":"FIELD_TYPE_UINT8_UNBOUNDED_SEQUENCE","value":147},{"type":"uint8","name":"FIELD_TYPE_INT16_UNBOUNDED_SEQUENCE","value":148},{"type":"uint8","name":"FIELD_TYPE_UINT16_UNBOUNDED_SEQUENCE","value":149},{"type":"uint8","name":"FIELD_TYPE_INT32_UNBOUNDED_SEQUENCE","value":150},{"type":"uint8","name":"FIELD_TYPE_UINT32_UNBOUNDED_SEQUENCE","value":151},{"type":"uint8","name":"FIELD_TYPE_INT64_UNBOUNDED_SEQUENCE","value":152},{"type":"uint8","name":"FIELD_TYPE_UINT64_UNBOUNDED_SEQUENCE","value":153},{"type":"uint8","name":"FIELD_TYPE_FLOAT_UNBOUNDED_SEQUENCE","value":154},{"type":"uint8","name":"FIELD_TYPE_DOUBLE_UNBOUNDED_SEQUENCE","value":155},{"type":"uint8","name":"FIELD_TYPE_LONG_DOUBLE_UNBOUNDED_SEQUENCE","value":156},{"type":"uint8","name":"FIELD_TYPE_CHAR_UNBOUNDED_SEQUENCE","value":157},{"type":"uint8","name":"FIELD_TYPE_WCHAR_UNBOUNDED_SEQUENCE","value":158},{"type":"uint8","name":"FIELD_TYPE_BOOLEAN_UNBOUNDED_SEQUENCE","value":159},{"type":"uint8","name":"FIELD_TYPE_BYTE_UNBOUNDED_SEQUENCE","value":160},{"type":"uint8","name":"FIELD_TYPE_STRING_UNBOUNDED_SEQUENCE","value":161},{"type":"uint8","name":"FIELD_TYPE_WSTRING_UNBOUNDED_SEQUENCE","value":162},{"type":"uint8","name":"FIELD_TYPE_FIXED_STRING_UNBOUNDED_SEQUENCE","value":163},{"type":"uint8","name":"FIELD_TYPE_FIXED_WSTRING_UNBOUNDED_SEQUENCE","value":164},{"type":"uint8","name":"FIELD_TYPE_BOUNDED_STRING_UNBOUNDED_SEQUENCE","value":165},{"type":"uint8","name":"FIELD_TYPE_BOUNDED_WSTRING_UNBOUNDED_SEQUENCE","value":166}],"fields":[{"name":"type_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":0},{"name":"capacity","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"string_capacity","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"nested_type_name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":255,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"type_description_interfaces","type":"FieldType","stringUpperBound":null,"isPrimitiveType":false},"msgName":"FieldType"};
  }
  hasMember(name) {
    let memberNames = ["type_id","capacity","string_capacity","nested_type_name"];
    return memberNames.indexOf(name) !== -1;
  }
}
class FieldTypeArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof FieldTypeWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new FieldTypeWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FieldTypeArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FieldTypeArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new FieldTypeRefArray(size);
    this._refObject = new FieldTypeRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new FieldTypeWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof FieldTypeArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "FieldTypeArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      FieldTypeWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return FieldTypeWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return FieldTypeArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_NOT_SET", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_NESTED_TYPE", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_INT8", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_UINT8", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_INT16", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_UINT16", {value: 5, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_INT32", {value: 6, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_UINT32", {value: 7, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_INT64", {value: 8, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_UINT64", {value: 9, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_FLOAT", {value: 10, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_DOUBLE", {value: 11, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_LONG_DOUBLE", {value: 12, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_CHAR", {value: 13, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_WCHAR", {value: 14, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_BOOLEAN", {value: 15, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_BYTE", {value: 16, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_STRING", {value: 17, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_WSTRING", {value: 18, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_FIXED_STRING", {value: 19, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_FIXED_WSTRING", {value: 20, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_BOUNDED_STRING", {value: 21, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_BOUNDED_WSTRING", {value: 22, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_NESTED_TYPE_ARRAY", {value: 49, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_INT8_ARRAY", {value: 50, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_UINT8_ARRAY", {value: 51, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_INT16_ARRAY", {value: 52, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_UINT16_ARRAY", {value: 53, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_INT32_ARRAY", {value: 54, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_UINT32_ARRAY", {value: 55, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_INT64_ARRAY", {value: 56, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_UINT64_ARRAY", {value: 57, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_FLOAT_ARRAY", {value: 58, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_DOUBLE_ARRAY", {value: 59, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_LONG_DOUBLE_ARRAY", {value: 60, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_CHAR_ARRAY", {value: 61, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_WCHAR_ARRAY", {value: 62, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_BOOLEAN_ARRAY", {value: 63, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_BYTE_ARRAY", {value: 64, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_STRING_ARRAY", {value: 65, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_WSTRING_ARRAY", {value: 66, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_FIXED_STRING_ARRAY", {value: 67, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_FIXED_WSTRING_ARRAY", {value: 68, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_BOUNDED_STRING_ARRAY", {value: 69, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_BOUNDED_WSTRING_ARRAY", {value: 70, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_NESTED_TYPE_BOUNDED_SEQUENCE", {value: 97, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_INT8_BOUNDED_SEQUENCE", {value: 98, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_UINT8_BOUNDED_SEQUENCE", {value: 99, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_INT16_BOUNDED_SEQUENCE", {value: 100, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_UINT16_BOUNDED_SEQUENCE", {value: 101, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_INT32_BOUNDED_SEQUENCE", {value: 102, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_UINT32_BOUNDED_SEQUENCE", {value: 103, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_INT64_BOUNDED_SEQUENCE", {value: 104, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_UINT64_BOUNDED_SEQUENCE", {value: 105, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_FLOAT_BOUNDED_SEQUENCE", {value: 106, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_DOUBLE_BOUNDED_SEQUENCE", {value: 107, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_LONG_DOUBLE_BOUNDED_SEQUENCE", {value: 108, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_CHAR_BOUNDED_SEQUENCE", {value: 109, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_WCHAR_BOUNDED_SEQUENCE", {value: 110, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_BOOLEAN_BOUNDED_SEQUENCE", {value: 111, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_BYTE_BOUNDED_SEQUENCE", {value: 112, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_STRING_BOUNDED_SEQUENCE", {value: 113, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_WSTRING_BOUNDED_SEQUENCE", {value: 114, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_FIXED_STRING_BOUNDED_SEQUENCE", {value: 115, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_FIXED_WSTRING_BOUNDED_SEQUENCE", {value: 116, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_BOUNDED_STRING_BOUNDED_SEQUENCE", {value: 117, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_BOUNDED_WSTRING_BOUNDED_SEQUENCE", {value: 118, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_NESTED_TYPE_UNBOUNDED_SEQUENCE", {value: 145, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_INT8_UNBOUNDED_SEQUENCE", {value: 146, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_UINT8_UNBOUNDED_SEQUENCE", {value: 147, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_INT16_UNBOUNDED_SEQUENCE", {value: 148, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_UINT16_UNBOUNDED_SEQUENCE", {value: 149, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_INT32_UNBOUNDED_SEQUENCE", {value: 150, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_UINT32_UNBOUNDED_SEQUENCE", {value: 151, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_INT64_UNBOUNDED_SEQUENCE", {value: 152, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_UINT64_UNBOUNDED_SEQUENCE", {value: 153, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_FLOAT_UNBOUNDED_SEQUENCE", {value: 154, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_DOUBLE_UNBOUNDED_SEQUENCE", {value: 155, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_LONG_DOUBLE_UNBOUNDED_SEQUENCE", {value: 156, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_CHAR_UNBOUNDED_SEQUENCE", {value: 157, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_WCHAR_UNBOUNDED_SEQUENCE", {value: 158, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_BOOLEAN_UNBOUNDED_SEQUENCE", {value: 159, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_BYTE_UNBOUNDED_SEQUENCE", {value: 160, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_STRING_UNBOUNDED_SEQUENCE", {value: 161, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_WSTRING_UNBOUNDED_SEQUENCE", {value: 162, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_FIXED_STRING_UNBOUNDED_SEQUENCE", {value: 163, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_FIXED_WSTRING_UNBOUNDED_SEQUENCE", {value: 164, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_BOUNDED_STRING_UNBOUNDED_SEQUENCE", {value: 165, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FieldTypeWrapper, "FIELD_TYPE_BOUNDED_WSTRING_UNBOUNDED_SEQUENCE", {value: 166, writable: false, enumerable: true, configurable: true});
module.exports = FieldTypeWrapper;
