// rclnodejs/generated/control_msgs/control_msgs__msg__MotionPrimitive.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const MotionArgumentWrapper = require('../../generated/control_msgs/control_msgs__msg__MotionArgument.js');
const PoseStampedWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__PoseStamped.js');
const Float64Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float64.js');
const MotionPrimitiveRefStruct = StructType({
  type: primitiveTypes.int8,
  blend_radius: primitiveTypes.float64,
  additional_arguments: MotionArgumentWrapper.refObjectArrayType,
  poses: PoseStampedWrapper.refObjectArrayType,
  joint_positions: Float64Wrapper.refObjectArrayType,
});
const MotionPrimitiveRefArray = ArrayType(MotionPrimitiveRefStruct);
const MotionPrimitiveRefStructArray = StructType({
  data: MotionPrimitiveRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class MotionPrimitiveWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new MotionPrimitiveRefStruct(other._refObject.toObject());
      this._wrapperFields.additional_arguments = MotionArgumentWrapper.createArray();
      this._wrapperFields.additional_arguments.copy(other._wrapperFields.additional_arguments);
      this._wrapperFields.poses = PoseStampedWrapper.createArray();
      this._wrapperFields.poses.copy(other._wrapperFields.poses);
      this._wrapperFields.joint_positions = Float64Wrapper.createArray();
      this._wrapperFields.joint_positions.copy(other._wrapperFields.joint_positions);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
    this._refObject.type = -1;
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new MotionPrimitiveRefStruct();
    this._wrapperFields.additional_arguments = MotionArgumentWrapper.createArray();
    this._wrapperFields.poses = PoseStampedWrapper.createArray();
    this._wrapperFields.joint_positions = Float64Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new MotionPrimitiveWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MotionPrimitiveArrayWrapper();
  }
  static get ArrayType() {
    return MotionPrimitiveArrayWrapper;
  }
  static get refObjectArrayType() {
    return MotionPrimitiveRefStructArray;
  }
  static get refObjectType() {
    return MotionPrimitiveRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.additional_arguments.fill([]);
    }
    this._wrapperFields.additional_arguments.freeze(own);
    this._refObject.additional_arguments = this._wrapperFields.additional_arguments.refObject;
    if (own) {
      this._wrapperFields.poses.fill([]);
    }
    this._wrapperFields.poses.freeze(own);
    this._refObject.poses = this._wrapperFields.poses.refObject;
    if (own) {
      this._wrapperFields.joint_positions.fill(Float64Array.from([]));
    }
    this._wrapperFields.joint_positions.freeze(own);
    this._refObject.joint_positions = this._wrapperFields.joint_positions.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.type = refObject.type;
    this._refObject.blend_radius = refObject.blend_radius;
    this._wrapperFields.additional_arguments.copyRefObject(refObject.additional_arguments);
    this._wrapperFields.poses.copyRefObject(refObject.poses);
    this._wrapperFields.joint_positions.copyRefObject(refObject.joint_positions);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.additional_arguments.size != 0) {
      MotionArgumentWrapper.ArrayType.freeArray(refObject.additional_arguments);
      if (!MotionArgumentWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.additional_arguments, MotionArgumentWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.poses.size != 0) {
      PoseStampedWrapper.ArrayType.freeArray(refObject.poses);
      if (!PoseStampedWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.poses, PoseStampedWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.joint_positions.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.joint_positions);
      if (!Float64Wrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.joint_positions, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    MotionPrimitiveWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'control_msgs', subFolder: 'msg', interfaceName: 'MotionPrimitive'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get type() {
    return this._refObject.type;
  }
  set type(value) {
    this._refObject.type = value;
  }
  get blend_radius() {
    return this._refObject.blend_radius;
  }
  set blend_radius(value) {
    this._refObject.blend_radius = value;
  }
  get additional_arguments() {
    return this._wrapperFields.additional_arguments;
  }
  set additional_arguments(value) {
    this._wrapperFields.additional_arguments.fill(value);
  }
  get poses() {
    return this._wrapperFields.poses;
  }
  set poses(value) {
    this._wrapperFields.poses.fill(value);
  }
  get joint_positions() {
    return this._wrapperFields['joint_positions'].data;
  }
  set joint_positions(value) {
    this._wrapperFields['joint_positions'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new MotionPrimitiveRefStruct(refObject.toObject());
    this._wrapperFields.additional_arguments.copyRefObject(this._refObject.additional_arguments);
    this._wrapperFields.poses.copyRefObject(this._refObject.poses);
    this._wrapperFields.joint_positions.copyRefObject(this._refObject.joint_positions);
  }
  copy(other) {
    this._refObject = new MotionPrimitiveRefStruct(other._refObject.toObject());
    this._wrapperFields.additional_arguments.copy(other._wrapperFields.additional_arguments);
    this._wrapperFields.poses.copy(other._wrapperFields.poses);
    this._wrapperFields.joint_positions.copy(other._wrapperFields.joint_positions);
  }
  static get classType() {
    return MotionPrimitiveWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"int8","name":"UNKNOWN","value":-1},{"type":"int8","name":"LINEAR_JOINT","value":0},{"type":"int8","name":"LINEAR_CARTESIAN","value":50},{"type":"int8","name":"CIRCULAR_CARTESIAN","value":51}],"fields":[{"name":"type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int8","stringUpperBound":null,"isPrimitiveType":true},"default_value":-1},{"name":"blend_radius","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"additional_arguments","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"control_msgs","type":"MotionArgument","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"poses","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"geometry_msgs","type":"PoseStamped","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"joint_positions","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"control_msgs","type":"MotionPrimitive","stringUpperBound":null,"isPrimitiveType":false},"msgName":"MotionPrimitive"};
  }
  hasMember(name) {
    let memberNames = ["type","blend_radius","additional_arguments","poses","joint_positions"];
    return memberNames.indexOf(name) !== -1;
  }
}
class MotionPrimitiveArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MotionPrimitiveWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MotionPrimitiveWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MotionPrimitiveArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MotionPrimitiveArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new MotionPrimitiveRefArray(size);
    this._refObject = new MotionPrimitiveRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MotionPrimitiveWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MotionPrimitiveArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MotionPrimitiveArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MotionPrimitiveWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MotionPrimitiveWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MotionPrimitiveArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
Object.defineProperty(MotionPrimitiveWrapper, "UNKNOWN", {value: -1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(MotionPrimitiveWrapper, "LINEAR_JOINT", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(MotionPrimitiveWrapper, "LINEAR_CARTESIAN", {value: 50, writable: false, enumerable: true, configurable: true});
Object.defineProperty(MotionPrimitiveWrapper, "CIRCULAR_CARTESIAN", {value: 51, writable: false, enumerable: true, configurable: true});
module.exports = MotionPrimitiveWrapper;
