// rclnodejs/generated/control_msgs/control_msgs__msg__VDA5050SafetyState.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const VDA5050SafetyStateRefStruct = StructType({
  e_stop: primitiveTypes.string,
  field_violation: primitiveTypes.bool,
});
const VDA5050SafetyStateRefArray = ArrayType(VDA5050SafetyStateRefStruct);
const VDA5050SafetyStateRefStructArray = StructType({
  data: VDA5050SafetyStateRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class VDA5050SafetyStateWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new VDA5050SafetyStateRefStruct(other._refObject.toObject());
      this._wrapperFields.e_stop =  new StringWrapper(other._wrapperFields.e_stop);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new VDA5050SafetyStateRefStruct();
    this._wrapperFields.e_stop = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new VDA5050SafetyStateWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new VDA5050SafetyStateArrayWrapper();
  }
  static get ArrayType() {
    return VDA5050SafetyStateArrayWrapper;
  }
  static get refObjectArrayType() {
    return VDA5050SafetyStateRefStructArray;
  }
  static get refObjectType() {
    return VDA5050SafetyStateRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.e_stop.freeze(own);
    }
    this._refObject.e_stop = this._wrapperFields.e_stop.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.e_stop.data = refObject.e_stop.data;
    this._refObject.field_violation = refObject.field_violation;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.e_stop);
  }
  static destroyRawROS(msg) {
    VDA5050SafetyStateWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'control_msgs', subFolder: 'msg', interfaceName: 'VDA5050SafetyState'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get e_stop() {
    return this._wrapperFields.e_stop.data;
  }
  set e_stop(value) {
    this._wrapperFields.e_stop.data = value;
  }
  get field_violation() {
    return this._refObject.field_violation;
  }
  set field_violation(value) {
    this._refObject.field_violation = value;
  }
  copyRefObject(refObject) {
    this._refObject = new VDA5050SafetyStateRefStruct(refObject.toObject());
    this._wrapperFields.e_stop.copyRefObject(this._refObject.e_stop);
  }
  copy(other) {
    this._refObject = new VDA5050SafetyStateRefStruct(other._refObject.toObject());
    this._wrapperFields.e_stop.copy(other._wrapperFields.e_stop);
  }
  static get classType() {
    return VDA5050SafetyStateWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"string","name":"AUTO_ACK","value":"autoAck"},{"type":"string","name":"MANUAL","value":"manual"},{"type":"string","name":"REMOTE","value":"remote"},{"type":"string","name":"NONE","value":"none"}],"fields":[{"name":"e_stop","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"field_violation","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"control_msgs","type":"VDA5050SafetyState","stringUpperBound":null,"isPrimitiveType":false},"msgName":"VDA5050SafetyState"};
  }
  hasMember(name) {
    let memberNames = ["e_stop","field_violation"];
    return memberNames.indexOf(name) !== -1;
  }
}
class VDA5050SafetyStateArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof VDA5050SafetyStateWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new VDA5050SafetyStateWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VDA5050SafetyStateArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VDA5050SafetyStateArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new VDA5050SafetyStateRefArray(size);
    this._refObject = new VDA5050SafetyStateRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new VDA5050SafetyStateWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof VDA5050SafetyStateArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "VDA5050SafetyStateArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      VDA5050SafetyStateWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return VDA5050SafetyStateWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return VDA5050SafetyStateArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
Object.defineProperty(VDA5050SafetyStateWrapper, "AUTO_ACK", {value: "autoAck", writable: false, enumerable: true, configurable: true});
Object.defineProperty(VDA5050SafetyStateWrapper, "MANUAL", {value: "manual", writable: false, enumerable: true, configurable: true});
Object.defineProperty(VDA5050SafetyStateWrapper, "REMOTE", {value: "remote", writable: false, enumerable: true, configurable: true});
Object.defineProperty(VDA5050SafetyStateWrapper, "NONE", {value: "none", writable: false, enumerable: true, configurable: true});
module.exports = VDA5050SafetyStateWrapper;
