// rclnodejs/generated/control_msgs/control_msgs__msg__MotionPrimitiveSequence.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const MotionPrimitiveWrapper = require('../../generated/control_msgs/control_msgs__msg__MotionPrimitive.js');
const MotionPrimitiveSequenceRefStruct = StructType({
  motions: MotionPrimitiveWrapper.refObjectArrayType,
});
const MotionPrimitiveSequenceRefArray = ArrayType(MotionPrimitiveSequenceRefStruct);
const MotionPrimitiveSequenceRefStructArray = StructType({
  data: MotionPrimitiveSequenceRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class MotionPrimitiveSequenceWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new MotionPrimitiveSequenceRefStruct(other._refObject.toObject());
      this._wrapperFields.motions = MotionPrimitiveWrapper.createArray();
      this._wrapperFields.motions.copy(other._wrapperFields.motions);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new MotionPrimitiveSequenceRefStruct();
    this._wrapperFields.motions = MotionPrimitiveWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new MotionPrimitiveSequenceWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MotionPrimitiveSequenceArrayWrapper();
  }
  static get ArrayType() {
    return MotionPrimitiveSequenceArrayWrapper;
  }
  static get refObjectArrayType() {
    return MotionPrimitiveSequenceRefStructArray;
  }
  static get refObjectType() {
    return MotionPrimitiveSequenceRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.motions.fill([]);
    }
    this._wrapperFields.motions.freeze(own);
    this._refObject.motions = this._wrapperFields.motions.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.motions.copyRefObject(refObject.motions);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.motions.size != 0) {
      MotionPrimitiveWrapper.ArrayType.freeArray(refObject.motions);
      if (!MotionPrimitiveWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.motions, MotionPrimitiveWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    MotionPrimitiveSequenceWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'control_msgs', subFolder: 'msg', interfaceName: 'MotionPrimitiveSequence'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get motions() {
    return this._wrapperFields.motions;
  }
  set motions(value) {
    this._wrapperFields.motions.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new MotionPrimitiveSequenceRefStruct(refObject.toObject());
    this._wrapperFields.motions.copyRefObject(this._refObject.motions);
  }
  copy(other) {
    this._refObject = new MotionPrimitiveSequenceRefStruct(other._refObject.toObject());
    this._wrapperFields.motions.copy(other._wrapperFields.motions);
  }
  static get classType() {
    return MotionPrimitiveSequenceWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"motions","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"control_msgs","type":"MotionPrimitive","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"control_msgs","type":"MotionPrimitiveSequence","stringUpperBound":null,"isPrimitiveType":false},"msgName":"MotionPrimitiveSequence"};
  }
  hasMember(name) {
    let memberNames = ["motions"];
    return memberNames.indexOf(name) !== -1;
  }
}
class MotionPrimitiveSequenceArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MotionPrimitiveSequenceWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MotionPrimitiveSequenceWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MotionPrimitiveSequenceArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MotionPrimitiveSequenceArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new MotionPrimitiveSequenceRefArray(size);
    this._refObject = new MotionPrimitiveSequenceRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MotionPrimitiveSequenceWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MotionPrimitiveSequenceArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MotionPrimitiveSequenceArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MotionPrimitiveSequenceWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MotionPrimitiveSequenceWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MotionPrimitiveSequenceArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = MotionPrimitiveSequenceWrapper;
