// rclnodejs/generated/control_msgs/control_msgs__msg__DynamicInterfaceValues.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const InterfaceValueWrapper = require('../../generated/control_msgs/control_msgs__msg__InterfaceValue.js');
const DynamicInterfaceValuesRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  states: InterfaceValueWrapper.refObjectType,
  commands: InterfaceValueWrapper.refObjectType,
});
const DynamicInterfaceValuesRefArray = ArrayType(DynamicInterfaceValuesRefStruct);
const DynamicInterfaceValuesRefStructArray = StructType({
  data: DynamicInterfaceValuesRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class DynamicInterfaceValuesWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new DynamicInterfaceValuesRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.states =  new InterfaceValueWrapper(other._wrapperFields.states);
      this._wrapperFields.commands =  new InterfaceValueWrapper(other._wrapperFields.commands);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new DynamicInterfaceValuesRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.states = new InterfaceValueWrapper();
    this._wrapperFields.commands = new InterfaceValueWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new DynamicInterfaceValuesWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new DynamicInterfaceValuesArrayWrapper();
  }
  static get ArrayType() {
    return DynamicInterfaceValuesArrayWrapper;
  }
  static get refObjectArrayType() {
    return DynamicInterfaceValuesRefStructArray;
  }
  static get refObjectType() {
    return DynamicInterfaceValuesRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.header.freeze(own);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.states.freeze(own);
    this._refObject.states = this._wrapperFields.states.refObject;
    this._wrapperFields.commands.freeze(own);
    this._refObject.commands = this._wrapperFields.commands.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.states.copyRefObject(refObject.states);
    this._wrapperFields.commands.copyRefObject(refObject.commands);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    InterfaceValueWrapper.freeStruct(refObject.states);
    InterfaceValueWrapper.freeStruct(refObject.commands);
  }
  static destroyRawROS(msg) {
    DynamicInterfaceValuesWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'control_msgs', subFolder: 'msg', interfaceName: 'DynamicInterfaceValues'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get states() {
    return this._wrapperFields.states;
  }
  set states(value) {
    if (value instanceof InterfaceValueWrapper) {
      this._wrapperFields.states.copy(value);
    } else {
      this._wrapperFields.states.copy(new InterfaceValueWrapper(value));
    }
  }
  get commands() {
    return this._wrapperFields.commands;
  }
  set commands(value) {
    if (value instanceof InterfaceValueWrapper) {
      this._wrapperFields.commands.copy(value);
    } else {
      this._wrapperFields.commands.copy(new InterfaceValueWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new DynamicInterfaceValuesRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.states.copyRefObject(this._refObject.states);
    this._wrapperFields.commands.copyRefObject(this._refObject.commands);
  }
  copy(other) {
    this._refObject = new DynamicInterfaceValuesRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.states.copy(other._wrapperFields.states);
    this._wrapperFields.commands.copy(other._wrapperFields.commands);
  }
  static get classType() {
    return DynamicInterfaceValuesWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"states","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"control_msgs","type":"InterfaceValue","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"commands","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"control_msgs","type":"InterfaceValue","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"control_msgs","type":"DynamicInterfaceValues","stringUpperBound":null,"isPrimitiveType":false},"msgName":"DynamicInterfaceValues"};
  }
  hasMember(name) {
    let memberNames = ["header","states","commands"];
    return memberNames.indexOf(name) !== -1;
  }
}
class DynamicInterfaceValuesArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof DynamicInterfaceValuesWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new DynamicInterfaceValuesWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to DynamicInterfaceValuesArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to DynamicInterfaceValuesArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new DynamicInterfaceValuesRefArray(size);
    this._refObject = new DynamicInterfaceValuesRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new DynamicInterfaceValuesWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof DynamicInterfaceValuesArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "DynamicInterfaceValuesArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      DynamicInterfaceValuesWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return DynamicInterfaceValuesWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return DynamicInterfaceValuesArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = DynamicInterfaceValuesWrapper;
