// rclnodejs/generated/control_msgs/control_msgs__msg__BatteryStateArray.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const BatteryStateWrapper = require('../../generated/sensor_msgs/sensor_msgs__msg__BatteryState.js');
const BatteryStateArrayRefStruct = StructType({
  battery_states: BatteryStateWrapper.refObjectArrayType,
});
const BatteryStateArrayRefArray = ArrayType(BatteryStateArrayRefStruct);
const BatteryStateArrayRefStructArray = StructType({
  data: BatteryStateArrayRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class BatteryStateArrayWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new BatteryStateArrayRefStruct(other._refObject.toObject());
      this._wrapperFields.battery_states = BatteryStateWrapper.createArray();
      this._wrapperFields.battery_states.copy(other._wrapperFields.battery_states);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new BatteryStateArrayRefStruct();
    this._wrapperFields.battery_states = BatteryStateWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new BatteryStateArrayWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new BatteryStateArrayArrayWrapper();
  }
  static get ArrayType() {
    return BatteryStateArrayArrayWrapper;
  }
  static get refObjectArrayType() {
    return BatteryStateArrayRefStructArray;
  }
  static get refObjectType() {
    return BatteryStateArrayRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.battery_states.fill([]);
    }
    this._wrapperFields.battery_states.freeze(own);
    this._refObject.battery_states = this._wrapperFields.battery_states.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.battery_states.copyRefObject(refObject.battery_states);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.battery_states.size != 0) {
      BatteryStateWrapper.ArrayType.freeArray(refObject.battery_states);
      if (!BatteryStateWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.battery_states, BatteryStateWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    BatteryStateArrayWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'control_msgs', subFolder: 'msg', interfaceName: 'BatteryStateArray'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get battery_states() {
    return this._wrapperFields.battery_states;
  }
  set battery_states(value) {
    this._wrapperFields.battery_states.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new BatteryStateArrayRefStruct(refObject.toObject());
    this._wrapperFields.battery_states.copyRefObject(this._refObject.battery_states);
  }
  copy(other) {
    this._refObject = new BatteryStateArrayRefStruct(other._refObject.toObject());
    this._wrapperFields.battery_states.copy(other._wrapperFields.battery_states);
  }
  static get classType() {
    return BatteryStateArrayWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"battery_states","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"sensor_msgs","type":"BatteryState","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"control_msgs","type":"BatteryStateArray","stringUpperBound":null,"isPrimitiveType":false},"msgName":"BatteryStateArray"};
  }
  hasMember(name) {
    let memberNames = ["battery_states"];
    return memberNames.indexOf(name) !== -1;
  }
}
class BatteryStateArrayArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof BatteryStateArrayWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new BatteryStateArrayWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to BatteryStateArrayArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to BatteryStateArrayArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new BatteryStateArrayRefArray(size);
    this._refObject = new BatteryStateArrayRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new BatteryStateArrayWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof BatteryStateArrayArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "BatteryStateArrayArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      BatteryStateArrayWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return BatteryStateArrayWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return BatteryStateArrayArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = BatteryStateArrayWrapper;
