// rclnodejs/generated/pose_to_moveit_interfaces/pose_to_moveit_interfaces__srv__MoveToPose_Request.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PoseWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Pose.js');
const MoveToPose_RequestRefStruct = StructType({
  target_pose: PoseWrapper.refObjectType,
  duration: primitiveTypes.float64,
});
const MoveToPose_RequestRefArray = ArrayType(MoveToPose_RequestRefStruct);
const MoveToPose_RequestRefStructArray = StructType({
  data: MoveToPose_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class MoveToPose_RequestWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new MoveToPose_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.target_pose =  new PoseWrapper(other._wrapperFields.target_pose);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new MoveToPose_RequestRefStruct();
    this._wrapperFields.target_pose = new PoseWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new MoveToPose_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MoveToPose_RequestArrayWrapper();
  }
  static get ArrayType() {
    return MoveToPose_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return MoveToPose_RequestRefStructArray;
  }
  static get refObjectType() {
    return MoveToPose_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.target_pose.freeze(own);
    this._refObject.target_pose = this._wrapperFields.target_pose.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.target_pose.copyRefObject(refObject.target_pose);
    this._refObject.duration = refObject.duration;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    PoseWrapper.freeStruct(refObject.target_pose);
  }
  static destroyRawROS(msg) {
    MoveToPose_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'pose_to_moveit_interfaces', subFolder: 'srv', interfaceName: 'MoveToPose_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get target_pose() {
    return this._wrapperFields.target_pose;
  }
  set target_pose(value) {
    if (value instanceof PoseWrapper) {
      this._wrapperFields.target_pose.copy(value);
    } else {
      this._wrapperFields.target_pose.copy(new PoseWrapper(value));
    }
  }
  get duration() {
    return this._refObject.duration;
  }
  set duration(value) {
    this._refObject.duration = value;
  }
  copyRefObject(refObject) {
    this._refObject = new MoveToPose_RequestRefStruct(refObject.toObject());
    this._wrapperFields.target_pose.copyRefObject(this._refObject.target_pose);
  }
  copy(other) {
    this._refObject = new MoveToPose_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.target_pose.copy(other._wrapperFields.target_pose);
  }
  static get classType() {
    return MoveToPose_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"target_pose","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Pose","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"duration","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"pose_to_moveit_interfaces","type":"MoveToPose_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"MoveToPose_Request"};
  }
  hasMember(name) {
    let memberNames = ["target_pose","duration"];
    return memberNames.indexOf(name) !== -1;
  }
}
class MoveToPose_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MoveToPose_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MoveToPose_RequestWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MoveToPose_RequestArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MoveToPose_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new MoveToPose_RequestRefArray(size);
    this._refObject = new MoveToPose_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MoveToPose_RequestWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MoveToPose_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MoveToPose_RequestArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MoveToPose_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MoveToPose_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MoveToPose_RequestArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = MoveToPose_RequestWrapper;
