// rclnodejs/generated/type_description_interfaces/type_description_interfaces__srv__GetTypeDescription_Response.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const TypeDescriptionWrapper = require('../../generated/type_description_interfaces/type_description_interfaces__msg__TypeDescription.js');
const TypeSourceWrapper = require('../../generated/type_description_interfaces/type_description_interfaces__msg__TypeSource.js');
const KeyValueWrapper = require('../../generated/type_description_interfaces/type_description_interfaces__msg__KeyValue.js');
const GetTypeDescription_ResponseRefStruct = StructType({
  successful: primitiveTypes.bool,
  failure_reason: primitiveTypes.string,
  type_description: TypeDescriptionWrapper.refObjectType,
  type_sources: TypeSourceWrapper.refObjectArrayType,
  extra_information: KeyValueWrapper.refObjectArrayType,
});
const GetTypeDescription_ResponseRefArray = ArrayType(GetTypeDescription_ResponseRefStruct);
const GetTypeDescription_ResponseRefStructArray = StructType({
  data: GetTypeDescription_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class GetTypeDescription_ResponseWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetTypeDescription_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.failure_reason =  new StringWrapper(other._wrapperFields.failure_reason);
      this._wrapperFields.type_description =  new TypeDescriptionWrapper(other._wrapperFields.type_description);
      this._wrapperFields.type_sources = TypeSourceWrapper.createArray();
      this._wrapperFields.type_sources.copy(other._wrapperFields.type_sources);
      this._wrapperFields.extra_information = KeyValueWrapper.createArray();
      this._wrapperFields.extra_information.copy(other._wrapperFields.extra_information);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new GetTypeDescription_ResponseRefStruct();
    this._wrapperFields.failure_reason = new StringWrapper();
    this._wrapperFields.type_description = new TypeDescriptionWrapper();
    this._wrapperFields.type_sources = TypeSourceWrapper.createArray();
    this._wrapperFields.extra_information = KeyValueWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new GetTypeDescription_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetTypeDescription_ResponseArrayWrapper();
  }
  static get ArrayType() {
    return GetTypeDescription_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetTypeDescription_ResponseRefStructArray;
  }
  static get refObjectType() {
    return GetTypeDescription_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.failure_reason.freeze(own);
    }
    this._refObject.failure_reason = this._wrapperFields.failure_reason.refObject;
    this._wrapperFields.type_description.freeze(own);
    this._refObject.type_description = this._wrapperFields.type_description.refObject;
    if (own) {
      this._wrapperFields.type_sources.fill([]);
    }
    this._wrapperFields.type_sources.freeze(own);
    this._refObject.type_sources = this._wrapperFields.type_sources.refObject;
    if (own) {
      this._wrapperFields.extra_information.fill([]);
    }
    this._wrapperFields.extra_information.freeze(own);
    this._refObject.extra_information = this._wrapperFields.extra_information.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.successful = refObject.successful;
    this._wrapperFields.failure_reason.data = refObject.failure_reason.data;
    this._wrapperFields.type_description.copyRefObject(refObject.type_description);
    this._wrapperFields.type_sources.copyRefObject(refObject.type_sources);
    this._wrapperFields.extra_information.copyRefObject(refObject.extra_information);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.failure_reason);
    TypeDescriptionWrapper.freeStruct(refObject.type_description);
    if (refObject.type_sources.size != 0) {
      TypeSourceWrapper.ArrayType.freeArray(refObject.type_sources);
      if (!TypeSourceWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.type_sources, TypeSourceWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.extra_information.size != 0) {
      KeyValueWrapper.ArrayType.freeArray(refObject.extra_information);
      if (!KeyValueWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.extra_information, KeyValueWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    GetTypeDescription_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'type_description_interfaces', subFolder: 'srv', interfaceName: 'GetTypeDescription_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get successful() {
    return this._refObject.successful;
  }
  set successful(value) {
    this._refObject.successful = value;
  }
  get failure_reason() {
    return this._wrapperFields.failure_reason.data;
  }
  set failure_reason(value) {
    this._wrapperFields.failure_reason.data = value;
  }
  get type_description() {
    return this._wrapperFields.type_description;
  }
  set type_description(value) {
    if (value instanceof TypeDescriptionWrapper) {
      this._wrapperFields.type_description.copy(value);
    } else {
      this._wrapperFields.type_description.copy(new TypeDescriptionWrapper(value));
    }
  }
  get type_sources() {
    return this._wrapperFields.type_sources;
  }
  set type_sources(value) {
    this._wrapperFields.type_sources.fill(value);
  }
  get extra_information() {
    return this._wrapperFields.extra_information;
  }
  set extra_information(value) {
    this._wrapperFields.extra_information.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new GetTypeDescription_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.failure_reason.copyRefObject(this._refObject.failure_reason);
    this._wrapperFields.type_description.copyRefObject(this._refObject.type_description);
    this._wrapperFields.type_sources.copyRefObject(this._refObject.type_sources);
    this._wrapperFields.extra_information.copyRefObject(this._refObject.extra_information);
  }
  copy(other) {
    this._refObject = new GetTypeDescription_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.failure_reason.copy(other._wrapperFields.failure_reason);
    this._wrapperFields.type_description.copy(other._wrapperFields.type_description);
    this._wrapperFields.type_sources.copy(other._wrapperFields.type_sources);
    this._wrapperFields.extra_information.copy(other._wrapperFields.extra_information);
  }
  static get classType() {
    return GetTypeDescription_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"successful","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"failure_reason","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"type_description","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"type_description_interfaces","type":"TypeDescription","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"type_sources","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"type_description_interfaces","type":"TypeSource","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"extra_information","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"type_description_interfaces","type":"KeyValue","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"type_description_interfaces","type":"GetTypeDescription_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetTypeDescription_Response"};
  }
  hasMember(name) {
    let memberNames = ["successful","failure_reason","type_description","type_sources","extra_information"];
    return memberNames.indexOf(name) !== -1;
  }
}
class GetTypeDescription_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetTypeDescription_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetTypeDescription_ResponseWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetTypeDescription_ResponseArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetTypeDescription_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new GetTypeDescription_ResponseRefArray(size);
    this._refObject = new GetTypeDescription_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetTypeDescription_ResponseWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetTypeDescription_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetTypeDescription_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetTypeDescription_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetTypeDescription_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetTypeDescription_ResponseArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = GetTypeDescription_ResponseWrapper;
