// rclnodejs/generated/type_description_interfaces/type_description_interfaces__msg__TypeDescription.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const IndividualTypeDescriptionWrapper = require('../../generated/type_description_interfaces/type_description_interfaces__msg__IndividualTypeDescription.js');
const TypeDescriptionRefStruct = StructType({
  type_description: IndividualTypeDescriptionWrapper.refObjectType,
  referenced_type_descriptions: IndividualTypeDescriptionWrapper.refObjectArrayType,
});
const TypeDescriptionRefArray = ArrayType(TypeDescriptionRefStruct);
const TypeDescriptionRefStructArray = StructType({
  data: TypeDescriptionRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class TypeDescriptionWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new TypeDescriptionRefStruct(other._refObject.toObject());
      this._wrapperFields.type_description =  new IndividualTypeDescriptionWrapper(other._wrapperFields.type_description);
      this._wrapperFields.referenced_type_descriptions = IndividualTypeDescriptionWrapper.createArray();
      this._wrapperFields.referenced_type_descriptions.copy(other._wrapperFields.referenced_type_descriptions);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new TypeDescriptionRefStruct();
    this._wrapperFields.type_description = new IndividualTypeDescriptionWrapper();
    this._wrapperFields.referenced_type_descriptions = IndividualTypeDescriptionWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new TypeDescriptionWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new TypeDescriptionArrayWrapper();
  }
  static get ArrayType() {
    return TypeDescriptionArrayWrapper;
  }
  static get refObjectArrayType() {
    return TypeDescriptionRefStructArray;
  }
  static get refObjectType() {
    return TypeDescriptionRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.type_description.freeze(own);
    this._refObject.type_description = this._wrapperFields.type_description.refObject;
    if (own) {
      this._wrapperFields.referenced_type_descriptions.fill([]);
    }
    this._wrapperFields.referenced_type_descriptions.freeze(own);
    this._refObject.referenced_type_descriptions = this._wrapperFields.referenced_type_descriptions.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.type_description.copyRefObject(refObject.type_description);
    this._wrapperFields.referenced_type_descriptions.copyRefObject(refObject.referenced_type_descriptions);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    IndividualTypeDescriptionWrapper.freeStruct(refObject.type_description);
    if (refObject.referenced_type_descriptions.size != 0) {
      IndividualTypeDescriptionWrapper.ArrayType.freeArray(refObject.referenced_type_descriptions);
      if (!IndividualTypeDescriptionWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.referenced_type_descriptions, IndividualTypeDescriptionWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    TypeDescriptionWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'type_description_interfaces', subFolder: 'msg', interfaceName: 'TypeDescription'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get type_description() {
    return this._wrapperFields.type_description;
  }
  set type_description(value) {
    if (value instanceof IndividualTypeDescriptionWrapper) {
      this._wrapperFields.type_description.copy(value);
    } else {
      this._wrapperFields.type_description.copy(new IndividualTypeDescriptionWrapper(value));
    }
  }
  get referenced_type_descriptions() {
    return this._wrapperFields.referenced_type_descriptions;
  }
  set referenced_type_descriptions(value) {
    this._wrapperFields.referenced_type_descriptions.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new TypeDescriptionRefStruct(refObject.toObject());
    this._wrapperFields.type_description.copyRefObject(this._refObject.type_description);
    this._wrapperFields.referenced_type_descriptions.copyRefObject(this._refObject.referenced_type_descriptions);
  }
  copy(other) {
    this._refObject = new TypeDescriptionRefStruct(other._refObject.toObject());
    this._wrapperFields.type_description.copy(other._wrapperFields.type_description);
    this._wrapperFields.referenced_type_descriptions.copy(other._wrapperFields.referenced_type_descriptions);
  }
  static get classType() {
    return TypeDescriptionWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"type_description","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"type_description_interfaces","type":"IndividualTypeDescription","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"referenced_type_descriptions","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"type_description_interfaces","type":"IndividualTypeDescription","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"type_description_interfaces","type":"TypeDescription","stringUpperBound":null,"isPrimitiveType":false},"msgName":"TypeDescription"};
  }
  hasMember(name) {
    let memberNames = ["type_description","referenced_type_descriptions"];
    return memberNames.indexOf(name) !== -1;
  }
}
class TypeDescriptionArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof TypeDescriptionWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new TypeDescriptionWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TypeDescriptionArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TypeDescriptionArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new TypeDescriptionRefArray(size);
    this._refObject = new TypeDescriptionRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new TypeDescriptionWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof TypeDescriptionArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "TypeDescriptionArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      TypeDescriptionWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return TypeDescriptionWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return TypeDescriptionArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = TypeDescriptionWrapper;
