// rclnodejs/generated/pal_statistics_msgs/pal_statistics_msgs__msg__StatisticsNames.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const StatisticsNamesRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  names: StringWrapper.refObjectArrayType,
  names_version: primitiveTypes.uint32,
});
const StatisticsNamesRefArray = ArrayType(StatisticsNamesRefStruct);
const StatisticsNamesRefStructArray = StructType({
  data: StatisticsNamesRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class StatisticsNamesWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new StatisticsNamesRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.names = StringWrapper.createArray();
      this._wrapperFields.names.copy(other._wrapperFields.names);
      this.names = other.names;
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new StatisticsNamesRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.names = StringWrapper.createArray();
    this._namesArray = [];
  }
  static createFromRefObject(refObject) {
    let self = new StatisticsNamesWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new StatisticsNamesArrayWrapper();
  }
  static get ArrayType() {
    return StatisticsNamesArrayWrapper;
  }
  static get refObjectArrayType() {
    return StatisticsNamesRefStructArray;
  }
  static get refObjectType() {
    return StatisticsNamesRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.header.freeze(own);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (own) {
      this._wrapperFields.names.fill([]);
    } else {
      this._wrapperFields.names.fill(this._namesArray);
    }
    this._wrapperFields.names.freeze(own);
    this._refObject.names = this._wrapperFields.names.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    refObject.names.data.length = refObject.names.size;
    for (let index = 0; index < refObject.names.size; index++) {
      this._namesArray[index] = refObject.names.data[index].data;
    }
    this._refObject.names_version = refObject.names_version;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    if (refObject.names.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.names);
      if (!StringWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.names, StringWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    StatisticsNamesWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'pal_statistics_msgs', subFolder: 'msg', interfaceName: 'StatisticsNames'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get names() {
    return this._namesArray;
  }
  set names(value) {
    this._namesArray = value;
  }
  get names_version() {
    return this._refObject.names_version;
  }
  set names_version(value) {
    this._refObject.names_version = value;
  }
  copyRefObject(refObject) {
    this._refObject = new StatisticsNamesRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    refObject.names.data.length = refObject.names.size;
    for (let index = 0; index < refObject.names.size; index++) {
      this._namesArray[index] = refObject.names.data[index].data;
    }
  }
  copy(other) {
    this._refObject = new StatisticsNamesRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._namesArray = other._namesArray.slice();
  }
  static get classType() {
    return StatisticsNamesWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"names","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"names_version","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"pal_statistics_msgs","type":"StatisticsNames","stringUpperBound":null,"isPrimitiveType":false},"msgName":"StatisticsNames"};
  }
  hasMember(name) {
    let memberNames = ["header","names","names_version"];
    return memberNames.indexOf(name) !== -1;
  }
}
class StatisticsNamesArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof StatisticsNamesWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new StatisticsNamesWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to StatisticsNamesArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to StatisticsNamesArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new StatisticsNamesRefArray(size);
    this._refObject = new StatisticsNamesRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new StatisticsNamesWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof StatisticsNamesArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "StatisticsNamesArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      StatisticsNamesWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return StatisticsNamesWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return StatisticsNamesArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = StatisticsNamesWrapper;
