// rclnodejs/generated/pal_statistics_msgs/pal_statistics_msgs__msg__Statistics.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const StatisticWrapper = require('../../generated/pal_statistics_msgs/pal_statistics_msgs__msg__Statistic.js');
const StatisticsRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  statistics: StatisticWrapper.refObjectArrayType,
});
const StatisticsRefArray = ArrayType(StatisticsRefStruct);
const StatisticsRefStructArray = StructType({
  data: StatisticsRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class StatisticsWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new StatisticsRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.statistics = StatisticWrapper.createArray();
      this._wrapperFields.statistics.copy(other._wrapperFields.statistics);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new StatisticsRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.statistics = StatisticWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new StatisticsWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new StatisticsArrayWrapper();
  }
  static get ArrayType() {
    return StatisticsArrayWrapper;
  }
  static get refObjectArrayType() {
    return StatisticsRefStructArray;
  }
  static get refObjectType() {
    return StatisticsRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.header.freeze(own);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (own) {
      this._wrapperFields.statistics.fill([]);
    }
    this._wrapperFields.statistics.freeze(own);
    this._refObject.statistics = this._wrapperFields.statistics.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.statistics.copyRefObject(refObject.statistics);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    if (refObject.statistics.size != 0) {
      StatisticWrapper.ArrayType.freeArray(refObject.statistics);
      if (!StatisticWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.statistics, StatisticWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    StatisticsWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'pal_statistics_msgs', subFolder: 'msg', interfaceName: 'Statistics'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get statistics() {
    return this._wrapperFields.statistics;
  }
  set statistics(value) {
    this._wrapperFields.statistics.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new StatisticsRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.statistics.copyRefObject(this._refObject.statistics);
  }
  copy(other) {
    this._refObject = new StatisticsRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.statistics.copy(other._wrapperFields.statistics);
  }
  static get classType() {
    return StatisticsWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"statistics","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"pal_statistics_msgs","type":"Statistic","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"pal_statistics_msgs","type":"Statistics","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Statistics"};
  }
  hasMember(name) {
    let memberNames = ["header","statistics"];
    return memberNames.indexOf(name) !== -1;
  }
}
class StatisticsArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof StatisticsWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new StatisticsWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to StatisticsArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to StatisticsArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new StatisticsRefArray(size);
    this._refObject = new StatisticsRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new StatisticsWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof StatisticsArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "StatisticsArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      StatisticsWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return StatisticsWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return StatisticsArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = StatisticsWrapper;
