// rclnodejs/generated/rcl_interfaces/rcl_interfaces__srv__GetLoggerLevels_Response.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const LoggerLevelWrapper = require('../../generated/rcl_interfaces/rcl_interfaces__msg__LoggerLevel.js');
const GetLoggerLevels_ResponseRefStruct = StructType({
  levels: LoggerLevelWrapper.refObjectArrayType,
});
const GetLoggerLevels_ResponseRefArray = ArrayType(GetLoggerLevels_ResponseRefStruct);
const GetLoggerLevels_ResponseRefStructArray = StructType({
  data: GetLoggerLevels_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class GetLoggerLevels_ResponseWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetLoggerLevels_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.levels = LoggerLevelWrapper.createArray();
      this._wrapperFields.levels.copy(other._wrapperFields.levels);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new GetLoggerLevels_ResponseRefStruct();
    this._wrapperFields.levels = LoggerLevelWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new GetLoggerLevels_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetLoggerLevels_ResponseArrayWrapper();
  }
  static get ArrayType() {
    return GetLoggerLevels_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetLoggerLevels_ResponseRefStructArray;
  }
  static get refObjectType() {
    return GetLoggerLevels_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.levels.fill([]);
    }
    this._wrapperFields.levels.freeze(own);
    this._refObject.levels = this._wrapperFields.levels.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.levels.copyRefObject(refObject.levels);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.levels.size != 0) {
      LoggerLevelWrapper.ArrayType.freeArray(refObject.levels);
      if (!LoggerLevelWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.levels, LoggerLevelWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    GetLoggerLevels_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rcl_interfaces', subFolder: 'srv', interfaceName: 'GetLoggerLevels_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get levels() {
    return this._wrapperFields.levels;
  }
  set levels(value) {
    this._wrapperFields.levels.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new GetLoggerLevels_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.levels.copyRefObject(this._refObject.levels);
  }
  copy(other) {
    this._refObject = new GetLoggerLevels_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.levels.copy(other._wrapperFields.levels);
  }
  static get classType() {
    return GetLoggerLevels_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"levels","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"rcl_interfaces","type":"LoggerLevel","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"rcl_interfaces","type":"GetLoggerLevels_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetLoggerLevels_Response"};
  }
  hasMember(name) {
    let memberNames = ["levels"];
    return memberNames.indexOf(name) !== -1;
  }
}
class GetLoggerLevels_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetLoggerLevels_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetLoggerLevels_ResponseWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetLoggerLevels_ResponseArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetLoggerLevels_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new GetLoggerLevels_ResponseRefArray(size);
    this._refObject = new GetLoggerLevels_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetLoggerLevels_ResponseWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetLoggerLevels_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetLoggerLevels_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetLoggerLevels_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetLoggerLevels_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetLoggerLevels_ResponseArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = GetLoggerLevels_ResponseWrapper;
