// rclnodejs/generated/rcl_interfaces/rcl_interfaces__msg__SetLoggerLevelsResult.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const SetLoggerLevelsResultRefStruct = StructType({
  successful: primitiveTypes.bool,
  reason: primitiveTypes.string,
});
const SetLoggerLevelsResultRefArray = ArrayType(SetLoggerLevelsResultRefStruct);
const SetLoggerLevelsResultRefStructArray = StructType({
  data: SetLoggerLevelsResultRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class SetLoggerLevelsResultWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new SetLoggerLevelsResultRefStruct(other._refObject.toObject());
      this._wrapperFields.reason =  new StringWrapper(other._wrapperFields.reason);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new SetLoggerLevelsResultRefStruct();
    this._wrapperFields.reason = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new SetLoggerLevelsResultWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SetLoggerLevelsResultArrayWrapper();
  }
  static get ArrayType() {
    return SetLoggerLevelsResultArrayWrapper;
  }
  static get refObjectArrayType() {
    return SetLoggerLevelsResultRefStructArray;
  }
  static get refObjectType() {
    return SetLoggerLevelsResultRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.reason.freeze(own);
    }
    this._refObject.reason = this._wrapperFields.reason.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.successful = refObject.successful;
    this._wrapperFields.reason.data = refObject.reason.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.reason);
  }
  static destroyRawROS(msg) {
    SetLoggerLevelsResultWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rcl_interfaces', subFolder: 'msg', interfaceName: 'SetLoggerLevelsResult'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get successful() {
    return this._refObject.successful;
  }
  set successful(value) {
    this._refObject.successful = value;
  }
  get reason() {
    return this._wrapperFields.reason.data;
  }
  set reason(value) {
    this._wrapperFields.reason.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new SetLoggerLevelsResultRefStruct(refObject.toObject());
    this._wrapperFields.reason.copyRefObject(this._refObject.reason);
  }
  copy(other) {
    this._refObject = new SetLoggerLevelsResultRefStruct(other._refObject.toObject());
    this._wrapperFields.reason.copy(other._wrapperFields.reason);
  }
  static get classType() {
    return SetLoggerLevelsResultWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"successful","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reason","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rcl_interfaces","type":"SetLoggerLevelsResult","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SetLoggerLevelsResult"};
  }
  hasMember(name) {
    let memberNames = ["successful","reason"];
    return memberNames.indexOf(name) !== -1;
  }
}
class SetLoggerLevelsResultArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SetLoggerLevelsResultWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SetLoggerLevelsResultWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetLoggerLevelsResultArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetLoggerLevelsResultArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new SetLoggerLevelsResultRefArray(size);
    this._refObject = new SetLoggerLevelsResultRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SetLoggerLevelsResultWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SetLoggerLevelsResultArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SetLoggerLevelsResultArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SetLoggerLevelsResultWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SetLoggerLevelsResultWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SetLoggerLevelsResultArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = SetLoggerLevelsResultWrapper;
