// rclnodejs/generated/rosbag2_interfaces/rosbag2_interfaces__srv__Record_Response.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const Record_ResponseRefStruct = StructType({
  return_code: primitiveTypes.int32,
  error_string: primitiveTypes.string,
});
const Record_ResponseRefArray = ArrayType(Record_ResponseRefStruct);
const Record_ResponseRefStructArray = StructType({
  data: Record_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class Record_ResponseWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new Record_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.error_string =  new StringWrapper(other._wrapperFields.error_string);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new Record_ResponseRefStruct();
    this._wrapperFields.error_string = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new Record_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new Record_ResponseArrayWrapper();
  }
  static get ArrayType() {
    return Record_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return Record_ResponseRefStructArray;
  }
  static get refObjectType() {
    return Record_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.error_string.freeze(own);
    }
    this._refObject.error_string = this._wrapperFields.error_string.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.return_code = refObject.return_code;
    this._wrapperFields.error_string.data = refObject.error_string.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.error_string);
  }
  static destroyRawROS(msg) {
    Record_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rosbag2_interfaces', subFolder: 'srv', interfaceName: 'Record_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get return_code() {
    return this._refObject.return_code;
  }
  set return_code(value) {
    this._refObject.return_code = value;
  }
  get error_string() {
    return this._wrapperFields.error_string.data;
  }
  set error_string(value) {
    this._wrapperFields.error_string.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new Record_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.error_string.copyRefObject(this._refObject.error_string);
  }
  copy(other) {
    this._refObject = new Record_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.error_string.copy(other._wrapperFields.error_string);
  }
  static get classType() {
    return Record_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"return_code","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"error_string","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rosbag2_interfaces","type":"Record_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Record_Response"};
  }
  hasMember(name) {
    let memberNames = ["return_code","error_string"];
    return memberNames.indexOf(name) !== -1;
  }
}
class Record_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof Record_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new Record_ResponseWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to Record_ResponseArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to Record_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new Record_ResponseRefArray(size);
    this._refObject = new Record_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new Record_ResponseWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof Record_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "Record_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      Record_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return Record_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return Record_ResponseArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = Record_ResponseWrapper;
