// rclnodejs/generated/rosbag2_interfaces/rosbag2_interfaces__srv__Play_Request.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const TimeWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Time.js');
const DurationWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Duration.js');
const Play_RequestRefStruct = StructType({
  start_offset: TimeWrapper.refObjectType,
  playback_duration: DurationWrapper.refObjectType,
  playback_until_timestamp: TimeWrapper.refObjectType,
});
const Play_RequestRefArray = ArrayType(Play_RequestRefStruct);
const Play_RequestRefStructArray = StructType({
  data: Play_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class Play_RequestWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new Play_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.start_offset =  new TimeWrapper(other._wrapperFields.start_offset);
      this._wrapperFields.playback_duration =  new DurationWrapper(other._wrapperFields.playback_duration);
      this._wrapperFields.playback_until_timestamp =  new TimeWrapper(other._wrapperFields.playback_until_timestamp);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new Play_RequestRefStruct();
    this._wrapperFields.start_offset = new TimeWrapper();
    this._wrapperFields.playback_duration = new DurationWrapper();
    this._wrapperFields.playback_until_timestamp = new TimeWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new Play_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new Play_RequestArrayWrapper();
  }
  static get ArrayType() {
    return Play_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return Play_RequestRefStructArray;
  }
  static get refObjectType() {
    return Play_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.start_offset.freeze(own);
    this._refObject.start_offset = this._wrapperFields.start_offset.refObject;
    this._wrapperFields.playback_duration.freeze(own);
    this._refObject.playback_duration = this._wrapperFields.playback_duration.refObject;
    this._wrapperFields.playback_until_timestamp.freeze(own);
    this._refObject.playback_until_timestamp = this._wrapperFields.playback_until_timestamp.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.start_offset.copyRefObject(refObject.start_offset);
    this._wrapperFields.playback_duration.copyRefObject(refObject.playback_duration);
    this._wrapperFields.playback_until_timestamp.copyRefObject(refObject.playback_until_timestamp);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    TimeWrapper.freeStruct(refObject.start_offset);
    DurationWrapper.freeStruct(refObject.playback_duration);
    TimeWrapper.freeStruct(refObject.playback_until_timestamp);
  }
  static destroyRawROS(msg) {
    Play_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rosbag2_interfaces', subFolder: 'srv', interfaceName: 'Play_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get start_offset() {
    return this._wrapperFields.start_offset;
  }
  set start_offset(value) {
    if (value instanceof TimeWrapper) {
      this._wrapperFields.start_offset.copy(value);
    } else {
      this._wrapperFields.start_offset.copy(new TimeWrapper(value));
    }
  }
  get playback_duration() {
    return this._wrapperFields.playback_duration;
  }
  set playback_duration(value) {
    if (value instanceof DurationWrapper) {
      this._wrapperFields.playback_duration.copy(value);
    } else {
      this._wrapperFields.playback_duration.copy(new DurationWrapper(value));
    }
  }
  get playback_until_timestamp() {
    return this._wrapperFields.playback_until_timestamp;
  }
  set playback_until_timestamp(value) {
    if (value instanceof TimeWrapper) {
      this._wrapperFields.playback_until_timestamp.copy(value);
    } else {
      this._wrapperFields.playback_until_timestamp.copy(new TimeWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new Play_RequestRefStruct(refObject.toObject());
    this._wrapperFields.start_offset.copyRefObject(this._refObject.start_offset);
    this._wrapperFields.playback_duration.copyRefObject(this._refObject.playback_duration);
    this._wrapperFields.playback_until_timestamp.copyRefObject(this._refObject.playback_until_timestamp);
  }
  copy(other) {
    this._refObject = new Play_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.start_offset.copy(other._wrapperFields.start_offset);
    this._wrapperFields.playback_duration.copy(other._wrapperFields.playback_duration);
    this._wrapperFields.playback_until_timestamp.copy(other._wrapperFields.playback_until_timestamp);
  }
  static get classType() {
    return Play_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"start_offset","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Time","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"playback_duration","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Duration","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"playback_until_timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Time","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"rosbag2_interfaces","type":"Play_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Play_Request"};
  }
  hasMember(name) {
    let memberNames = ["start_offset","playback_duration","playback_until_timestamp"];
    return memberNames.indexOf(name) !== -1;
  }
}
class Play_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof Play_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new Play_RequestWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to Play_RequestArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to Play_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new Play_RequestRefArray(size);
    this._refObject = new Play_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new Play_RequestWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof Play_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "Play_RequestArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      Play_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return Play_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return Play_RequestArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = Play_RequestWrapper;
