// rclnodejs/generated/nav_msgs/nav_msgs__msg__TrajectoryPoint.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const PoseWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Pose.js');
const TwistWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Twist.js');
const AccelWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Accel.js');
const WrenchWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Wrench.js');
const TrajectoryPointRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  pose: PoseWrapper.refObjectType,
  velocity: TwistWrapper.refObjectType,
  acceleration: AccelWrapper.refObjectType,
  effort: WrenchWrapper.refObjectType,
});
const TrajectoryPointRefArray = ArrayType(TrajectoryPointRefStruct);
const TrajectoryPointRefStructArray = StructType({
  data: TrajectoryPointRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class TrajectoryPointWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new TrajectoryPointRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.pose =  new PoseWrapper(other._wrapperFields.pose);
      this._wrapperFields.velocity =  new TwistWrapper(other._wrapperFields.velocity);
      this._wrapperFields.acceleration =  new AccelWrapper(other._wrapperFields.acceleration);
      this._wrapperFields.effort =  new WrenchWrapper(other._wrapperFields.effort);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new TrajectoryPointRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.pose = new PoseWrapper();
    this._wrapperFields.velocity = new TwistWrapper();
    this._wrapperFields.acceleration = new AccelWrapper();
    this._wrapperFields.effort = new WrenchWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new TrajectoryPointWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new TrajectoryPointArrayWrapper();
  }
  static get ArrayType() {
    return TrajectoryPointArrayWrapper;
  }
  static get refObjectArrayType() {
    return TrajectoryPointRefStructArray;
  }
  static get refObjectType() {
    return TrajectoryPointRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.header.freeze(own);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.pose.freeze(own);
    this._refObject.pose = this._wrapperFields.pose.refObject;
    this._wrapperFields.velocity.freeze(own);
    this._refObject.velocity = this._wrapperFields.velocity.refObject;
    this._wrapperFields.acceleration.freeze(own);
    this._refObject.acceleration = this._wrapperFields.acceleration.refObject;
    this._wrapperFields.effort.freeze(own);
    this._refObject.effort = this._wrapperFields.effort.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.pose.copyRefObject(refObject.pose);
    this._wrapperFields.velocity.copyRefObject(refObject.velocity);
    this._wrapperFields.acceleration.copyRefObject(refObject.acceleration);
    this._wrapperFields.effort.copyRefObject(refObject.effort);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    PoseWrapper.freeStruct(refObject.pose);
    TwistWrapper.freeStruct(refObject.velocity);
    AccelWrapper.freeStruct(refObject.acceleration);
    WrenchWrapper.freeStruct(refObject.effort);
  }
  static destroyRawROS(msg) {
    TrajectoryPointWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav_msgs', subFolder: 'msg', interfaceName: 'TrajectoryPoint'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get pose() {
    return this._wrapperFields.pose;
  }
  set pose(value) {
    if (value instanceof PoseWrapper) {
      this._wrapperFields.pose.copy(value);
    } else {
      this._wrapperFields.pose.copy(new PoseWrapper(value));
    }
  }
  get velocity() {
    return this._wrapperFields.velocity;
  }
  set velocity(value) {
    if (value instanceof TwistWrapper) {
      this._wrapperFields.velocity.copy(value);
    } else {
      this._wrapperFields.velocity.copy(new TwistWrapper(value));
    }
  }
  get acceleration() {
    return this._wrapperFields.acceleration;
  }
  set acceleration(value) {
    if (value instanceof AccelWrapper) {
      this._wrapperFields.acceleration.copy(value);
    } else {
      this._wrapperFields.acceleration.copy(new AccelWrapper(value));
    }
  }
  get effort() {
    return this._wrapperFields.effort;
  }
  set effort(value) {
    if (value instanceof WrenchWrapper) {
      this._wrapperFields.effort.copy(value);
    } else {
      this._wrapperFields.effort.copy(new WrenchWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new TrajectoryPointRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.pose.copyRefObject(this._refObject.pose);
    this._wrapperFields.velocity.copyRefObject(this._refObject.velocity);
    this._wrapperFields.acceleration.copyRefObject(this._refObject.acceleration);
    this._wrapperFields.effort.copyRefObject(this._refObject.effort);
  }
  copy(other) {
    this._refObject = new TrajectoryPointRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.pose.copy(other._wrapperFields.pose);
    this._wrapperFields.velocity.copy(other._wrapperFields.velocity);
    this._wrapperFields.acceleration.copy(other._wrapperFields.acceleration);
    this._wrapperFields.effort.copy(other._wrapperFields.effort);
  }
  static get classType() {
    return TrajectoryPointWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"pose","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Pose","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"velocity","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Twist","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"acceleration","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Accel","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"effort","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Wrench","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"nav_msgs","type":"TrajectoryPoint","stringUpperBound":null,"isPrimitiveType":false},"msgName":"TrajectoryPoint"};
  }
  hasMember(name) {
    let memberNames = ["header","pose","velocity","acceleration","effort"];
    return memberNames.indexOf(name) !== -1;
  }
}
class TrajectoryPointArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof TrajectoryPointWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new TrajectoryPointWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TrajectoryPointArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TrajectoryPointArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new TrajectoryPointRefArray(size);
    this._refObject = new TrajectoryPointRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new TrajectoryPointWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof TrajectoryPointArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "TrajectoryPointArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      TrajectoryPointWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return TrajectoryPointWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return TrajectoryPointArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = TrajectoryPointWrapper;
