// rclnodejs/generated/service_msgs/service_msgs__msg__ServiceEventInfo.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const TimeWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Time.js');
const CharWrapper = require('../../generated/std_msgs/std_msgs__msg__Char.js');
const ServiceEventInfoRefStruct = StructType({
  event_type: primitiveTypes.uint8,
  stamp: TimeWrapper.refObjectType,
  client_gid: ArrayType(primitiveTypes.char, 16),
  sequence_number: primitiveTypes.int64,
});
const ServiceEventInfoRefArray = ArrayType(ServiceEventInfoRefStruct);
const ServiceEventInfoRefStructArray = StructType({
  data: ServiceEventInfoRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class ServiceEventInfoWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new ServiceEventInfoRefStruct(other._refObject.toObject());
      this._wrapperFields.stamp =  new TimeWrapper(other._wrapperFields.stamp);
      this._wrapperFields.client_gid = CharWrapper.createArray();
      this._wrapperFields.client_gid.copy(other._wrapperFields.client_gid);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new ServiceEventInfoRefStruct();
    this._wrapperFields.stamp = new TimeWrapper();
    this._wrapperFields.client_gid = CharWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new ServiceEventInfoWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ServiceEventInfoArrayWrapper();
  }
  static get ArrayType() {
    return ServiceEventInfoArrayWrapper;
  }
  static get refObjectArrayType() {
    return ServiceEventInfoRefStructArray;
  }
  static get refObjectType() {
    return ServiceEventInfoRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.stamp.freeze(own);
    this._refObject.stamp = this._wrapperFields.stamp.refObject;
    this._refObject.client_gid = Array.from(this._wrapperFields.client_gid.data);
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.event_type = refObject.event_type;
    this._wrapperFields.stamp.copyRefObject(refObject.stamp);
    this._wrapperFields.client_gid.fill(refObject.client_gid.toArray());
    this._refObject.sequence_number = refObject.sequence_number;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    TimeWrapper.freeStruct(refObject.stamp);
  }
  static destroyRawROS(msg) {
    ServiceEventInfoWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'service_msgs', subFolder: 'msg', interfaceName: 'ServiceEventInfo'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get event_type() {
    return this._refObject.event_type;
  }
  set event_type(value) {
    this._refObject.event_type = value;
  }
  get stamp() {
    return this._wrapperFields.stamp;
  }
  set stamp(value) {
    if (value instanceof TimeWrapper) {
      this._wrapperFields.stamp.copy(value);
    } else {
      this._wrapperFields.stamp.copy(new TimeWrapper(value));
    }
  }
  get client_gid() {
    return this._wrapperFields['client_gid'].data;
  }
  set client_gid(value) {
    if (value.length !== 16) {
      throw new RangeError('The length of the array must be 16.');
    }
    this._wrapperFields['client_gid'].fill(value);
  }
  get sequence_number() {
    return BigInt(this._refObject.sequence_number);
  }
  set sequence_number(value) {
    if (typeof value !== "bigint") {
      throw new TypeError('sequence_number must be type of bigint');
    }
    this._refObject.sequence_number = value.toString();
  }
  copyRefObject(refObject) {
    this._refObject = new ServiceEventInfoRefStruct(refObject.toObject());
    this._wrapperFields.stamp.copyRefObject(this._refObject.stamp);
    this._wrapperFields.client_gid.fill(refObject.client_gid.toArray());
  }
  copy(other) {
    this._refObject = new ServiceEventInfoRefStruct(other._refObject.toObject());
    this._wrapperFields.stamp.copy(other._wrapperFields.stamp);
    this._wrapperFields.client_gid.copy(other._wrapperFields.client_gid);
  }
  static get classType() {
    return ServiceEventInfoWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"REQUEST_SENT","value":0},{"type":"uint8","name":"REQUEST_RECEIVED","value":1},{"type":"uint8","name":"RESPONSE_SENT","value":2},{"type":"uint8","name":"RESPONSE_RECEIVED","value":3}],"fields":[{"name":"event_type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"stamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Time","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"client_gid","type":{"isArray":true,"arraySize":16,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"char","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"sequence_number","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"service_msgs","type":"ServiceEventInfo","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ServiceEventInfo"};
  }
  hasMember(name) {
    let memberNames = ["event_type","stamp","client_gid","sequence_number"];
    return memberNames.indexOf(name) !== -1;
  }
}
class ServiceEventInfoArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ServiceEventInfoWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ServiceEventInfoWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ServiceEventInfoArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ServiceEventInfoArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new ServiceEventInfoRefArray(size);
    this._refObject = new ServiceEventInfoRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ServiceEventInfoWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ServiceEventInfoArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ServiceEventInfoArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ServiceEventInfoWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ServiceEventInfoWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ServiceEventInfoArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
Object.defineProperty(ServiceEventInfoWrapper, "REQUEST_SENT", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ServiceEventInfoWrapper, "REQUEST_RECEIVED", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ServiceEventInfoWrapper, "RESPONSE_SENT", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ServiceEventInfoWrapper, "RESPONSE_RECEIVED", {value: 3, writable: false, enumerable: true, configurable: true});
module.exports = ServiceEventInfoWrapper;
