// rclnodejs/generated/ur_dashboard_msgs/ur_dashboard_msgs__action__SetMode.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
class ur_dashboard_msgs__action__SetMode {
  static get Goal() {
    return require('./ur_dashboard_msgs__action__SetMode_Goal.js');
  }
  static get Result() {
    return require('./ur_dashboard_msgs__action__SetMode_Result.js');
  }
  static get Feedback() {
    return require('./ur_dashboard_msgs__action__SetMode_Feedback.js');
  }
  static get impl() {
    return {
      get SendGoalService() {
        return require('./ur_dashboard_msgs__action__SetMode_SendGoal.js');
      },
      get GetResultService() {
        return require('./ur_dashboard_msgs__action__SetMode_GetResult.js');
      },
      get FeedbackMessage() {
        return require('./ur_dashboard_msgs__action__SetMode_FeedbackMessage.js');
      },
      get CancelGoal() {
        return require('../action_msgs/action_msgs__srv__CancelGoal.js');
      },
      get GoalStatusArray() {
        return require('../action_msgs/action_msgs__msg__GoalStatusArray.js');
      }
    };
  }
  static type() {
    return {pkgName: 'ur_dashboard_msgs', subFolder: 'action', interfaceName: 'SetMode'};
  }
}
module.exports = ur_dashboard_msgs__action__SetMode;
