// rclnodejs/generated/ur_dashboard_msgs/ur_dashboard_msgs__srv__RawRequest_Request.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const RawRequest_RequestRefStruct = StructType({
  query: primitiveTypes.string,
});
const RawRequest_RequestRefArray = ArrayType(RawRequest_RequestRefStruct);
const RawRequest_RequestRefStructArray = StructType({
  data: RawRequest_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class RawRequest_RequestWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new RawRequest_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.query =  new StringWrapper(other._wrapperFields.query);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new RawRequest_RequestRefStruct();
    this._wrapperFields.query = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new RawRequest_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new RawRequest_RequestArrayWrapper();
  }
  static get ArrayType() {
    return RawRequest_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return RawRequest_RequestRefStructArray;
  }
  static get refObjectType() {
    return RawRequest_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.query.freeze(own);
    }
    this._refObject.query = this._wrapperFields.query.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.query.data = refObject.query.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.query);
  }
  static destroyRawROS(msg) {
    RawRequest_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'ur_dashboard_msgs', subFolder: 'srv', interfaceName: 'RawRequest_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get query() {
    return this._wrapperFields.query.data;
  }
  set query(value) {
    this._wrapperFields.query.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new RawRequest_RequestRefStruct(refObject.toObject());
    this._wrapperFields.query.copyRefObject(this._refObject.query);
  }
  copy(other) {
    this._refObject = new RawRequest_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.query.copy(other._wrapperFields.query);
  }
  static get classType() {
    return RawRequest_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"query","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"ur_dashboard_msgs","type":"RawRequest_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"RawRequest_Request"};
  }
  hasMember(name) {
    let memberNames = ["query"];
    return memberNames.indexOf(name) !== -1;
  }
}
class RawRequest_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof RawRequest_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new RawRequest_RequestWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RawRequest_RequestArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RawRequest_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new RawRequest_RequestRefArray(size);
    this._refObject = new RawRequest_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new RawRequest_RequestWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof RawRequest_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "RawRequest_RequestArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      RawRequest_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return RawRequest_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return RawRequest_RequestArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = RawRequest_RequestWrapper;
