// rclnodejs/generated/ur_dashboard_msgs/ur_dashboard_msgs__srv__IsProgramSaved_Response.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const IsProgramSaved_ResponseRefStruct = StructType({
  answer: primitiveTypes.string,
  program_name: primitiveTypes.string,
  program_saved: primitiveTypes.bool,
  success: primitiveTypes.bool,
});
const IsProgramSaved_ResponseRefArray = ArrayType(IsProgramSaved_ResponseRefStruct);
const IsProgramSaved_ResponseRefStructArray = StructType({
  data: IsProgramSaved_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class IsProgramSaved_ResponseWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new IsProgramSaved_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.answer =  new StringWrapper(other._wrapperFields.answer);
      this._wrapperFields.program_name =  new StringWrapper(other._wrapperFields.program_name);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new IsProgramSaved_ResponseRefStruct();
    this._wrapperFields.answer = new StringWrapper();
    this._wrapperFields.program_name = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new IsProgramSaved_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new IsProgramSaved_ResponseArrayWrapper();
  }
  static get ArrayType() {
    return IsProgramSaved_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return IsProgramSaved_ResponseRefStructArray;
  }
  static get refObjectType() {
    return IsProgramSaved_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.answer.freeze(own);
    }
    this._refObject.answer = this._wrapperFields.answer.refObject;
    if (own) {
      this._wrapperFields.program_name.freeze(own);
    }
    this._refObject.program_name = this._wrapperFields.program_name.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.answer.data = refObject.answer.data;
    this._wrapperFields.program_name.data = refObject.program_name.data;
    this._refObject.program_saved = refObject.program_saved;
    this._refObject.success = refObject.success;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.answer);
    StringWrapper.freeStruct(refObject.program_name);
  }
  static destroyRawROS(msg) {
    IsProgramSaved_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'ur_dashboard_msgs', subFolder: 'srv', interfaceName: 'IsProgramSaved_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get answer() {
    return this._wrapperFields.answer.data;
  }
  set answer(value) {
    this._wrapperFields.answer.data = value;
  }
  get program_name() {
    return this._wrapperFields.program_name.data;
  }
  set program_name(value) {
    this._wrapperFields.program_name.data = value;
  }
  get program_saved() {
    return this._refObject.program_saved;
  }
  set program_saved(value) {
    this._refObject.program_saved = value;
  }
  get success() {
    return this._refObject.success;
  }
  set success(value) {
    this._refObject.success = value;
  }
  copyRefObject(refObject) {
    this._refObject = new IsProgramSaved_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.answer.copyRefObject(this._refObject.answer);
    this._wrapperFields.program_name.copyRefObject(this._refObject.program_name);
  }
  copy(other) {
    this._refObject = new IsProgramSaved_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.answer.copy(other._wrapperFields.answer);
    this._wrapperFields.program_name.copy(other._wrapperFields.program_name);
  }
  static get classType() {
    return IsProgramSaved_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"answer","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"program_name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"program_saved","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"success","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"ur_dashboard_msgs","type":"IsProgramSaved_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"IsProgramSaved_Response"};
  }
  hasMember(name) {
    let memberNames = ["answer","program_name","program_saved","success"];
    return memberNames.indexOf(name) !== -1;
  }
}
class IsProgramSaved_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof IsProgramSaved_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new IsProgramSaved_ResponseWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to IsProgramSaved_ResponseArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to IsProgramSaved_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new IsProgramSaved_ResponseRefArray(size);
    this._refObject = new IsProgramSaved_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new IsProgramSaved_ResponseWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof IsProgramSaved_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "IsProgramSaved_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      IsProgramSaved_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return IsProgramSaved_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return IsProgramSaved_ResponseArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = IsProgramSaved_ResponseWrapper;
