// rclnodejs/generated/ur_dashboard_msgs/ur_dashboard_msgs__action__SetMode_Result.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const SetMode_ResultRefStruct = StructType({
  success: primitiveTypes.bool,
  message: primitiveTypes.string,
});
const SetMode_ResultRefArray = ArrayType(SetMode_ResultRefStruct);
const SetMode_ResultRefStructArray = StructType({
  data: SetMode_ResultRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class SetMode_ResultWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new SetMode_ResultRefStruct(other._refObject.toObject());
      this._wrapperFields.message =  new StringWrapper(other._wrapperFields.message);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new SetMode_ResultRefStruct();
    this._wrapperFields.message = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new SetMode_ResultWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SetMode_ResultArrayWrapper();
  }
  static get ArrayType() {
    return SetMode_ResultArrayWrapper;
  }
  static get refObjectArrayType() {
    return SetMode_ResultRefStructArray;
  }
  static get refObjectType() {
    return SetMode_ResultRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.message.freeze(own);
    }
    this._refObject.message = this._wrapperFields.message.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.success = refObject.success;
    this._wrapperFields.message.data = refObject.message.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.message);
  }
  static destroyRawROS(msg) {
    SetMode_ResultWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'ur_dashboard_msgs', subFolder: 'action', interfaceName: 'SetMode_Result'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get success() {
    return this._refObject.success;
  }
  set success(value) {
    this._refObject.success = value;
  }
  get message() {
    return this._wrapperFields.message.data;
  }
  set message(value) {
    this._wrapperFields.message.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new SetMode_ResultRefStruct(refObject.toObject());
    this._wrapperFields.message.copyRefObject(this._refObject.message);
  }
  copy(other) {
    this._refObject = new SetMode_ResultRefStruct(other._refObject.toObject());
    this._wrapperFields.message.copy(other._wrapperFields.message);
  }
  static get classType() {
    return SetMode_ResultWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"success","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"message","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"ur_dashboard_msgs","type":"SetMode_Result","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SetMode_Result"};
  }
  hasMember(name) {
    let memberNames = ["success","message"];
    return memberNames.indexOf(name) !== -1;
  }
}
class SetMode_ResultArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SetMode_ResultWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SetMode_ResultWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetMode_ResultArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetMode_ResultArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new SetMode_ResultRefArray(size);
    this._refObject = new SetMode_ResultRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SetMode_ResultWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SetMode_ResultArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SetMode_ResultArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SetMode_ResultWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SetMode_ResultWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SetMode_ResultArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = SetMode_ResultWrapper;
