// rclnodejs/generated/ur_dashboard_msgs/ur_dashboard_msgs__msg__RobotMode.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const RobotModeRefStruct = StructType({
  mode: primitiveTypes.int8,
});
const RobotModeRefArray = ArrayType(RobotModeRefStruct);
const RobotModeRefStructArray = StructType({
  data: RobotModeRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class RobotModeWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new RobotModeRefStruct(other._refObject.toObject());
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new RobotModeRefStruct();
  }
  static createFromRefObject(refObject) {
    let self = new RobotModeWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new RobotModeArrayWrapper();
  }
  static get ArrayType() {
    return RobotModeArrayWrapper;
  }
  static get refObjectArrayType() {
    return RobotModeRefStructArray;
  }
  static get refObjectType() {
    return RobotModeRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.mode = refObject.mode;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destroyRawROS(msg) {
    RobotModeWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'ur_dashboard_msgs', subFolder: 'msg', interfaceName: 'RobotMode'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get mode() {
    return this._refObject.mode;
  }
  set mode(value) {
    this._refObject.mode = value;
  }
  copyRefObject(refObject) {
    this._refObject = new RobotModeRefStruct(refObject.toObject());
  }
  copy(other) {
    this._refObject = new RobotModeRefStruct(other._refObject.toObject());
  }
  static get classType() {
    return RobotModeWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"int8","name":"NO_CONTROLLER","value":-1},{"type":"int8","name":"DISCONNECTED","value":0},{"type":"int8","name":"CONFIRM_SAFETY","value":1},{"type":"int8","name":"BOOTING","value":2},{"type":"int8","name":"POWER_OFF","value":3},{"type":"int8","name":"POWER_ON","value":4},{"type":"int8","name":"IDLE","value":5},{"type":"int8","name":"BACKDRIVE","value":6},{"type":"int8","name":"RUNNING","value":7},{"type":"int8","name":"UPDATING_FIRMWARE","value":8}],"fields":[{"name":"mode","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"ur_dashboard_msgs","type":"RobotMode","stringUpperBound":null,"isPrimitiveType":false},"msgName":"RobotMode"};
  }
  hasMember(name) {
    let memberNames = ["mode"];
    return memberNames.indexOf(name) !== -1;
  }
}
class RobotModeArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof RobotModeWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new RobotModeWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RobotModeArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RobotModeArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new RobotModeRefArray(size);
    this._refObject = new RobotModeRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new RobotModeWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof RobotModeArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "RobotModeArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      RobotModeWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return RobotModeWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return RobotModeArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
Object.defineProperty(RobotModeWrapper, "NO_CONTROLLER", {value: -1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(RobotModeWrapper, "DISCONNECTED", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(RobotModeWrapper, "CONFIRM_SAFETY", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(RobotModeWrapper, "BOOTING", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(RobotModeWrapper, "POWER_OFF", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(RobotModeWrapper, "POWER_ON", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(RobotModeWrapper, "IDLE", {value: 5, writable: false, enumerable: true, configurable: true});
Object.defineProperty(RobotModeWrapper, "BACKDRIVE", {value: 6, writable: false, enumerable: true, configurable: true});
Object.defineProperty(RobotModeWrapper, "RUNNING", {value: 7, writable: false, enumerable: true, configurable: true});
Object.defineProperty(RobotModeWrapper, "UPDATING_FIRMWARE", {value: 8, writable: false, enumerable: true, configurable: true});
module.exports = RobotModeWrapper;
