// rclnodejs/generated/moveit_msgs/moveit_msgs__msg__PipelineState.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const MotionPlanRequestWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__MotionPlanRequest.js');
const MotionPlanResponseWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__MotionPlanResponse.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const PipelineStateRefStruct = StructType({
  request: MotionPlanRequestWrapper.refObjectType,
  response: MotionPlanResponseWrapper.refObjectType,
  pipeline_stage: primitiveTypes.string,
});
const PipelineStateRefArray = ArrayType(PipelineStateRefStruct);
const PipelineStateRefStructArray = StructType({
  data: PipelineStateRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class PipelineStateWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new PipelineStateRefStruct(other._refObject.toObject());
      this._wrapperFields.request =  new MotionPlanRequestWrapper(other._wrapperFields.request);
      this._wrapperFields.response =  new MotionPlanResponseWrapper(other._wrapperFields.response);
      this._wrapperFields.pipeline_stage =  new StringWrapper(other._wrapperFields.pipeline_stage);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new PipelineStateRefStruct();
    this._wrapperFields.request = new MotionPlanRequestWrapper();
    this._wrapperFields.response = new MotionPlanResponseWrapper();
    this._wrapperFields.pipeline_stage = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new PipelineStateWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new PipelineStateArrayWrapper();
  }
  static get ArrayType() {
    return PipelineStateArrayWrapper;
  }
  static get refObjectArrayType() {
    return PipelineStateRefStructArray;
  }
  static get refObjectType() {
    return PipelineStateRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.request.freeze(own);
    this._refObject.request = this._wrapperFields.request.refObject;
    this._wrapperFields.response.freeze(own);
    this._refObject.response = this._wrapperFields.response.refObject;
    if (own) {
      this._wrapperFields.pipeline_stage.freeze(own);
    }
    this._refObject.pipeline_stage = this._wrapperFields.pipeline_stage.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.request.copyRefObject(refObject.request);
    this._wrapperFields.response.copyRefObject(refObject.response);
    this._wrapperFields.pipeline_stage.data = refObject.pipeline_stage.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    MotionPlanRequestWrapper.freeStruct(refObject.request);
    MotionPlanResponseWrapper.freeStruct(refObject.response);
    StringWrapper.freeStruct(refObject.pipeline_stage);
  }
  static destroyRawROS(msg) {
    PipelineStateWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'msg', interfaceName: 'PipelineState'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get request() {
    return this._wrapperFields.request;
  }
  set request(value) {
    if (value instanceof MotionPlanRequestWrapper) {
      this._wrapperFields.request.copy(value);
    } else {
      this._wrapperFields.request.copy(new MotionPlanRequestWrapper(value));
    }
  }
  get response() {
    return this._wrapperFields.response;
  }
  set response(value) {
    if (value instanceof MotionPlanResponseWrapper) {
      this._wrapperFields.response.copy(value);
    } else {
      this._wrapperFields.response.copy(new MotionPlanResponseWrapper(value));
    }
  }
  get pipeline_stage() {
    return this._wrapperFields.pipeline_stage.data;
  }
  set pipeline_stage(value) {
    this._wrapperFields.pipeline_stage.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new PipelineStateRefStruct(refObject.toObject());
    this._wrapperFields.request.copyRefObject(this._refObject.request);
    this._wrapperFields.response.copyRefObject(this._refObject.response);
    this._wrapperFields.pipeline_stage.copyRefObject(this._refObject.pipeline_stage);
  }
  copy(other) {
    this._refObject = new PipelineStateRefStruct(other._refObject.toObject());
    this._wrapperFields.request.copy(other._wrapperFields.request);
    this._wrapperFields.response.copy(other._wrapperFields.response);
    this._wrapperFields.pipeline_stage.copy(other._wrapperFields.pipeline_stage);
  }
  static get classType() {
    return PipelineStateWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"request","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"MotionPlanRequest","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"response","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"MotionPlanResponse","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"pipeline_stage","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"PipelineState","stringUpperBound":null,"isPrimitiveType":false},"msgName":"PipelineState"};
  }
  hasMember(name) {
    let memberNames = ["request","response","pipeline_stage"];
    return memberNames.indexOf(name) !== -1;
  }
}
class PipelineStateArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof PipelineStateWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new PipelineStateWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PipelineStateArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PipelineStateArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new PipelineStateRefArray(size);
    this._refObject = new PipelineStateRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new PipelineStateWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof PipelineStateArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "PipelineStateArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      PipelineStateWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return PipelineStateWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return PipelineStateArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = PipelineStateWrapper;
