// rclnodejs/generated/moveit_msgs/moveit_msgs__srv__GetGroupUrdf_Request.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const GetGroupUrdf_RequestRefStruct = StructType({
  group_name: primitiveTypes.string,
});
const GetGroupUrdf_RequestRefArray = ArrayType(GetGroupUrdf_RequestRefStruct);
const GetGroupUrdf_RequestRefStructArray = StructType({
  data: GetGroupUrdf_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class GetGroupUrdf_RequestWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetGroupUrdf_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.group_name =  new StringWrapper(other._wrapperFields.group_name);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new GetGroupUrdf_RequestRefStruct();
    this._wrapperFields.group_name = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new GetGroupUrdf_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetGroupUrdf_RequestArrayWrapper();
  }
  static get ArrayType() {
    return GetGroupUrdf_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetGroupUrdf_RequestRefStructArray;
  }
  static get refObjectType() {
    return GetGroupUrdf_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.group_name.freeze(own);
    }
    this._refObject.group_name = this._wrapperFields.group_name.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.group_name.data = refObject.group_name.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.group_name);
  }
  static destroyRawROS(msg) {
    GetGroupUrdf_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'srv', interfaceName: 'GetGroupUrdf_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get group_name() {
    return this._wrapperFields.group_name.data;
  }
  set group_name(value) {
    this._wrapperFields.group_name.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GetGroupUrdf_RequestRefStruct(refObject.toObject());
    this._wrapperFields.group_name.copyRefObject(this._refObject.group_name);
  }
  copy(other) {
    this._refObject = new GetGroupUrdf_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.group_name.copy(other._wrapperFields.group_name);
  }
  static get classType() {
    return GetGroupUrdf_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"group_name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"GetGroupUrdf_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetGroupUrdf_Request"};
  }
  hasMember(name) {
    let memberNames = ["group_name"];
    return memberNames.indexOf(name) !== -1;
  }
}
class GetGroupUrdf_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetGroupUrdf_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetGroupUrdf_RequestWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetGroupUrdf_RequestArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetGroupUrdf_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new GetGroupUrdf_RequestRefArray(size);
    this._refObject = new GetGroupUrdf_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetGroupUrdf_RequestWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetGroupUrdf_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetGroupUrdf_RequestArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetGroupUrdf_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetGroupUrdf_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetGroupUrdf_RequestArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = GetGroupUrdf_RequestWrapper;
