// rclnodejs/generated/moveit_msgs/moveit_msgs__srv__UpdatePointcloudOctomap_Event.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const ServiceEventInfoWrapper = require('../../generated/service_msgs/service_msgs__msg__ServiceEventInfo.js');
const UpdatePointcloudOctomap_RequestWrapper = require('../../generated/moveit_msgs/moveit_msgs__srv__UpdatePointcloudOctomap_Request.js');
const UpdatePointcloudOctomap_ResponseWrapper = require('../../generated/moveit_msgs/moveit_msgs__srv__UpdatePointcloudOctomap_Response.js');
const UpdatePointcloudOctomapRefStruct = StructType({
  info: ServiceEventInfoWrapper.refObjectType,
  request: UpdatePointcloudOctomap_RequestWrapper.refObjectArrayType,
  response: UpdatePointcloudOctomap_ResponseWrapper.refObjectArrayType,
});
const UpdatePointcloudOctomapRefArray = ArrayType(UpdatePointcloudOctomapRefStruct);
const UpdatePointcloudOctomapRefStructArray = StructType({
  data: UpdatePointcloudOctomapRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class UpdatePointcloudOctomapWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new UpdatePointcloudOctomapRefStruct(other._refObject.toObject());
      this._wrapperFields.info =  new ServiceEventInfoWrapper(other._wrapperFields.info);
      this._wrapperFields.request = UpdatePointcloudOctomap_RequestWrapper.createArray();
      this._wrapperFields.request.copy(other._wrapperFields.request);
      this._wrapperFields.response = UpdatePointcloudOctomap_ResponseWrapper.createArray();
      this._wrapperFields.response.copy(other._wrapperFields.response);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new UpdatePointcloudOctomapRefStruct();
    this._wrapperFields.info = new ServiceEventInfoWrapper();
    this._wrapperFields.request = UpdatePointcloudOctomap_RequestWrapper.createArray();
    this._wrapperFields.response = UpdatePointcloudOctomap_ResponseWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new UpdatePointcloudOctomapWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new UpdatePointcloudOctomapArrayWrapper();
  }
  static get ArrayType() {
    return UpdatePointcloudOctomapArrayWrapper;
  }
  static get refObjectArrayType() {
    return UpdatePointcloudOctomapRefStructArray;
  }
  static get refObjectType() {
    return UpdatePointcloudOctomapRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.info.freeze(own);
    this._refObject.info = this._wrapperFields.info.refObject;
    if (own) {
      this._wrapperFields.request.fill([]);
    }
    this._wrapperFields.request.freeze(own);
    this._refObject.request = this._wrapperFields.request.refObject;
    if (own) {
      this._wrapperFields.response.fill([]);
    }
    this._wrapperFields.response.freeze(own);
    this._refObject.response = this._wrapperFields.response.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.info.copyRefObject(refObject.info);
    this._wrapperFields.request.copyRefObject(refObject.request);
    this._wrapperFields.response.copyRefObject(refObject.response);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    ServiceEventInfoWrapper.freeStruct(refObject.info);
    if (refObject.request.size != 0) {
      UpdatePointcloudOctomap_RequestWrapper.ArrayType.freeArray(refObject.request);
      if (!UpdatePointcloudOctomap_RequestWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.request, UpdatePointcloudOctomap_RequestWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.response.size != 0) {
      UpdatePointcloudOctomap_ResponseWrapper.ArrayType.freeArray(refObject.response);
      if (!UpdatePointcloudOctomap_ResponseWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.response, UpdatePointcloudOctomap_ResponseWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    UpdatePointcloudOctomapWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'srv', interfaceName: 'UpdatePointcloudOctomap_Event'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get info() {
    return this._wrapperFields.info;
  }
  set info(value) {
    if (value instanceof ServiceEventInfoWrapper) {
      this._wrapperFields.info.copy(value);
    } else {
      this._wrapperFields.info.copy(new ServiceEventInfoWrapper(value));
    }
  }
  get request() {
    return this._wrapperFields.request;
  }
  set request(value) {
    if (value.length > 1) {
      throw new RangeError('The length of array request must be <= 1.');
    }
    this._wrapperFields.request.fill(value);
  }
  get response() {
    return this._wrapperFields.response;
  }
  set response(value) {
    if (value.length > 1) {
      throw new RangeError('The length of array response must be <= 1.');
    }
    this._wrapperFields.response.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new UpdatePointcloudOctomapRefStruct(refObject.toObject());
    this._wrapperFields.info.copyRefObject(this._refObject.info);
    this._wrapperFields.request.copyRefObject(this._refObject.request);
    this._wrapperFields.response.copyRefObject(this._refObject.response);
  }
  copy(other) {
    this._refObject = new UpdatePointcloudOctomapRefStruct(other._refObject.toObject());
    this._wrapperFields.info.copy(other._wrapperFields.info);
    this._wrapperFields.request.copy(other._wrapperFields.request);
    this._wrapperFields.response.copy(other._wrapperFields.response);
  }
  static get classType() {
    return UpdatePointcloudOctomapWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"info","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"service_msgs","type":"ServiceEventInfo","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"request","type":{"isArray":true,"arraySize":1,"isUpperBound":true,"isDynamicArray":true,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"UpdatePointcloudOctomap_Request","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"response","type":{"isArray":true,"arraySize":1,"isUpperBound":true,"isDynamicArray":true,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"UpdatePointcloudOctomap_Response","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"UpdatePointcloudOctomap","stringUpperBound":null,"isPrimitiveType":false},"msgName":"UpdatePointcloudOctomap","subFolder":"srv"};
  }
  hasMember(name) {
    let memberNames = ["info","request","response"];
    return memberNames.indexOf(name) !== -1;
  }
}
class UpdatePointcloudOctomapArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof UpdatePointcloudOctomapWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new UpdatePointcloudOctomapWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to UpdatePointcloudOctomapArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to UpdatePointcloudOctomapArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new UpdatePointcloudOctomapRefArray(size);
    this._refObject = new UpdatePointcloudOctomapRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new UpdatePointcloudOctomapWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof UpdatePointcloudOctomapArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "UpdatePointcloudOctomapArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      UpdatePointcloudOctomapWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return UpdatePointcloudOctomapWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return UpdatePointcloudOctomapArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = UpdatePointcloudOctomapWrapper;
