// @chart-sg/node-red-ros2-manager/lib/ros2-shared-manager.js
/**
 * Shared ROS2 Context Manager
 * 
 * This module provides a centralized way to manage ROS2 initialization and node creation
 * across multiple Node-RED plugins (RMF and node-red-ros2).
 * 
 * Key features:
 * - Single ROS2 context per Node-RED process
 * - Automatic initialization and cleanup
 * - Node lifecycle management
 * - Reference counting for safe cleanup
 * - Error handling and recovery
 * - Global configuration from ros2-config node
 */

// Try to import rclnodejs with graceful fallback
let rclnodejs = null;
let rclnodejsAvailable = false;

try {
    rclnodejs = require('rclnodejs');
    rclnodejsAvailable = true;
    console.log('[ROS2SharedManager] rclnodejs loaded successfully');
} catch (error) {
    console.warn('[ROS2SharedManager] rclnodejs not available:', error.message);
    console.warn('[ROS2SharedManager] ROS2 functionality will be disabled until rclnodejs is properly installed');
}

class ROS2SharedManager {
    constructor() {
        this.initialized = false;
        this.available = rclnodejsAvailable;
        this.initializing = false;
        this.context = null;
        this.nodes = new Map(); // nodeId -> node instance
        this.nodeCounter = 0;
        this.referenceCount = 0;
        this.spinning = false;
        this.spinStarted = false; // Prevent multiple spin attempts
        this.spinInterval = null; // For multi-node spinning
        this.domain = null;
        this.namespace = '';
        this.shutdownCallbacks = new Set();
        this.initializationPromise = null;
        this.configuration = {
            domain: 0,
            namespace: ''
        };
        this.lastHealthLog = null; // Track last health log time
        this.lastHealthStatus = null; // Track last health status to avoid duplicate logs
        
        // Action client management
        this.actionClients = new Map(); // actionClientId -> action client info
        this.actionClientCounter = 0;
    }

    /**
     * Configure SharedManager with global settings from ros2-config node
     * @param {Object} config - Configuration object
     * @param {number} config.domain - ROS2 domain ID
     * @param {string} config.namespace - ROS2 namespace
     */
    configure(config) {
        const isDebug = process.env.ROS2_MANAGER_DEBUG === 'true';
        
        // Validate domain
        if (config.domain !== undefined) {
            const domain = parseInt(config.domain);
            if (isNaN(domain) || domain < 0 || domain > 232) {
                throw new Error(`Invalid domain ID: ${config.domain}. Must be between 0-232.`);
            }
            
            // Check for domain conflicts if already initialized
            if (this.initialized && this.domain !== null && this.domain !== domain) {
                console.warn(`[ROS2SharedManager] Domain conflict detected: Already using domain ${this.domain}, ignoring new domain ${domain}`);
                return;
            }
            
            this.configuration.domain = domain;
            if (isDebug) {
                console.log(`[ROS2SharedManager] Domain configured: ${domain}`);
            }
        }
        
        // Configure namespace
        if (config.namespace !== undefined) {
            let namespace = config.namespace.trim();
            
            // Ensure namespace starts with / if not empty
            if (namespace && !namespace.startsWith('/')) {
                namespace = '/' + namespace;
            }
            
            this.configuration.namespace = namespace;
            this.namespace = namespace;
            
            if (isDebug) {
                console.log(`[ROS2SharedManager] Namespace configured: "${namespace}"`);
            }
        }
        
        // Log effective configuration
        if (isDebug) {
            console.log(`[ROS2SharedManager] Effective configuration: domain=${this.configuration.domain}, namespace="${this.configuration.namespace}"`);
        }
    }

    /**
     * Get current configuration
     * @returns {Object} Current configuration object
     */
    getConfiguration() {
        return {
            domain: this.configuration.domain,
            namespace: this.configuration.namespace,
            initialized: this.initialized,
            nodeCount: this.nodes.size
        };
    }

    /**
     * Initialize ROS2 context with domain configuration
     * @param {Object} options - Initialization options
     * @param {number} options.domain - ROS2 domain ID (optional)
     * @param {Array} options.args - Additional ROS2 arguments (optional)
     * @returns {Promise<void>}
     */
    async initialize(options = {}) {
        // Check if rclnodejs is available
        if (!this.available) {
            throw new Error(`
ROS2 functionality is not available. rclnodejs is not properly installed.

To enable ROS2 functionality:
1. Source your ROS2 environment:
   source /opt/ros/jazzy/setup.bash  # (or your ROS2 distro)
   
2. For RMF users, also source RMF workspace:
   source ~/rmf_ws/install/setup.bash

3. Install rclnodejs in your Node-RED directory:
   cd ~/.node-red
   npm install rclnodejs

4. Restart Node-RED

Current ROS2 environment: ${process.env.ROS_DISTRO || 'Not sourced'}
            `.trim());
        }

        if (this.initialized) {
            this.referenceCount++;
            return;
        }

        if (this.initializing) {
            // Wait for ongoing initialization
            return this.initializationPromise;
        }

        this.initializing = true;
        this.initializationPromise = this._doInitialize(options);
        
        try {
            await this.initializationPromise;
            this.initialized = true;
            this.referenceCount++;
            
            // Start spinning if we have nodes and haven't started yet
            if (this.nodes.size > 0 && !this.spinStarted) {
                console.log('[ROS2SharedManager] Starting spinning after initialization completion');
                // Don't await - let it run in background
                this.startSpinning().catch(error => {
                    console.error('[ROS2SharedManager] Post-init spinning failed:', error);
                });
            }
        } catch (error) {
            this.initializing = false;
            this.initializationPromise = null;
            throw error;
        } finally {
            this.initializing = false;
        }
    }

    async _doInitialize(options) {
        try {
            const args = options.args || [];
            
            // Use configured domain if available, otherwise fall back to options.domain
            const effectiveDomain = this.configuration.domain !== undefined ? this.configuration.domain : options.domain;
            
            // Set domain if specified
            if (effectiveDomain !== undefined) {
                process.env.ROS_DOMAIN_ID = effectiveDomain.toString();
                this.domain = effectiveDomain;
            }

            // Initialize ROS2
            await rclnodejs.init(...args);
            
            // Store context reference (rclnodejs manages context internally)
            // We don't need to access the context directly
            this.context = true; // Just mark as initialized
            
            const isDebug = process.env.ROS2_MANAGER_DEBUG === 'true';
            if (isDebug) {
                console.log(`[ROS2SharedManager] Initialized with domain ${this.domain || 'default'}, namespace "${this.configuration.namespace}"`);
            }
            // Don't start spinning here - wait for nodes to be created
        } catch (error) {
            console.error('[ROS2SharedManager] Initialization failed:', error);
            throw error;
        }
    }

    /**
     * Create a new ROS2 node
     * @param {string} nodeName - Name of the node
     * @returns {Promise<{nodeId: string, node: Object}>}
     */
    async createNode(nodeName) {
        if (!this.initialized) {
            throw new Error('ROS2SharedManager not initialized. Call initialize() first.');
        }

        try {
            // Apply configured namespace to node name if specified
            let effectiveNodeName = nodeName;
            if (this.configuration.namespace) {
                // Node names get namespace differently than topics
                // For now, we'll let topics/services/actions get the namespace
                // Node names typically don't use namespace prefix
                effectiveNodeName = nodeName;
            }
            
            // Create node using the correct rclnodejs API
            const node = rclnodejs.createNode(effectiveNodeName);
            
            const nodeId = `node_${++this.nodeCounter}_${nodeName}`;
            this.nodes.set(nodeId, {
                node,
                name: nodeName,
                namespace: this.configuration.namespace,
                fullName: effectiveNodeName,
                createdAt: new Date()
            });

            // Start spinning if this is the first node
            if (this.nodes.size === 1) {
                const isDebug = process.env.ROS2_MANAGER_DEBUG === 'true';
                if (isDebug) {
                    console.log('[ROS2SharedManager] Starting spinning for ROS2 nodes...');
                    console.log('[ROS2SharedManager] Debug - initialized:', this.initialized, 'spinning:', this.spinning);
                }
                // Don't await - let it run in background to avoid blocking node creation
                this.startSpinning().catch(error => {
                    console.error('[ROS2SharedManager] Background spinning failed:', error);
                });
            }

            const isDebug = process.env.ROS2_MANAGER_DEBUG === 'true';
            if (isDebug) {
                console.log(`[ROS2SharedManager] Created node: ${nodeName} (ID: ${nodeId})`);
            }
            return { nodeId, node };
        } catch (error) {
            console.error(`[ROS2SharedManager] Failed to create node ${nodeName}:`, error);
            throw error;
        }
    }

    /**
     * Destroy a specific node
     * @param {string} nodeId - ID of the node to destroy
     */
    destroyNode(nodeId) {
        const nodeInfo = this.nodes.get(nodeId);
        if (nodeInfo) {
            try {
                console.log(`[ROS2SharedManager] Destroying node: ${nodeInfo.fullName} (ID: ${nodeId})`);
                
                // Clean up associated action clients first
                this.cleanupActionClientsForNode(nodeId);
                
                // Mark the node as destroyed before actually destroying it
                if (nodeInfo.node) {
                    nodeInfo.node._destroyed = true;
                    nodeInfo.node.destroy();
                }
                
                // Remove from our tracking immediately
                this.nodes.delete(nodeId);
                
                console.log(`[ROS2SharedManager] Successfully destroyed node: ${nodeInfo.fullName} (ID: ${nodeId})`);
            } catch (error) {
                console.error(`[ROS2SharedManager] Error destroying node ${nodeId}:`, error);
                // Still remove it from tracking even if destroy failed
                this.nodes.delete(nodeId);
            }
        } else {
            console.log(`[ROS2SharedManager] Node ${nodeId} not found for destruction`);
        }
    }

    /**
     * Get a node by its ID
     * @param {string} nodeId - Node ID to retrieve
     * @returns {Object|null} The ROS2 node or null if not found
     */
    getNode(nodeId) {
        if (!this.initialized) {
            console.warn('[ROS2SharedManager] Manager not initialized');
            return null;
        }
        
        const nodeInfo = this.nodes.get(nodeId);
        return nodeInfo ? nodeInfo.node : null;
    }

    /**
     * Get a node by its name (returns the first match)
     * @param {string} nodeName - Node name to search for
     * @returns {Object|null} The ROS2 node or null if not found
     */
    getNodeByName(nodeName) {
        if (!this.initialized) {
            console.warn('[ROS2SharedManager] Manager not initialized');
            return null;
        }
        
        for (const [nodeId, nodeInfo] of this.nodes.entries()) {
            if (nodeInfo.name === nodeName) {
                return nodeInfo.node;
            }
        }
        
        return null;
    }

    /**
     * Get all active nodes
     * @returns {Array} Array of node information objects
     */
    getActiveNodes() {
        return Array.from(this.nodes.entries()).map(([nodeId, nodeInfo]) => ({
            nodeId,
            name: nodeInfo.name,
            namespace: nodeInfo.namespace,
            fullName: nodeInfo.fullName,
            createdAt: nodeInfo.createdAt
        }));
    }

    /**
     * Start spinning (processing callbacks) for all nodes
     */
    async startSpinning() {
        // Only log debug info if ROS2_MANAGER_DEBUG is enabled
        if (process.env.ROS2_MANAGER_DEBUG) {
            console.log('[ROS2SharedManager] startSpinning called - initialized:', this.initialized, 'spinning:', this.spinning, 'spinStarted:', this.spinStarted);
        }
        
        if (this.initialized && !this.spinning && !this.spinStarted) {
            try {
                // Only log debug info if ROS2_MANAGER_DEBUG is enabled
                if (process.env.ROS2_MANAGER_DEBUG) {
                    console.log('[ROS2SharedManager] Attempting to start ROS2 spinning...');
                }
                this.spinStarted = true; // Mark that we've attempted to start
                
                // Add a longer delay to ensure rclnodejs context is fully ready
                await new Promise(resolve => setTimeout(resolve, 200));
                
                // Try different spinning approaches
                console.log('[ROS2SharedManager] Available nodes:', this.nodes.size);
                console.log('[ROS2SharedManager] Node names:', Array.from(this.nodes.values()).map(n => n.name));
                
                // Use multi-node spinning approach (spinOnce per node)
                // Note: rclnodejs.spin() is not reliable, so we use the per-node approach
                console.log('[ROS2SharedManager] Starting multi-node spinning approach');
                this.startMultiNodeSpinning();
                
                this.spinning = true;
                console.log('[ROS2SharedManager] Started spinning');
            } catch (error) {
                console.error('[ROS2SharedManager] Failed to start spinning:', error);
                this.spinStarted = false; // Reset flag on failure
                throw error;
            }
        } else if (!this.initialized) {
            console.warn('[ROS2SharedManager] Cannot start spinning - not initialized yet');
        } else if (this.spinning) {
            console.log('[ROS2SharedManager] Already spinning');
        } else if (this.spinStarted) {
            console.log('[ROS2SharedManager] Spin already attempted');
        }
    }

    /**
     * Start spinning for multiple nodes using spinOnce approach
     */
    startMultiNodeSpinning() {
        console.log('[ROS2SharedManager] Starting multi-node spinning for', this.nodes.size, 'nodes');
        
        // Don't start spinning if there are no nodes
        if (this.nodes.size === 0) {
            console.log('[ROS2SharedManager] No nodes to spin, skipping multi-node spinning');
            return;
        }
        
        // Create a spinning interval that processes all nodes
        this.spinInterval = setInterval(() => {
            try {
                let totalNodes = this.nodes.size;
                let successfulSpins = 0;
                let skippedSpins = 0;
                
                // Use spinOnce to process callbacks for all nodes
                for (const [nodeId, nodeInfo] of this.nodes) {
                    // Check if node and nodeInfo are still valid
                    if (nodeInfo && nodeInfo.node && typeof nodeInfo.node.spin === 'function') {
                        
                        // Check if node has been explicitly destroyed
                        if (nodeInfo.node._destroyed) {
                            console.log(`[ROS2SharedManager] Removing explicitly destroyed node: ${nodeInfo.name} (ID: ${nodeId})`);
                            this.nodes.delete(nodeId);
                            continue;
                        }
                        
                        // Simple backoff for failing nodes
                        const now = Date.now();
                        if (nodeInfo.consecutiveFailures > 0) {
                            // Conservative backoff: 100ms, 200ms, 500ms, max 1000ms
                            let backoffTime;
                            if (nodeInfo.consecutiveFailures < 3) {
                                backoffTime = 100 * nodeInfo.consecutiveFailures; // 100ms, 200ms
                            } else if (nodeInfo.consecutiveFailures < 10) {
                                backoffTime = 500; // 500ms
                            } else {
                                backoffTime = 1000; // Max 1 second
                            }
                            
                            if (nodeInfo.lastFailureTime && (now - nodeInfo.lastFailureTime) < backoffTime) {
                                skippedSpins++;
                                continue; // Skip this node for now
                            }
                        }
                        
                        try {
                            // Process callbacks for this node
                            rclnodejs.spinOnce(nodeInfo.node, 1); // Very short timeout
                            
                            // Success! Reset failure tracking
                            if (nodeInfo.consecutiveFailures > 0) {
                                console.log(`[ROS2SharedManager] Node ${nodeInfo.name} spinning recovered after ${nodeInfo.consecutiveFailures} failures`);
                                nodeInfo.consecutiveFailures = 0;
                                nodeInfo.lastFailureTime = 0;
                            }
                            successfulSpins++;
                            
                        } catch (spinError) {
                            // Handle specific errors gracefully without stopping spinning
                            if (spinError.message.includes('reinterpret') || spinError.message.includes('nullptr')) {
                                // These are ROS2 internal errors - likely node corruption or timing issues
                                // Track failures and use backoff to avoid overwhelming the system
                                
                                nodeInfo.consecutiveFailures = (nodeInfo.consecutiveFailures || 0) + 1;
                                nodeInfo.lastFailureTime = now;
                                
                                // Log occasionally to avoid spam
                                if (nodeInfo.consecutiveFailures === 1 || nodeInfo.consecutiveFailures % 10 === 0) {
                                    console.log(`[ROS2SharedManager] Node ${nodeInfo.name} has ${nodeInfo.consecutiveFailures} consecutive spin failures (ROS2 internal error)`);
                                }
                                
                                skippedSpins++;
                                
                                // If a node fails too much, it might be genuinely broken
                                if (nodeInfo.consecutiveFailures > 50) { // Very high threshold
                                    console.warn(`[ROS2SharedManager] Node ${nodeInfo.name} has failed ${nodeInfo.consecutiveFailures} times, may be corrupted`);
                                    // Still don't remove it - let it try to recover
                                }
                                
                            } else if (spinError.message.includes('destroyed') || spinError.message.includes('invalid')) {
                                // Node is actually destroyed, remove it
                                console.log(`[ROS2SharedManager] Node ${nodeInfo.name} is destroyed/invalid, removing: ${spinError.message}`);
                                this.nodes.delete(nodeId);
                                
                            } else {
                                // Other spinning errors - handle gracefully  
                                nodeInfo.consecutiveFailures = (nodeInfo.consecutiveFailures || 0) + 1;
                                nodeInfo.lastFailureTime = now;
                                
                                // Log more frequently for first few failures, then reduce frequency
                                if (nodeInfo.consecutiveFailures <= 5 || nodeInfo.consecutiveFailures % 20 === 0) {
                                    console.log(`[ROS2SharedManager] Node ${nodeInfo.name} has ${nodeInfo.consecutiveFailures} consecutive spin failures`);
                                }
                                
                                skippedSpins++;
                                
                                // Warning at 30 failures
                                if (nodeInfo.consecutiveFailures === 30) {
                                    console.warn(`[ROS2SharedManager] Node ${nodeInfo.name} has failed ${nodeInfo.consecutiveFailures} times - may indicate system issues`);
                                }
                            }
                        }
                    } else {
                        // Invalid node entry, remove it
                        const reason = !nodeInfo ? 'no nodeInfo' : 
                                     !nodeInfo.node ? 'no node' : 
                                     'node missing spin function';
                        console.log(`[ROS2SharedManager] Removing invalid node entry: ${nodeId} (${reason})`);
                        this.nodes.delete(nodeId);
                    }
                }
                
                // Health logging: Only log when status changes or during issues
                const healthInterval = (skippedSpins > 0) ? 10000 : 60000; // Reduced frequency: 1 minute for healthy, 10s for issues
                if (!this.lastHealthLog || (Date.now() - this.lastHealthLog) > healthInterval) {
                    if (totalNodes > 0) {
                        // Only log if status changed or there are issues
                        const currentStatus = `${successfulSpins}/${totalNodes}`;
                        const hasIssues = skippedSpins > 0;
                        
                        if (hasIssues || this.lastHealthStatus !== currentStatus) {
                            let message = `[ROS2SharedManager] Spinning health: ${successfulSpins}/${totalNodes} nodes spinning successfully`;
                            
                            // Add details only when there are issues
                            if (skippedSpins > 0) {
                                message += `, ${skippedSpins} temporarily skipped`;
                                const skippedDetails = [];
                                for (const [nodeId, nodeInfo] of this.nodes) {
                                    if (nodeInfo.consecutiveFailures > 0) {
                                        let backoffTime;
                                        if (nodeInfo.consecutiveFailures < 3) {
                                            backoffTime = 100 * nodeInfo.consecutiveFailures;
                                        } else if (nodeInfo.consecutiveFailures < 10) {
                                            backoffTime = 500;
                                        } else {
                                            backoffTime = 1000;
                                        }
                                        skippedDetails.push(`${nodeInfo.name}(${nodeInfo.consecutiveFailures} fails, ${backoffTime}ms backoff)`);
                                    }
                                }
                                if (skippedDetails.length > 0) {
                                    message += ` - Details: ${skippedDetails.join(', ')}`;
                                }
                            }
                            
                            console.log(message);
                            this.lastHealthStatus = currentStatus;
                            this.lastHealthLog = Date.now();
                        }
                    }
                }
                
            } catch (error) {
                console.error('[ROS2SharedManager] Error in spinning interval:', error);
            }
        }, 50); // 50ms interval - maintains responsive callback processing
        
        console.log('[ROS2SharedManager] Multi-node spinning started - callbacks will be processed continuously');
    }    /**
     * Stop spinning
     */
    stopSpinning() {
        if (this.spinning) {
            // Clear the multi-node spinning interval if it exists
            if (this.spinInterval) {
                clearInterval(this.spinInterval);
                this.spinInterval = null;
                console.log('[ROS2SharedManager] Cleared multi-node spinning interval');
            }
            
            // Note: rclnodejs doesn't have a direct stop spin method for global spin
            // The spinning will stop when shutdown is called
            this.spinning = false;
            this.spinStarted = false;
            console.log('[ROS2SharedManager] Stopped spinning');
        }
    }

    /**
     * Register a callback to be called during shutdown
     * @param {Function} callback - Shutdown callback
     */
    onShutdown(callback) {
        this.shutdownCallbacks.add(callback);
    }

    /**
     * Unregister a shutdown callback
     * @param {Function} callback - Shutdown callback to remove
     */
    offShutdown(callback) {
        this.shutdownCallbacks.delete(callback);
    }

    /**
     * Shutdown the manager (decrements reference count)
     * Only performs actual shutdown when reference count reaches 0
     */
    async shutdown() {
        if (!this.initialized) {
            return;
        }

        this.referenceCount--;
        console.log(`[ROS2SharedManager] Shutdown called, reference count: ${this.referenceCount}`);

        if (this.referenceCount <= 0) {
            await this._doShutdown();
        }
    }

    async _doShutdown() {
        console.log('[ROS2SharedManager] Performing shutdown...');

        // Call shutdown callbacks
        for (const callback of this.shutdownCallbacks) {
            try {
                await callback();
            } catch (error) {
                console.error('[ROS2SharedManager] Error in shutdown callback:', error);
            }
        }

        // Stop spinning
        this.stopSpinning();

        // Destroy all action clients first
        console.log(`[ROS2SharedManager] Destroying ${this.actionClients.size} action clients...`);
        for (const [actionClientId] of this.actionClients) {
            this.destroyActionClient(actionClientId);
        }

        // Destroy all nodes
        for (const [nodeId] of this.nodes) {
            this.destroyNode(nodeId);
        }

        // Shutdown ROS2
        try {
            await rclnodejs.shutdown();
            console.log('[ROS2SharedManager] ROS2 shutdown complete');
        } catch (error) {
            console.error('[ROS2SharedManager] Error during ROS2 shutdown:', error);
        }

        // Reset state
        this.initialized = false;
        this.initializing = false;
        this.context = null;
        this.nodes.clear();
        this.nodeCounter = 0;
        this.actionClients.clear(); // Clear action clients map
        this.actionClientCounter = 0; // Reset action client counter
        this.referenceCount = 0;
        this.shutdownCallbacks.clear();
        this.initializationPromise = null;
    }

    /**
     * Force shutdown regardless of reference count
     * Use with caution - only for emergency cleanup
     */
    async forceShutdown() {
        console.log('[ROS2SharedManager] Force shutdown requested');
        this.referenceCount = 0;
        await this._doShutdown();
    }

    /**
     * Get current manager status
     * @returns {Object} Status information
     */
    getStatus() {
        return {
            initialized: this.initialized,
            initializing: this.initializing,
            nodeCount: this.nodes.size,
            referenceCount: this.referenceCount,
            domain: this.domain
        };
    }

    /**
     * Get the rclnodejs instance used by this manager
     * This ensures all ROS2 operations use the same context to prevent instance mismatches
     * @returns {Object} rclnodejs instance
     * @throws {Error} If rclnodejs is not available
     */
    getRclnodejs() {
        if (!this.available) {
            throw new Error('rclnodejs is not available. Please ensure rclnodejs is properly installed.');
        }
        
        if (!rclnodejs) {
            throw new Error('rclnodejs instance is not loaded in SharedManager');
        }
        
        console.log('[ROS2SharedManager] Providing rclnodejs instance to external consumer');
        return rclnodejs;
    }

    /**
     * Check if ROS2 is available
     * @returns {boolean}
     */
    static isROS2Available() {
        try {
            require('rclnodejs');
            return true;
        } catch (error) {
            return false;
        }
    }

    /**
     * Get the current domain ID
     * @returns {number|null} Current domain ID
     */
    getDomain() {
        return this.domain;
    }

    /**
     * Create an action client
     * @param {string} nodeId - Node ID to create action client on
     * @param {string} actionType - ROS2 action type (e.g., 'rmf_task_msgs/action/DynamicEvent')
     * @param {string} actionName - Action name/topic
     * @param {Object} callbacks - Callback functions
     * @param {Function} callbacks.onFeedback - Feedback callback
     * @param {Function} callbacks.onResult - Result callback
     * @returns {Promise<string>} Action client ID
     */
    async createActionClient(nodeId, actionType, actionName, callbacks = {}) {
        const nodeInfo = this.nodes.get(nodeId);
        if (!nodeInfo) {
            throw new Error(`Node ${nodeId} not found`);
        }

        try {
            const rclnodejs = require('rclnodejs');
            const { ActionClient } = rclnodejs;
            
            console.log(`[ROS2SharedManager] Creating action client for ${actionType} at ${actionName}`);
            
            const actionClient = new ActionClient(nodeInfo.node, actionType, actionName);
            
            const actionClientId = `action_${++this.actionClientCounter}_${nodeId}`;
            
            this.actionClients.set(actionClientId, {
                client: actionClient,
                nodeId: nodeId,
                actionType: actionType,
                actionName: actionName,
                callbacks: callbacks
            });
            
            console.log(`[ROS2SharedManager] Action client created: ${actionClientId}`);
            return actionClientId;
            
        } catch (error) {
            console.error(`[ROS2SharedManager] Failed to create action client:`, error);
            throw error;
        }
    }

    /**
     * Send a goal to an action client
     * @param {string} actionClientId - Action client ID
     * @param {Object} goal - Goal data
     * @param {Function} feedbackCallback - Optional feedback callback (overrides default)
     * @returns {Promise<Object>} Goal result
     */
    async sendGoal(actionClientId, goal, feedbackCallback = null) {
        const actionInfo = this.actionClients.get(actionClientId);
        if (!actionInfo) {
            throw new Error(`Action client ${actionClientId} not found`);
        }

        try {
            console.log(`[ROS2SharedManager] Sending goal to ${actionClientId}:`, goal);
            
            // Use provided feedback callback or default from creation
            const feedbackFn = feedbackCallback || actionInfo.callbacks.onFeedback;
            
            const goalHandle = await actionInfo.client.sendGoal(goal, feedbackFn);
            
            if (!goalHandle.isAccepted()) {
                throw new Error('Goal was rejected by action server');
            }
            
            console.log(`[ROS2SharedManager] Goal accepted for ${actionClientId}`);
            
            // Wait for result - spinning continues normally to process callbacks
            const result = await goalHandle.getResult();
            
            // Call result callback if provided
            if (actionInfo.callbacks.onResult) {
                actionInfo.callbacks.onResult({
                    success: goalHandle.isSucceeded(),
                    canceled: goalHandle.isCanceled(),
                    aborted: goalHandle.isAborted(),
                    result: result,
                    goalHandle: goalHandle
                });
            }
            
            console.log(`[ROS2SharedManager] Goal completed for ${actionClientId}`);
            
            return {
                success: goalHandle.isSucceeded(),
                canceled: goalHandle.isCanceled(),
                aborted: goalHandle.isAborted(),
                result: result,
                goalHandle: goalHandle
            };
            
        } catch (error) {
            console.error(`[ROS2SharedManager] Goal failed for ${actionClientId}:`, error);
            throw error;
        }
    }

    /**
     * Check if action server is available
     * @param {string} actionClientId - Action client ID
     * @returns {boolean} True if action server is available
     */
    isActionServerAvailable(actionClientId) {
        const actionInfo = this.actionClients.get(actionClientId);
        if (!actionInfo) {
            return false;
        }
        
        try {
            return actionInfo.client.isActionServerAvailable();
        } catch (error) {
            console.warn(`[ROS2SharedManager] Error checking action server availability:`, error);
            return false;
        }
    }

    /**
     * Destroy an action client
     * @param {string} actionClientId - Action client ID
     */
    destroyActionClient(actionClientId) {
        const actionInfo = this.actionClients.get(actionClientId);
        if (!actionInfo) {
            console.warn(`[ROS2SharedManager] Action client ${actionClientId} not found for destruction`);
            return;
        }

        try {
            console.log(`[ROS2SharedManager] Destroying action client: ${actionClientId}`);
            actionInfo.client.destroy();
            this.actionClients.delete(actionClientId);
        } catch (error) {
            console.warn(`[ROS2SharedManager] Error destroying action client ${actionClientId}:`, error);
            // Remove from map anyway
            this.actionClients.delete(actionClientId);
        }
    }

    /**
     * Clean up action clients for a specific node
     * @param {string} nodeId - Node ID
     */
    cleanupActionClientsForNode(nodeId) {
        const clientsToRemove = [];
        
        for (const [actionClientId, actionInfo] of this.actionClients) {
            if (actionInfo.nodeId === nodeId) {
                clientsToRemove.push(actionClientId);
            }
        }
        
        for (const actionClientId of clientsToRemove) {
            this.destroyActionClient(actionClientId);
        }
        
        if (clientsToRemove.length > 0) {
            console.log(`[ROS2SharedManager] Cleaned up ${clientsToRemove.length} action clients for node ${nodeId}`);
        }
    }
}

// Singleton instance
let sharedManager = null;

/**
 * Get the shared ROS2 manager instance
 * @returns {ROS2SharedManager}
 */
function getSharedManager() {
    if (!sharedManager) {
        sharedManager = new ROS2SharedManager();
    }
    return sharedManager;
}

/**
 * Initialize the shared ROS2 context
 * @param {Object} options - Initialization options
 * @returns {Promise<ROS2SharedManager>}
 */
async function initializeSharedROS2(options = {}) {
    const manager = getSharedManager();
    await manager.initialize(options);
    return manager;
}

/**
 * Shutdown the shared ROS2 context
 */
async function shutdownSharedROS2() {
    if (sharedManager) {
        await sharedManager.shutdown();
    }
}

// Create singleton instance
const ros2SharedManager = new ROS2SharedManager();

module.exports = {
    ROS2SharedManager,
    ros2SharedManager, // Export singleton instance
    getSharedManager,
    initializeSharedROS2,
    shutdownSharedROS2
};
