// rclnodejs/generated/controller_manager_msgs/controller_manager_msgs__msg__ControllerManagerActivity.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const NamedLifecycleStateWrapper = require('../../generated/controller_manager_msgs/controller_manager_msgs__msg__NamedLifecycleState.js');
const ControllerManagerActivityRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  controllers: NamedLifecycleStateWrapper.refObjectArrayType,
  hardware_components: NamedLifecycleStateWrapper.refObjectArrayType,
});
const ControllerManagerActivityRefArray = ArrayType(ControllerManagerActivityRefStruct);
const ControllerManagerActivityRefStructArray = StructType({
  data: ControllerManagerActivityRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class ControllerManagerActivityWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new ControllerManagerActivityRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.controllers = NamedLifecycleStateWrapper.createArray();
      this._wrapperFields.controllers.copy(other._wrapperFields.controllers);
      this._wrapperFields.hardware_components = NamedLifecycleStateWrapper.createArray();
      this._wrapperFields.hardware_components.copy(other._wrapperFields.hardware_components);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new ControllerManagerActivityRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.controllers = NamedLifecycleStateWrapper.createArray();
    this._wrapperFields.hardware_components = NamedLifecycleStateWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new ControllerManagerActivityWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ControllerManagerActivityArrayWrapper();
  }
  static get ArrayType() {
    return ControllerManagerActivityArrayWrapper;
  }
  static get refObjectArrayType() {
    return ControllerManagerActivityRefStructArray;
  }
  static get refObjectType() {
    return ControllerManagerActivityRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.header.freeze(own);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (own) {
      this._wrapperFields.controllers.fill([]);
    }
    this._wrapperFields.controllers.freeze(own);
    this._refObject.controllers = this._wrapperFields.controllers.refObject;
    if (own) {
      this._wrapperFields.hardware_components.fill([]);
    }
    this._wrapperFields.hardware_components.freeze(own);
    this._refObject.hardware_components = this._wrapperFields.hardware_components.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.controllers.copyRefObject(refObject.controllers);
    this._wrapperFields.hardware_components.copyRefObject(refObject.hardware_components);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    if (refObject.controllers.size != 0) {
      NamedLifecycleStateWrapper.ArrayType.freeArray(refObject.controllers);
      if (!NamedLifecycleStateWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.controllers, NamedLifecycleStateWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.hardware_components.size != 0) {
      NamedLifecycleStateWrapper.ArrayType.freeArray(refObject.hardware_components);
      if (!NamedLifecycleStateWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.hardware_components, NamedLifecycleStateWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    ControllerManagerActivityWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'controller_manager_msgs', subFolder: 'msg', interfaceName: 'ControllerManagerActivity'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get controllers() {
    return this._wrapperFields.controllers;
  }
  set controllers(value) {
    this._wrapperFields.controllers.fill(value);
  }
  get hardware_components() {
    return this._wrapperFields.hardware_components;
  }
  set hardware_components(value) {
    this._wrapperFields.hardware_components.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new ControllerManagerActivityRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.controllers.copyRefObject(this._refObject.controllers);
    this._wrapperFields.hardware_components.copyRefObject(this._refObject.hardware_components);
  }
  copy(other) {
    this._refObject = new ControllerManagerActivityRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.controllers.copy(other._wrapperFields.controllers);
    this._wrapperFields.hardware_components.copy(other._wrapperFields.hardware_components);
  }
  static get classType() {
    return ControllerManagerActivityWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"controllers","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"controller_manager_msgs","type":"NamedLifecycleState","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"hardware_components","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"controller_manager_msgs","type":"NamedLifecycleState","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"controller_manager_msgs","type":"ControllerManagerActivity","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ControllerManagerActivity"};
  }
  hasMember(name) {
    let memberNames = ["header","controllers","hardware_components"];
    return memberNames.indexOf(name) !== -1;
  }
}
class ControllerManagerActivityArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ControllerManagerActivityWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ControllerManagerActivityWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ControllerManagerActivityArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ControllerManagerActivityArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new ControllerManagerActivityRefArray(size);
    this._refObject = new ControllerManagerActivityRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ControllerManagerActivityWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ControllerManagerActivityArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ControllerManagerActivityArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ControllerManagerActivityWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ControllerManagerActivityWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ControllerManagerActivityArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = ControllerManagerActivityWrapper;
