// rclnodejs/generated/rcl_interfaces/rcl_interfaces__msg__LoggerLevel.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const LoggerLevelRefStruct = StructType({
  name: primitiveTypes.string,
  level: primitiveTypes.uint32,
});
const LoggerLevelRefArray = ArrayType(LoggerLevelRefStruct);
const LoggerLevelRefStructArray = StructType({
  data: LoggerLevelRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class LoggerLevelWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new LoggerLevelRefStruct(other._refObject.toObject());
      this._wrapperFields.name =  new StringWrapper(other._wrapperFields.name);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new LoggerLevelRefStruct();
    this._wrapperFields.name = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new LoggerLevelWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new LoggerLevelArrayWrapper();
  }
  static get ArrayType() {
    return LoggerLevelArrayWrapper;
  }
  static get refObjectArrayType() {
    return LoggerLevelRefStructArray;
  }
  static get refObjectType() {
    return LoggerLevelRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.name.freeze(own);
    }
    this._refObject.name = this._wrapperFields.name.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.name.data = refObject.name.data;
    this._refObject.level = refObject.level;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.name);
  }
  static destroyRawROS(msg) {
    LoggerLevelWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rcl_interfaces', subFolder: 'msg', interfaceName: 'LoggerLevel'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get name() {
    return this._wrapperFields.name.data;
  }
  set name(value) {
    this._wrapperFields.name.data = value;
  }
  get level() {
    return this._refObject.level;
  }
  set level(value) {
    this._refObject.level = value;
  }
  copyRefObject(refObject) {
    this._refObject = new LoggerLevelRefStruct(refObject.toObject());
    this._wrapperFields.name.copyRefObject(this._refObject.name);
  }
  copy(other) {
    this._refObject = new LoggerLevelRefStruct(other._refObject.toObject());
    this._wrapperFields.name.copy(other._wrapperFields.name);
  }
  static get classType() {
    return LoggerLevelWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"LOG_LEVEL_UNKNOWN","value":0},{"type":"uint8","name":"LOG_LEVEL_DEBUG","value":10},{"type":"uint8","name":"LOG_LEVEL_INFO","value":20},{"type":"uint8","name":"LOG_LEVEL_WARN","value":30},{"type":"uint8","name":"LOG_LEVEL_ERROR","value":40},{"type":"uint8","name":"LOG_LEVEL_FATAL","value":50}],"fields":[{"name":"name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"level","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rcl_interfaces","type":"LoggerLevel","stringUpperBound":null,"isPrimitiveType":false},"msgName":"LoggerLevel"};
  }
  hasMember(name) {
    let memberNames = ["name","level"];
    return memberNames.indexOf(name) !== -1;
  }
}
class LoggerLevelArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof LoggerLevelWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new LoggerLevelWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LoggerLevelArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LoggerLevelArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new LoggerLevelRefArray(size);
    this._refObject = new LoggerLevelRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new LoggerLevelWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof LoggerLevelArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "LoggerLevelArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      LoggerLevelWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return LoggerLevelWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return LoggerLevelArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
Object.defineProperty(LoggerLevelWrapper, "LOG_LEVEL_UNKNOWN", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LoggerLevelWrapper, "LOG_LEVEL_DEBUG", {value: 10, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LoggerLevelWrapper, "LOG_LEVEL_INFO", {value: 20, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LoggerLevelWrapper, "LOG_LEVEL_WARN", {value: 30, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LoggerLevelWrapper, "LOG_LEVEL_ERROR", {value: 40, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LoggerLevelWrapper, "LOG_LEVEL_FATAL", {value: 50, writable: false, enumerable: true, configurable: true});
module.exports = LoggerLevelWrapper;
