// rclnodejs/generated/rcl_interfaces/rcl_interfaces__srv__SetLoggerLevels_Response.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const SetLoggerLevelsResultWrapper = require('../../generated/rcl_interfaces/rcl_interfaces__msg__SetLoggerLevelsResult.js');
const SetLoggerLevels_ResponseRefStruct = StructType({
  results: SetLoggerLevelsResultWrapper.refObjectArrayType,
});
const SetLoggerLevels_ResponseRefArray = ArrayType(SetLoggerLevels_ResponseRefStruct);
const SetLoggerLevels_ResponseRefStructArray = StructType({
  data: SetLoggerLevels_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class SetLoggerLevels_ResponseWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new SetLoggerLevels_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.results = SetLoggerLevelsResultWrapper.createArray();
      this._wrapperFields.results.copy(other._wrapperFields.results);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new SetLoggerLevels_ResponseRefStruct();
    this._wrapperFields.results = SetLoggerLevelsResultWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new SetLoggerLevels_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SetLoggerLevels_ResponseArrayWrapper();
  }
  static get ArrayType() {
    return SetLoggerLevels_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return SetLoggerLevels_ResponseRefStructArray;
  }
  static get refObjectType() {
    return SetLoggerLevels_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.results.fill([]);
    }
    this._wrapperFields.results.freeze(own);
    this._refObject.results = this._wrapperFields.results.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.results.copyRefObject(refObject.results);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.results.size != 0) {
      SetLoggerLevelsResultWrapper.ArrayType.freeArray(refObject.results);
      if (!SetLoggerLevelsResultWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.results, SetLoggerLevelsResultWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    SetLoggerLevels_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rcl_interfaces', subFolder: 'srv', interfaceName: 'SetLoggerLevels_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get results() {
    return this._wrapperFields.results;
  }
  set results(value) {
    this._wrapperFields.results.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new SetLoggerLevels_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.results.copyRefObject(this._refObject.results);
  }
  copy(other) {
    this._refObject = new SetLoggerLevels_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.results.copy(other._wrapperFields.results);
  }
  static get classType() {
    return SetLoggerLevels_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"results","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"rcl_interfaces","type":"SetLoggerLevelsResult","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"rcl_interfaces","type":"SetLoggerLevels_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SetLoggerLevels_Response"};
  }
  hasMember(name) {
    let memberNames = ["results"];
    return memberNames.indexOf(name) !== -1;
  }
}
class SetLoggerLevels_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SetLoggerLevels_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SetLoggerLevels_ResponseWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetLoggerLevels_ResponseArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetLoggerLevels_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new SetLoggerLevels_ResponseRefArray(size);
    this._refObject = new SetLoggerLevels_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SetLoggerLevels_ResponseWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SetLoggerLevels_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SetLoggerLevels_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SetLoggerLevels_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SetLoggerLevels_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SetLoggerLevels_ResponseArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = SetLoggerLevels_ResponseWrapper;
