// rclnodejs/generated/moveit_msgs/moveit_msgs__srv__ServoCommandType_Request.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const ServoCommandType_RequestRefStruct = StructType({
  command_type: primitiveTypes.int8,
});
const ServoCommandType_RequestRefArray = ArrayType(ServoCommandType_RequestRefStruct);
const ServoCommandType_RequestRefStructArray = StructType({
  data: ServoCommandType_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class ServoCommandType_RequestWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new ServoCommandType_RequestRefStruct(other._refObject.toObject());
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new ServoCommandType_RequestRefStruct();
  }
  static createFromRefObject(refObject) {
    let self = new ServoCommandType_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ServoCommandType_RequestArrayWrapper();
  }
  static get ArrayType() {
    return ServoCommandType_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return ServoCommandType_RequestRefStructArray;
  }
  static get refObjectType() {
    return ServoCommandType_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.command_type = refObject.command_type;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destroyRawROS(msg) {
    ServoCommandType_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'srv', interfaceName: 'ServoCommandType_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get command_type() {
    return this._refObject.command_type;
  }
  set command_type(value) {
    this._refObject.command_type = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ServoCommandType_RequestRefStruct(refObject.toObject());
  }
  copy(other) {
    this._refObject = new ServoCommandType_RequestRefStruct(other._refObject.toObject());
  }
  static get classType() {
    return ServoCommandType_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"int8","name":"JOINT_JOG","value":0},{"type":"int8","name":"TWIST","value":1},{"type":"int8","name":"POSE","value":2}],"fields":[{"name":"command_type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"ServoCommandType_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ServoCommandType_Request"};
  }
  hasMember(name) {
    let memberNames = ["command_type"];
    return memberNames.indexOf(name) !== -1;
  }
}
class ServoCommandType_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ServoCommandType_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ServoCommandType_RequestWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ServoCommandType_RequestArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ServoCommandType_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new ServoCommandType_RequestRefArray(size);
    this._refObject = new ServoCommandType_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ServoCommandType_RequestWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ServoCommandType_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ServoCommandType_RequestArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ServoCommandType_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ServoCommandType_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ServoCommandType_RequestArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
Object.defineProperty(ServoCommandType_RequestWrapper, "JOINT_JOG", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ServoCommandType_RequestWrapper, "TWIST", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ServoCommandType_RequestWrapper, "POSE", {value: 2, writable: false, enumerable: true, configurable: true});
module.exports = ServoCommandType_RequestWrapper;
