// rclnodejs/generated/moveit_msgs/moveit_msgs__msg__ServoStatus.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const ServoStatusRefStruct = StructType({
  code: primitiveTypes.int8,
  message: primitiveTypes.string,
});
const ServoStatusRefArray = ArrayType(ServoStatusRefStruct);
const ServoStatusRefStructArray = StructType({
  data: ServoStatusRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class ServoStatusWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new ServoStatusRefStruct(other._refObject.toObject());
      this._wrapperFields.message =  new StringWrapper(other._wrapperFields.message);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new ServoStatusRefStruct();
    this._wrapperFields.message = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new ServoStatusWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ServoStatusArrayWrapper();
  }
  static get ArrayType() {
    return ServoStatusArrayWrapper;
  }
  static get refObjectArrayType() {
    return ServoStatusRefStructArray;
  }
  static get refObjectType() {
    return ServoStatusRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.message.freeze(own);
    }
    this._refObject.message = this._wrapperFields.message.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.code = refObject.code;
    this._wrapperFields.message.data = refObject.message.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.message);
  }
  static destroyRawROS(msg) {
    ServoStatusWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'msg', interfaceName: 'ServoStatus'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get code() {
    return this._refObject.code;
  }
  set code(value) {
    this._refObject.code = value;
  }
  get message() {
    return this._wrapperFields.message.data;
  }
  set message(value) {
    this._wrapperFields.message.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ServoStatusRefStruct(refObject.toObject());
    this._wrapperFields.message.copyRefObject(this._refObject.message);
  }
  copy(other) {
    this._refObject = new ServoStatusRefStruct(other._refObject.toObject());
    this._wrapperFields.message.copy(other._wrapperFields.message);
  }
  static get classType() {
    return ServoStatusWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"int8","name":"INVALID","value":-1},{"type":"int8","name":"NO_WARNING","value":0},{"type":"int8","name":"DECELERATE_FOR_APPROACHING_SINGULARITY","value":1},{"type":"int8","name":"HALT_FOR_SINGULARITY","value":2},{"type":"int8","name":"DECELERATE_FOR_LEAVING_SINGULARITY","value":3},{"type":"int8","name":"DECELERATE_FOR_COLLISION","value":4},{"type":"int8","name":"HALT_FOR_COLLISION","value":5},{"type":"int8","name":"JOINT_BOUND","value":6}],"fields":[{"name":"code","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"message","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"ServoStatus","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ServoStatus"};
  }
  hasMember(name) {
    let memberNames = ["code","message"];
    return memberNames.indexOf(name) !== -1;
  }
}
class ServoStatusArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ServoStatusWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ServoStatusWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ServoStatusArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ServoStatusArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new ServoStatusRefArray(size);
    this._refObject = new ServoStatusRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ServoStatusWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ServoStatusArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ServoStatusArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ServoStatusWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ServoStatusWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ServoStatusArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
Object.defineProperty(ServoStatusWrapper, "INVALID", {value: -1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ServoStatusWrapper, "NO_WARNING", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ServoStatusWrapper, "DECELERATE_FOR_APPROACHING_SINGULARITY", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ServoStatusWrapper, "HALT_FOR_SINGULARITY", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ServoStatusWrapper, "DECELERATE_FOR_LEAVING_SINGULARITY", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ServoStatusWrapper, "DECELERATE_FOR_COLLISION", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ServoStatusWrapper, "HALT_FOR_COLLISION", {value: 5, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ServoStatusWrapper, "JOINT_BOUND", {value: 6, writable: false, enumerable: true, configurable: true});
module.exports = ServoStatusWrapper;
