// rclnodejs/generated/ur_dashboard_msgs/ur_dashboard_msgs__action__SetMode_Feedback.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const SetMode_FeedbackRefStruct = StructType({
  current_robot_mode: primitiveTypes.int8,
  current_safety_mode: primitiveTypes.int8,
});
const SetMode_FeedbackRefArray = ArrayType(SetMode_FeedbackRefStruct);
const SetMode_FeedbackRefStructArray = StructType({
  data: SetMode_FeedbackRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class SetMode_FeedbackWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new SetMode_FeedbackRefStruct(other._refObject.toObject());
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new SetMode_FeedbackRefStruct();
  }
  static createFromRefObject(refObject) {
    let self = new SetMode_FeedbackWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SetMode_FeedbackArrayWrapper();
  }
  static get ArrayType() {
    return SetMode_FeedbackArrayWrapper;
  }
  static get refObjectArrayType() {
    return SetMode_FeedbackRefStructArray;
  }
  static get refObjectType() {
    return SetMode_FeedbackRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.current_robot_mode = refObject.current_robot_mode;
    this._refObject.current_safety_mode = refObject.current_safety_mode;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destroyRawROS(msg) {
    SetMode_FeedbackWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'ur_dashboard_msgs', subFolder: 'action', interfaceName: 'SetMode_Feedback'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get current_robot_mode() {
    return this._refObject.current_robot_mode;
  }
  set current_robot_mode(value) {
    this._refObject.current_robot_mode = value;
  }
  get current_safety_mode() {
    return this._refObject.current_safety_mode;
  }
  set current_safety_mode(value) {
    this._refObject.current_safety_mode = value;
  }
  copyRefObject(refObject) {
    this._refObject = new SetMode_FeedbackRefStruct(refObject.toObject());
  }
  copy(other) {
    this._refObject = new SetMode_FeedbackRefStruct(other._refObject.toObject());
  }
  static get classType() {
    return SetMode_FeedbackWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"current_robot_mode","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"current_safety_mode","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"ur_dashboard_msgs","type":"SetMode_Feedback","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SetMode_Feedback"};
  }
  hasMember(name) {
    let memberNames = ["current_robot_mode","current_safety_mode"];
    return memberNames.indexOf(name) !== -1;
  }
}
class SetMode_FeedbackArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SetMode_FeedbackWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SetMode_FeedbackWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetMode_FeedbackArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetMode_FeedbackArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new SetMode_FeedbackRefArray(size);
    this._refObject = new SetMode_FeedbackRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SetMode_FeedbackWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SetMode_FeedbackArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SetMode_FeedbackArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SetMode_FeedbackWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SetMode_FeedbackWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SetMode_FeedbackArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = SetMode_FeedbackWrapper;
