// rclnodejs/generated/ur_dashboard_msgs/ur_dashboard_msgs__msg__SafetyMode.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const SafetyModeRefStruct = StructType({
  mode: primitiveTypes.uint8,
});
const SafetyModeRefArray = ArrayType(SafetyModeRefStruct);
const SafetyModeRefStructArray = StructType({
  data: SafetyModeRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class SafetyModeWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new SafetyModeRefStruct(other._refObject.toObject());
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new SafetyModeRefStruct();
  }
  static createFromRefObject(refObject) {
    let self = new SafetyModeWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SafetyModeArrayWrapper();
  }
  static get ArrayType() {
    return SafetyModeArrayWrapper;
  }
  static get refObjectArrayType() {
    return SafetyModeRefStructArray;
  }
  static get refObjectType() {
    return SafetyModeRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.mode = refObject.mode;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destroyRawROS(msg) {
    SafetyModeWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'ur_dashboard_msgs', subFolder: 'msg', interfaceName: 'SafetyMode'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get mode() {
    return this._refObject.mode;
  }
  set mode(value) {
    this._refObject.mode = value;
  }
  copyRefObject(refObject) {
    this._refObject = new SafetyModeRefStruct(refObject.toObject());
  }
  copy(other) {
    this._refObject = new SafetyModeRefStruct(other._refObject.toObject());
  }
  static get classType() {
    return SafetyModeWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"NORMAL","value":1},{"type":"uint8","name":"REDUCED","value":2},{"type":"uint8","name":"PROTECTIVE_STOP","value":3},{"type":"uint8","name":"RECOVERY","value":4},{"type":"uint8","name":"SAFEGUARD_STOP","value":5},{"type":"uint8","name":"SYSTEM_EMERGENCY_STOP","value":6},{"type":"uint8","name":"ROBOT_EMERGENCY_STOP","value":7},{"type":"uint8","name":"VIOLATION","value":8},{"type":"uint8","name":"FAULT","value":9},{"type":"uint8","name":"VALIDATE_JOINT_ID","value":10},{"type":"uint8","name":"UNDEFINED_SAFETY_MODE","value":11},{"type":"uint8","name":"AUTOMATIC_MODE_SAFEGUARD_STOP","value":12},{"type":"uint8","name":"SYSTEM_THREE_POSITION_ENABLING_STOP","value":13}],"fields":[{"name":"mode","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"ur_dashboard_msgs","type":"SafetyMode","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SafetyMode"};
  }
  hasMember(name) {
    let memberNames = ["mode"];
    return memberNames.indexOf(name) !== -1;
  }
}
class SafetyModeArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SafetyModeWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SafetyModeWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SafetyModeArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SafetyModeArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new SafetyModeRefArray(size);
    this._refObject = new SafetyModeRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SafetyModeWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SafetyModeArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SafetyModeArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SafetyModeWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SafetyModeWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SafetyModeArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
Object.defineProperty(SafetyModeWrapper, "NORMAL", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SafetyModeWrapper, "REDUCED", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SafetyModeWrapper, "PROTECTIVE_STOP", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SafetyModeWrapper, "RECOVERY", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SafetyModeWrapper, "SAFEGUARD_STOP", {value: 5, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SafetyModeWrapper, "SYSTEM_EMERGENCY_STOP", {value: 6, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SafetyModeWrapper, "ROBOT_EMERGENCY_STOP", {value: 7, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SafetyModeWrapper, "VIOLATION", {value: 8, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SafetyModeWrapper, "FAULT", {value: 9, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SafetyModeWrapper, "VALIDATE_JOINT_ID", {value: 10, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SafetyModeWrapper, "UNDEFINED_SAFETY_MODE", {value: 11, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SafetyModeWrapper, "AUTOMATIC_MODE_SAFEGUARD_STOP", {value: 12, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SafetyModeWrapper, "SYSTEM_THREE_POSITION_ENABLING_STOP", {value: 13, writable: false, enumerable: true, configurable: true});
module.exports = SafetyModeWrapper;
