// rclnodejs/generated/ur_dashboard_msgs/ur_dashboard_msgs__action__SetMode_Goal.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const SetMode_GoalRefStruct = StructType({
  target_robot_mode: primitiveTypes.int8,
  stop_program: primitiveTypes.bool,
  play_program: primitiveTypes.bool,
});
const SetMode_GoalRefArray = ArrayType(SetMode_GoalRefStruct);
const SetMode_GoalRefStructArray = StructType({
  data: SetMode_GoalRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class SetMode_GoalWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new SetMode_GoalRefStruct(other._refObject.toObject());
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new SetMode_GoalRefStruct();
  }
  static createFromRefObject(refObject) {
    let self = new SetMode_GoalWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SetMode_GoalArrayWrapper();
  }
  static get ArrayType() {
    return SetMode_GoalArrayWrapper;
  }
  static get refObjectArrayType() {
    return SetMode_GoalRefStructArray;
  }
  static get refObjectType() {
    return SetMode_GoalRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.target_robot_mode = refObject.target_robot_mode;
    this._refObject.stop_program = refObject.stop_program;
    this._refObject.play_program = refObject.play_program;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destroyRawROS(msg) {
    SetMode_GoalWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'ur_dashboard_msgs', subFolder: 'action', interfaceName: 'SetMode_Goal'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get target_robot_mode() {
    return this._refObject.target_robot_mode;
  }
  set target_robot_mode(value) {
    this._refObject.target_robot_mode = value;
  }
  get stop_program() {
    return this._refObject.stop_program;
  }
  set stop_program(value) {
    this._refObject.stop_program = value;
  }
  get play_program() {
    return this._refObject.play_program;
  }
  set play_program(value) {
    this._refObject.play_program = value;
  }
  copyRefObject(refObject) {
    this._refObject = new SetMode_GoalRefStruct(refObject.toObject());
  }
  copy(other) {
    this._refObject = new SetMode_GoalRefStruct(other._refObject.toObject());
  }
  static get classType() {
    return SetMode_GoalWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"target_robot_mode","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"stop_program","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"play_program","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"ur_dashboard_msgs","type":"SetMode_Goal","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SetMode_Goal"};
  }
  hasMember(name) {
    let memberNames = ["target_robot_mode","stop_program","play_program"];
    return memberNames.indexOf(name) !== -1;
  }
}
class SetMode_GoalArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SetMode_GoalWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SetMode_GoalWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetMode_GoalArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetMode_GoalArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new SetMode_GoalRefArray(size);
    this._refObject = new SetMode_GoalRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SetMode_GoalWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SetMode_GoalArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SetMode_GoalArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SetMode_GoalWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SetMode_GoalWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SetMode_GoalArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = SetMode_GoalWrapper;
